{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [
    {
     "ename": "ImportError",
     "evalue": "No module named rpy2.ipython",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mImportError\u001b[0m                               Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-3-261134f77704>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      3\u001b[0m \u001b[0;32mimport\u001b[0m \u001b[0mpandas\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0mpd\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      4\u001b[0m \u001b[0;32mimport\u001b[0m \u001b[0mnumpy\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0mnp\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 5\u001b[0;31m \u001b[0mget_ipython\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mmagic\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34mu'load_ext rpy2.ipython'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;32m/Users/bryketos/anaconda/lib/python2.7/site-packages/IPython/core/interactiveshell.pyc\u001b[0m in \u001b[0;36mmagic\u001b[0;34m(self, arg_s)\u001b[0m\n\u001b[1;32m   2334\u001b[0m         \u001b[0mmagic_name\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0m_\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mmagic_arg_s\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0marg_s\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mpartition\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m' '\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   2335\u001b[0m         \u001b[0mmagic_name\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mmagic_name\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mlstrip\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mprefilter\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mESC_MAGIC\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 2336\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mrun_line_magic\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mmagic_name\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mmagic_arg_s\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   2337\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   2338\u001b[0m     \u001b[0;31m#-------------------------------------------------------------------------\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/Users/bryketos/anaconda/lib/python2.7/site-packages/IPython/core/interactiveshell.pyc\u001b[0m in \u001b[0;36mrun_line_magic\u001b[0;34m(self, magic_name, line)\u001b[0m\n\u001b[1;32m   2255\u001b[0m                 \u001b[0mkwargs\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'local_ns'\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0msys\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_getframe\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mstack_depth\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mf_locals\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   2256\u001b[0m             \u001b[0;32mwith\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mbuiltin_trap\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 2257\u001b[0;31m                 \u001b[0mresult\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mfn\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   2258\u001b[0m             \u001b[0;32mreturn\u001b[0m \u001b[0mresult\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   2259\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m<decorator-gen-64>\u001b[0m in \u001b[0;36mload_ext\u001b[0;34m(self, module_str)\u001b[0m\n",
      "\u001b[0;32m/Users/bryketos/anaconda/lib/python2.7/site-packages/IPython/core/magic.pyc\u001b[0m in \u001b[0;36m<lambda>\u001b[0;34m(f, *a, **k)\u001b[0m\n\u001b[1;32m    191\u001b[0m     \u001b[0;31m# but it's overkill for just that one bit of state.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    192\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0mmagic_deco\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0marg\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 193\u001b[0;31m         \u001b[0mcall\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;32mlambda\u001b[0m \u001b[0mf\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m*\u001b[0m\u001b[0ma\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mk\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mf\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0ma\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mk\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    194\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    195\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mcallable\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0marg\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/Users/bryketos/anaconda/lib/python2.7/site-packages/IPython/core/magics/extension.pyc\u001b[0m in \u001b[0;36mload_ext\u001b[0;34m(self, module_str)\u001b[0m\n\u001b[1;32m     64\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0mmodule_str\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     65\u001b[0m             \u001b[0;32mraise\u001b[0m \u001b[0mUsageError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'Missing module name.'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 66\u001b[0;31m         \u001b[0mres\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mshell\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mextension_manager\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mload_extension\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mmodule_str\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     67\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     68\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mres\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0;34m'already loaded'\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/Users/bryketos/anaconda/lib/python2.7/site-packages/IPython/core/extensions.pyc\u001b[0m in \u001b[0;36mload_extension\u001b[0;34m(self, module_str)\u001b[0m\n\u001b[1;32m     82\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0mmodule_str\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0;32min\u001b[0m \u001b[0msys\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mmodules\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     83\u001b[0m                 \u001b[0;32mwith\u001b[0m \u001b[0mprepended_to_syspath\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mipython_extension_dir\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 84\u001b[0;31m                     \u001b[0m__import__\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mmodule_str\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     85\u001b[0m             \u001b[0mmod\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0msys\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mmodules\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mmodule_str\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     86\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_call_load_ipython_extension\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mmod\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mImportError\u001b[0m: No module named rpy2.ipython"
     ]
    }
   ],
   "source": [
    "'''Prepare the notebook. Run this cell first.'''\n",
    "import os, re, sys\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "%load_ext rpy2.ipython"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "'''Set parameters.  Run this cell before all others.'''\n",
    "min_maf = 0.01 #Remove variants below this frequency. Set to 0 to keep all variants.\n",
    "max_missing_sample_callrate = 0.03 #Remove samples that have greater than this percent of missing genotypes (\"sample callrate\").\n",
    "max_missing_marker_callrate = 0.03 #Remove SNPs that have greater than this percent of missing calls (\"marker callrate\").\n",
    "max_IBD = 0.125 #Remove samples that have identity by state greater than this coefficient of relationship (first cousins).\n",
    "min_HWE_p = 0.00001 #Remove SNPs with HWE exact test p-values lower than this.\n",
    "min_diffmiss_p = 0.00001 #Remove SNPs with differential missingness test p-values lower than this."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/home/brian/clink/NEIGHBOR\n"
     ]
    }
   ],
   "source": [
    "'''Make analysis directory and upload your data'''\n",
    "\n",
    "import os\n",
    "study_dir = '/home/brian/clink/NEIGHBOR/' #Can change this variable to suit your study.\n",
    "if not os.path.isdir(study_dir): os.mkdir(study_dir)\n",
    "%cd {study_dir}\n",
    "\n",
    "#You can either upload data to the EC2 instance using the key pair you specified at instance launch,\n",
    "#  for instance using scp from a system that has your data,\n",
    "#  or you can fetch the data from this instance using SFTP/rsync/curl/wget/s3\n",
    "\n",
    "#Remember that if you want to pull data from a private S3 bucket, the EC2 instance needs to assume an IAM role,\n",
    "#  and that bucket must be configured for the EC2 instance's IAM role.\n",
    "\n",
    "#In this example, I used scp from a firewalled and secured workstation to securely upload a Plink study to the EC2 instance.\n",
    "#I used a command like this from my workstation:\n",
    "#scp -i my_key_pair.pem my_study.* ec2-user@<IPv4_public_IP>:/gwas/my_study/\n",
    "#  ..where <IPv4_public_IP> is the public IPv4 address of the EC2 instance running this notebook.\n",
    "\n",
    "input_filestem  = 'merged_neighbor' #Set this variable to the Plink filestem of your study.\n",
    "output_filestem = 'clean-' + input_filestem #Generate a new filestem for post-QC data."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "PLINK v1.90b4.1 64-bit (30 Mar 2017)           www.cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2017 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to merged_neighbor.log.\n",
      "Options in effect:\n",
      "  --bfile merged_neighbor\n",
      "  --check-sex\n",
      "  --out merged_neighbor\n",
      "\n",
      "64329 MB RAM detected; reserving 32164 MB for main workspace.\n",
      "528478 variants loaded from .bim file.\n",
      "5183 people (2299 males, 2752 females, 132 ambiguous) loaded from .fam.\n",
      "Ambiguous sex IDs written to merged_neighbor.nosex .\n",
      "5051 phenotype values loaded from .fam.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 5183 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 0%\b\b1%\b\b2%\b\b3%\b\b4%\b\b5%\b\b6%\b\b7%\b\b8%\b\b9%\b\b10%\b\b\b11%\b\b\b12%\b\b\b13%\b\b\b14%\b\b\b15%\b\b\b16%\b\b\b17%\b\b\b18%\b\b\b19%\b\b\b20%\b\b\b21%\b\b\b22%\b\b\b23%\b\b\b24%\b\b\b25%\b\b\b26%\b\b\b27%\b\b\b28%\b\b\b29%\b\b\b30%\b\b\b31%\b\b\b32%\b\b\b33%\b\b\b34%\b\b\b35%\b\b\b36%\b\b\b37%\b\b\b38%\b\b\b39%\b\b\b40%\b\b\b41%\b\b\b42%\b\b\b43%\b\b\b44%\b\b\b45%\b\b\b46%\b\b\b47%\b\b\b48%\b\b\b49%\b\b\b50%\b\b\b51%\b\b\b52%\b\b\b53%\b\b\b54%\b\b\b55%\b\b\b56%\b\b\b57%\b\b\b58%\b\b\b59%\b\b\b60%\b\b\b61%\b\b\b62%\b\b\b63%\b\b\b64%\b\b\b65%\b\b\b66%\b\b\b67%\b\b\b68%\b\b\b69%\b\b\b70%\b\b\b71%\b\b\b72%\b\b\b73%\b\b\b74%\b\b\b75%\b\b\b76%\b\b\b77%\b\b\b78%\b\b\b79%\b\b\b80%\b\b\b81%\b\b\b82%\b\b\b83%\b\b\b84%\b\b\b85%\b\b\b86%\b\b\b87%\b\b\b88%\b\b\b89%\b\b\b90%\b\b\b91%\b\b\b92%\b\b\b93%\b\b\b94%\b\b\b95%\b\b\b96%\b\b\b97%\b\b\b98%\b\b\b99%\b\b\b\b done.\n",
      "Warning: 11557 het. haploid genotypes present (see merged_neighbor.hh ); many\n",
      "commands treat these as missing.\n",
      "Warning: Nonmissing nonmale Y chromosome genotype(s) present; many commands\n",
      "treat these as missing.\n",
      "Total genotyping rate is 0.999482.\n",
      "528478 variants and 5183 people pass filters and QC.\n",
      "Phenotype data is quantitative.\n",
      "--check-sex: 12379 Xchr and 0 Ychr variant(s) scanned, 151 problems detected.\n",
      "Report written to merged_neighbor.sexcheck .\n"
     ]
    }
   ],
   "source": [
    "'''1) Run sex check using plink.'''\n",
    "\n",
    "#This should generate a *.sexcheck file for the study:\n",
    "!plink --bfile $input_filestem --check-sex --out $input_filestem"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of gender discordances detected: ['151']\n",
      "         FID          IID       PEDSEX       SNPSEX       STATUS            F\n",
      "      ME8001       ME8001            0            2      PROBLEM     -0.03751\n",
      "      ME8002       ME8002            0            1      PROBLEM            1\n",
      "      ME8003       ME8003            0            1      PROBLEM            1\n",
      "      ME8004       ME8004            0            2      PROBLEM    -0.005969\n",
      "      ME8005       ME8005            0            2      PROBLEM     -0.04629\n"
     ]
    }
   ],
   "source": [
    "sexcheck_file = input_filestem + '.sexcheck'\n",
    "sexprobs_file = input_filestem + '.sexprobs'\n",
    "\n",
    "!fgrep PROBLEM $sexcheck_file > $sexprobs_file\n",
    "sexproblems = !cat $sexprobs_file | wc -l\n",
    "print(\"Number of gender discordances detected: \" + str(sexproblems))\n",
    "\n",
    "#Save the FID and IID of the individuals that failed qc, if any.\n",
    "#It's okay to generate an empty file here - we'll count the lines of each fail-* file later for a sumamry.\n",
    "!perl -lne 's/^\\s+//; @f = split; print join \"\\t\", (@f[0..1])' $sexprobs_file >fail-sexcheck-qc.txt\n",
    "\n",
    "#At this point, inspect the reasons for any gender discordances.  In this case, some genders are missing in the ped file:\n",
    "!head -1 $sexcheck_file #For the header.\n",
    "!head -5 $sexprobs_file #Check out the first 5 discordances."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "PLINK v1.90b4.1 64-bit (30 Mar 2017)           www.cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2017 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to merged_neighbor.log.\n",
      "Options in effect:\n",
      "  --bfile merged_neighbor\n",
      "  --geno 0.01\n",
      "  --make-bed\n",
      "  --out merged_neighbor\n",
      "\n",
      "64329 MB RAM detected; reserving 32164 MB for main workspace.\n",
      "Note: --make-bed input and output filenames match.  Appending '~' to input\n",
      "filenames.\n",
      "655216 variants loaded from .bim file.\n",
      "5183 people (2299 males, 2752 females, 132 ambiguous) loaded from .fam.\n",
      "Ambiguous sex IDs written to merged_neighbor.nosex .\n",
      "5051 phenotype values loaded from .fam.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 5183 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 0%\b\b1%\b\b2%\b\b3%\b\b4%\b\b5%\b\b6%\b\b7%\b\b8%\b\b9%\b\b10%\b\b\b11%\b\b\b12%\b\b\b13%\b\b\b14%\b\b\b15%\b\b\b16%\b\b\b17%\b\b\b18%\b\b\b19%\b\b\b20%\b\b\b21%\b\b\b22%\b\b\b23%\b\b\b24%\b\b\b25%\b\b\b26%\b\b\b27%\b\b\b28%\b\b\b29%\b\b\b30%\b\b\b31%\b\b\b32%\b\b\b33%\b\b\b34%\b\b\b35%\b\b\b36%\b\b\b37%\b\b\b38%\b\b\b39%\b\b\b40%\b\b\b41%\b\b\b42%\b\b\b43%\b\b\b44%\b\b\b45%\b\b\b46%\b\b\b47%\b\b\b48%\b\b\b49%\b\b\b50%\b\b\b51%\b\b\b52%\b\b\b53%\b\b\b54%\b\b\b55%\b\b\b56%\b\b\b57%\b\b\b58%\b\b\b59%\b\b\b60%\b\b\b61%\b\b\b62%\b\b\b63%\b\b\b64%\b\b\b65%\b\b\b66%\b\b\b67%\b\b\b68%\b\b\b69%\b\b\b70%\b\b\b71%\b\b\b72%\b\b\b73%\b\b\b74%\b\b\b75%\b\b\b76%\b\b\b77%\b\b\b78%\b\b\b79%\b\b\b80%\b\b\b81%\b\b\b82%\b\b\b83%\b\b\b84%\b\b\b85%\b\b\b86%\b\b\b87%\b\b\b88%\b\b\b89%\b\b\b90%\b\b\b91%\b\b\b92%\b\b\b93%\b\b\b94%\b\b\b95%\b\b\b96%\b\b\b97%\b\b\b98%\b\b\b99%\b\b\b\b done.\n",
      "Warning: 11950 het. haploid genotypes present (see merged_neighbor.hh ); many\n",
      "commands treat these as missing.\n",
      "Warning: Nonmissing nonmale Y chromosome genotype(s) present; many commands\n",
      "treat these as missing.\n",
      "Total genotyping rate is 0.849377.\n",
      "107404 variants removed due to missing genotype data (--geno).\n",
      "547812 variants and 5183 people pass filters and QC.\n",
      "Phenotype data is quantitative.\n",
      "--make-bed to merged_neighbor.bed + merged_neighbor.bim + merged_neighbor.fam\n",
      "... 0%\b\b1%\b\b2%\b\b3%\b\b4%\b\b5%\b\b6%\b\b7%\b\b8%\b\b9%\b\b10%\b\b\b11%\b\b\b12%\b\b\b13%\b\b\b14%\b\b\b15%\b\b\b16%\b\b\b17%\b\b\b18%\b\b\b19%\b\b\b20%\b\b\b21%\b\b\b22%\b\b\b23%\b\b\b24%\b\b\b25%\b\b\b26%\b\b\b27%\b\b\b28%\b\b\b29%\b\b\b30%\b\b\b31%\b\b\b32%\b\b\b33%\b\b\b34%\b\b\b35%\b\b\b36%\b\b\b37%\b\b\b38%\b\b\b39%\b\b\b40%\b\b\b41%\b\b\b42%\b\b\b43%\b\b\b44%\b\b\b45%\b\b\b46%\b\b\b47%\b\b\b48%\b\b\b49%\b\b\b50%\b\b\b51%\b\b\b52%\b\b\b53%\b\b\b54%\b\b\b55%\b\b\b56%\b\b\b57%\b\b\b58%\b\b\b59%\b\b\b60%\b\b\b61%\b\b\b62%\b\b\b63%\b\b\b64%\b\b\b65%\b\b\b66%\b\b\b67%\b\b\b68%\b\b\b69%\b\b\b70%\b\b\b71%\b\b\b72%\b\b\b73%\b\b\b74%\b\b\b75%\b\b\b76%\b\b\b77%\b\b\b78%\b\b\b79%\b\b\b80%\b\b\b81%\b\b\b82%\b\b\b83%\b\b\b84%\b\b\b85%\b\b\b86%\b\b\b87%\b\b\b88%\b\b\b89%\b\b\b90%\b\b\b91%\b\b\b92%\b\b\b93%\b\b\b94%\b\b\b95%\b\b\b96%\b\b\b97%\b\b\b98%\b\b\b99%\b\b\bdone.\n"
     ]
    }
   ],
   "source": [
    "'''2) Marker callrate.'''\n",
    "!plink --bfile $input_filestem --geno $max_missing_marker_callrate --make-bed --out $input_filestem"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "PLINK v1.90b4.1 64-bit (30 Mar 2017)           www.cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2017 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to merged_neighbor.log.\n",
      "Options in effect:\n",
      "  --bfile merged_neighbor\n",
      "  --het\n",
      "  --missing\n",
      "  --out merged_neighbor\n",
      "\n",
      "64329 MB RAM detected; reserving 32164 MB for main workspace.\n",
      "547812 variants loaded from .bim file.\n",
      "5183 people (2299 males, 2752 females, 132 ambiguous) loaded from .fam.\n",
      "Ambiguous sex IDs written to merged_neighbor.nosex .\n",
      "5051 phenotype values loaded from .fam.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 5183 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 0%\b\b1%\b\b2%\b\b3%\b\b4%\b\b5%\b\b6%\b\b7%\b\b8%\b\b9%\b\b10%\b\b\b11%\b\b\b12%\b\b\b13%\b\b\b14%\b\b\b15%\b\b\b16%\b\b\b17%\b\b\b18%\b\b\b19%\b\b\b20%\b\b\b21%\b\b\b22%\b\b\b23%\b\b\b24%\b\b\b25%\b\b\b26%\b\b\b27%\b\b\b28%\b\b\b29%\b\b\b30%\b\b\b31%\b\b\b32%\b\b\b33%\b\b\b34%\b\b\b35%\b\b\b36%\b\b\b37%\b\b\b38%\b\b\b39%\b\b\b40%\b\b\b41%\b\b\b42%\b\b\b43%\b\b\b44%\b\b\b45%\b\b\b46%\b\b\b47%\b\b\b48%\b\b\b49%\b\b\b50%\b\b\b51%\b\b\b52%\b\b\b53%\b\b\b54%\b\b\b55%\b\b\b56%\b\b\b57%\b\b\b58%\b\b\b59%\b\b\b60%\b\b\b61%\b\b\b62%\b\b\b63%\b\b\b64%\b\b\b65%\b\b\b66%\b\b\b67%\b\b\b68%\b\b\b69%\b\b\b70%\b\b\b71%\b\b\b72%\b\b\b73%\b\b\b74%\b\b\b75%\b\b\b76%\b\b\b77%\b\b\b78%\b\b\b79%\b\b\b80%\b\b\b81%\b\b\b82%\b\b\b83%\b\b\b84%\b\b\b85%\b\b\b86%\b\b\b87%\b\b\b88%\b\b\b89%\b\b\b90%\b\b\b91%\b\b\b92%\b\b\b93%\b\b\b94%\b\b\b95%\b\b\b96%\b\b\b97%\b\b\b98%\b\b\b99%\b\b\b\b done.\n",
      "Warning: 11574 het. haploid genotypes present (see merged_neighbor.hh ); many\n",
      "commands treat these as missing.\n",
      "Warning: Nonmissing nonmale Y chromosome genotype(s) present; many commands\n",
      "treat these as missing.\n",
      "Total genotyping rate is 0.999488.\n",
      "--missing: Sample missing data report written to merged_neighbor.imiss, and\n",
      "variant-based missing data report written to merged_neighbor.lmiss.\n",
      "547812 variants and 5183 people pass filters and QC.\n",
      "Phenotype data is quantitative.\n",
      "--het: 534198 variants scanned, report written to merged_neighbor.het .\n"
     ]
    }
   ],
   "source": [
    "'''3) Sample callrate.'''\n",
    "!plink --bfile $input_filestem --mind $max_missing_sample_callrate --out $input_filestem"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>N_MISS</th>\n",
       "      <th>N_GENO</th>\n",
       "      <th>F_MISS</th>\n",
       "      <th>logF_MISS</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>count</th>\n",
       "      <td>5183.000000</td>\n",
       "      <td>5183.000000</td>\n",
       "      <td>5183.000000</td>\n",
       "      <td>5183.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mean</th>\n",
       "      <td>280.452827</td>\n",
       "      <td>547807.548524</td>\n",
       "      <td>0.000512</td>\n",
       "      <td>-3.781140</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>std</th>\n",
       "      <td>713.359390</td>\n",
       "      <td>3.974823</td>\n",
       "      <td>0.001302</td>\n",
       "      <td>0.574419</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>min</th>\n",
       "      <td>10.000000</td>\n",
       "      <td>547804.000000</td>\n",
       "      <td>0.000018</td>\n",
       "      <td>-4.738737</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25%</th>\n",
       "      <td>31.000000</td>\n",
       "      <td>547804.000000</td>\n",
       "      <td>0.000057</td>\n",
       "      <td>-4.247260</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50%</th>\n",
       "      <td>58.000000</td>\n",
       "      <td>547804.000000</td>\n",
       "      <td>0.000106</td>\n",
       "      <td>-3.975104</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75%</th>\n",
       "      <td>220.000000</td>\n",
       "      <td>547812.000000</td>\n",
       "      <td>0.000402</td>\n",
       "      <td>-3.396206</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>max</th>\n",
       "      <td>14615.000000</td>\n",
       "      <td>547812.000000</td>\n",
       "      <td>0.026680</td>\n",
       "      <td>-1.573814</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             N_MISS         N_GENO       F_MISS    logF_MISS\n",
       "count   5183.000000    5183.000000  5183.000000  5183.000000\n",
       "mean     280.452827  547807.548524     0.000512    -3.781140\n",
       "std      713.359390       3.974823     0.001302     0.574419\n",
       "min       10.000000  547804.000000     0.000018    -4.738737\n",
       "25%       31.000000  547804.000000     0.000057    -4.247260\n",
       "50%       58.000000  547804.000000     0.000106    -3.975104\n",
       "75%      220.000000  547812.000000     0.000402    -3.396206\n",
       "max    14615.000000  547812.000000     0.026680    -1.573814"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "'''Visualize missingness and heterozygosity from reports.'''\n",
    "\n",
    "imiss_file = input_filestem + '.imiss'\n",
    "het_file   = input_filestem + '.het'\n",
    "\n",
    "\n",
    "#!head {imiss_file} #See how Plink generates some extra whitespace...\n",
    "!perl -i.bak -lpe 's/^\\s+//; s/\\s+/\\t/g;' {imiss_file} {het_file} \n",
    "#!head {imiss_file} #Uncomment to see how this Perl one-liner cleaned up these two files for easy reading.\n",
    "#Now the whitespace is cleaned up, and the original het and imiss files are backed up as *.het.bak and *.imiss.bak\n",
    "\n",
    "imiss = pd.read_table(imiss_file, sep=\"\\t\")\n",
    "het   = pd.read_table(het_file,   sep=\"\\t\")\n",
    "\n",
    "het['meanHet'] = (het['N(NM)'] - het['O(HOM)']) / het['N(NM)'] #Compute mean heterozygosity.\n",
    "imiss['logF_MISS'] = np.log10(imiss['F_MISS']) #This might generate a RuntimeWarning: \"divide by zero encountered in log10\" for genotypes that weren't missing in anybody.\n",
    "\n",
    "#imiss.head()\n",
    "\n",
    "imiss.describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Summary of log10(missingness): [-4.739 -4.247 -3.975 -3.781 -3.396 -1.574]\n",
      "Summary of mean heterozygosity: [ 0.2344  0.3188  0.3202  0.3203  0.3218  0.3605]\n"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAeAAAAHgCAIAAADytinCAAAgAElEQVR4nOzdaVwTV9sH4MlCCEvY\nd9mCghuLCK5sKqIW10oeBVsVN7Ro1bZWa2tVatWnatVqtVUUbbG1WqWIWlqwCqKCCy6VTQsIKsgS\nBCHsIfN+mD55EeIGksyY//Whv+RkZs6dMb05ObnnDIskSQIAAOiHreoAAABAMSRoAACaQoIGAKAp\nJGgAAJpCggYAoCkkaAAAmkKCBgCgKSRoAACaQoIGAKApJGgAAJpCggYAoCkkaAAAmkKCBgCgKSRo\nAACaQoIGAKApJGgAAJpCggYAoCkkaAAAmkKCBgCgKSRoAACaQoIGAKApJGgAAJpCggYAoCkkaAAA\nmkKCBgCgKSRoAACaQoIGAKApJGgAAJpCggYAoCkkaAAAmkKCBgCgKSRoAACaQoIGAKApJGgAAJpC\nggYAoCkkaAAAmkKCBgCgKSRoAACaQoIGAKApJGgAAJpCggYAoCkkaAAAmkKCBgCgKSRoAACaQoIG\nAKApJGgAAJpCggYAoCkkaAAAmkKCBgCgKSRoAACaQoIGAKApJGgAAJpCggYAoCkkaAAAmkKCBgCg\nKSRoAACaQoIGAKApJGgAAJpCggYAoCmlJmiSJKurq2UymTI7BQBgKGUk6Lq6ug0bNjg5OfH5fH19\nfR6P5+joGBER0djYqITeAZ6yciWBIQIwhDISdHh4+Llz5yIjI4uLi5uamkpLSw8cOJCenh4eHq6E\n3gGecu0aEjQwBYskya7uw8DAICsry8rKqnVjTU2NUCgUi8Vd3TvAUwICiPh4gstVdRwAL6aMEbRQ\nKIyPj2/T+Mcff9ja2iqhd4Cn2NgQbPw2DsygjBH0tWvXRCKRlpaWi4uLQCCQSCSZmZmVlZVxcXEe\nHh5d3TsAAEMpI0ETBCGVSpOTkwsKCsRisaGhYffu3f38/Lj4mgkA8GxKStCtlZeXa2hoGBgYKLlf\nAIIgiJUrifXrMcsBjKCMj+mYMWNKS0sJgigqKvLy8rKwsDAzMxs5cuSjR4+U0DvAU1DFAcyhjAT9\n559/1tfXEwTx0UcfOTg4VFdXSyQSNze3RYsWKaF3AACGUuos8NWrV3///XcdHR2CIFauXNmjRw9l\n9g5AEKjiACZRUoIuLi62trbu27dvQUFBz549CYLIzMw0Nzd/zi5Xrlw5d+5cm8bKykpfX9/AwMAu\njBXebFFRqo4A4GUpI0H7+voGBweXlZVpaWndv39/9OjR58+fnzRp0ldfffWcvSwtLdsX4SUnJ2dk\nZCBBA4A6UEaCTk5OJgiiqanp/v375eXlBEFoaWmdPHnS29v7OXvZ2NjY2Ni0aXz8+DEuPoROQRUH\nMIfyPqY8Hq9Hjx5DhgwhCGLAgAFCofDUqVNK6x3gX6jiAOZQ2TgiNTU1ODhYVb0DANCfyhK0SCSS\nSCSq6h3UF6o4gDmwYD+omagoJGhgCizYDwBAU1iwH9QM7qgCzKGMMrvY2NjWC/YbGxt7e3u7ubkJ\nhcL9+/crIQCA/0dVcWCWA5gAC/YDANCUMkbQkZGRIpFoy5Yt7RfsV0LvAE9BFQcwB8MW7D969KhY\nLMbkNQCoAyUtlsTlcv39/ZXTFwDAmwHf9UDNoIoDmAMJGtQM1uIA5kCCBgCgKSRoUDOo4gDmUOot\nrwBUD3dUAebAUAIAgKaQoEHNoIoDmAMJGtQMqjiAOZCgAQBoCgka1AyqOIA5UMUBagZVHMAcGEoA\nANAUEjSoGVRxAHMgQYOaQRUHMAcSNAAATSFBg5pBFQcwB6o4QM2gigOYA0MJAACaou8I+sGDB3fu\n3GnTePv2bYFAoJJ44A2xciWxfj1mOYAR6JugHz16lJ6e3qYxPz+/R48eKokH3hBUFQcSNDABfRP0\nwIEDBw4c2KaRuqu3SuIBAFAyjCNAzaCKA5iDviNogC6BKg5gDgwlAABoCgka1AzW4gDmQIIGNYO1\nOIA5kKABAGgKCRrUDKo4gDlQxQFqBlUcwBwYSgAA0BQSNKgZVHEAcyBBg5pBFQcwBxI0AABNIUGD\nmkEVBzAHqjhAzaCKA5gDQwkAAJpCggY1gyoOYA4kaFAzqOIA5kCCBgCgKaUmaJIkq6urZRi/gAqh\nigOYQxmf1Lq6ug0bNjg5OfH5fH19fR6P5+joGBER0djYqITeAZ4SFYUEDUyhjE9qeHj4uXPnIiMj\ni4uLm5qaSktLDxw4kJ6eHh4eroTeAQAYikWSZFf3YWBgkJWVZWVl1bqxpqZGKBS+6i26qbt6I7ND\nx61cSaxfj0E0MIIyPqZCoTA+Pr5N4x9//GFra6uE3gGegioOYA5lXEkYGRkpEom2bNni4uIiEAgk\nEklmZmZlZWVcXJwSegcAYChlJGhPT8/c3Nzk5OSCggKxWGxoaBgWFubn58fl4kJzUDpUcQBzKClF\ncrlcf39/+dO0tLSWlhYkaFABrMUBzKGaocS4cePKy8tV0jUAAFMoI0Hr6upyn1ZRUWFnZ4cRNKgA\n1uIA5lBGgr569erAgQMnT5589+7dkpKSkpISQ0PDGzdulJSUKKF3gKegigOYQxlj2N69e6ekpOzc\nuTMwMHDr1q2BgYFsNtvIyMjExOQ5e9XV1bXP4KWlpSwWqyuDBQCgCyVNMnA4nKVLl44fP37u3LmH\nDx9uamp64S5nz5798ccf2zQ+ePDAx8ena2IE9YAqDmAOZVxJ2JpMJtu3b9/Zs2d37dplbGz8qrvj\nSkIAUB/KHkqw2eywsLBffvmloaHh1KlTSu4dAIBBVPZdLzU1NTg4WFW9g/pCFQcwh8oStEgkkkgk\nquod1BeqOIA5sGA/AABNYcF+UDOo4gDmwIL9oGZwRxVgDizYDwBAU1iwH9QMqjiAObBgP6gZqooD\nsxzABFiwHwCAplSzYD+AyqCKA5gDY1hQM7ijCjAHhhIAADSFBA1qBlUcwBxI0KBmsBYHMAcSNAAA\nTSFBg5pBFQcwB6o4QM2gigOYA0MJAACaQoIGNYMqDmAOJGhQM6jiAOZAggYAoCkkaFAzqOIA5kAV\nB6gZVHEAc2AoAQBAU0jQoGZQxQHMgQQNagZVHMAcSNAAADSFBA1qBlUcwByo4lAlkiSPHz+elpZm\nZWUVGhpqZGSk6ojUAKo4gDlYJEmqOgbFjh8//v3337dpLC0tffLkSffu3a2trbdu3WpiYqKS2F6X\nGTNmODo6jhkzprCw8Ouvvz527Fi3bt1UHRQA0AV9E7RCR48eFYvF4eHh58+f37p1a2xsrKoj6rgb\nN2589913e/fupZ5euHDh+PHj27Zte6WDyGSyqqoqDL1fwcqVxPr1mOUARmDqx9TX15fFYlVVVak6\nkI5obGz86aefNm/ezOfz5Y39+/e/e/fuKx1n8+bN3t7ec+fO9fb2/vPPP193mG8oVHEAczA1QRME\noaGh0dzcrOooXll1dXVAQIBYLPbx8UlKSpo9ezbVfuXKld69eyvc5erVq8eOHWuTvmNjYx88eHDx\n4sWYmJg///xz06ZNRUVFXR49ACgRU38kvHPnTlVVlamp6bM2+Oeff+Li4jgczsSJE4VCYZtX8/Pz\nU1NTTU1NR4wYweUq9STs3LlzyZIlQUFBBEHcvHnz2rVru3fv1tbWjoyM/O2339psLJVKp02bZmpq\n2qdPn88//9zR0fHLL7+kXjp16tTKlStZLBZBEDo6OiEhIUlJSe+8806bIzQ3N5eUlFhaWir5bdIX\nqjiAOZj3Sd27d+/kyZOXLFmyZ88eeaNMJtu7d+/o0aNHjx69ffv2mJiYxYsXOzk52dvbz5s3LyEh\nofURvvvuuw8++KCuri41NXXkyJGPHz/u0oBra2tXrVo1bNiw0aNHR0dHZ2ZmDh06lHppz549o0aN\n+v3335uamv744w8zM7M2+0ZFRfn4+OzatWvhwoVHjhwpLi6+evUq9ZKGhoZUKpVvKZVK26fgyMhI\nPz+/VatW+fn5tT5dai0qCgkaGINklCNHjmzbtu3x48dt2r/88suVK1fW19c3NTVt2rTJysqqrq6O\neqm6utrPz0++ZWlp6ciRI2UyGfX07Nmzixcv7tKYp06d+vPPP5MkWV9fv3DhwrFjx/7555/yVz/5\n5JMzZ848a985c+bk5ubKnx47dmznzp3U44SEhOnTpzc1NZEkWVZW5uvrW1ZW1nrfy5cvBwcHt7S0\nkCTZ0tLyzjvvXLp06bW+MwDoWswbSvB4PENDwzaNJ06cuHfvXmBg4KRJk/r27dvY2Mjj8aiXBAJB\n68Hm33//PWzYMGpmgCAIPz+/jIyMrov28ePHjY2NISEhBEHw+fxvvvnmyZMna9asiYmJyc7O3rFj\nR2Zm5vDhw5+1u6Wl5f379+VPCwsLraysqMcBAQEj/mfatGmbNm1qM+GTmJg4Z84cNptNEASbzZ43\nbx5+SCQIrMUBTPImzEtWVlbevXv34MGDffr0qaioePfdd5ubm588eUIVnzU2NjY0NMi//ltbW//y\nyy/yfYuLi1vnNZlMVlZWZm5uLs/gnfT48ePWxdocDofL5Z44cSIyMvKvv/7y8PCIiYlhP/sb9+zZ\ns6dPn/7tt9/26dMnKSnpxIkT8fHx8ldDQ0NDQ0Nbb19fX6+lpUU91tLSqq+vV/hSh5EkGR8ff/Pm\nTaFQKBKJNDQ0OnlAFaCqODDLAUzAvARtYGBQXFxMEISRkRGfz29qarp06dLUqVOlUmlDQ4OxsfFH\nH32kr6+/Y8eOUaNGEQTx1VdfLVmyhNqFIAg7O7vc3Nzw8PCgoCB9ff0jR44sXbqUIIjGxsbdu3ff\nunXL1NT09u3bM2bMEIlEd+7cKSkpMTc3d3JyovoSi8XUceS9P7/FwcGhoKAgLy+PSo6FhYX29va6\nurrTp0+ntqHmjhUep6ioSFNT89tvv/3hhx+ysrIGDRq0b9++qqoqqvC5TV/R0dEnTpywsrISCoVB\nQUG2traBgYGLFi0aNGiQVCqtq6s7d+7czJkzCYLowLvgcrkRERE1NTVlZWU6Ojr+/v6PHj2aNGnS\nrl27qG8qL3kcOrSQjY0VxcUEl0uTeNDC9JaKigpzc/PnDLM6g3kJ+q+//rpz5864cePkV2fU1dUN\nGTIkKirKwMBAKpVevHhx6NChI0eOvH79ekVFxZo1a5ydnanTSpLkkiVLnJ2d6+vrDx06xOFw5s2b\n5+bm1tzcPH78eCq15efnV1VVHTp06NKlS4aGhlZWVpcuXaqrq/vqq6/q6uqSkpIeP35sa2s7cuRI\ngiAqKiouXbrU0tLSp08fhVeLsNnszz//fOfOnU5OTvX19RcvXnzOj3VSqfTatWu5ubkPHz7Mysoa\nOHBgXV3d2LFjN27c2ObD0cbFixebm5up/+bn53///fdLliyxt7dftWpVWFiYg4MDl8sdM2aMg4ND\nx875+vXrLS0thw0blp2dPXDgwP/+978xMTHm5ubnz5+nzgODtFhZYfgMjKHqSfBXc+TIkTFjxuTn\n5/v4+BQVFVGNd+7cmTBhQkNDQ2pqakpKSkRExMGDB1vvFRMTM3z48OHDhw8YMGD69Omurq5mZmYW\nFhb9+/ePjY0lSXL//v2jRo26ePGiTCZbsWKFjY2NnZ0dj8fr16/f5MmTfX19Z8yY8dNPP3l5eR04\ncCAtLW3jxo1vv/321atXvb29o6KiDh8+PGrUqJiYmNadlpWVhYaG+vr6+vr6rlq16uLFi7du3ZL/\nONlefn6+t7f3zp07N2/e7OrqGhISIpPJCgoKevXqtXz58qtXryrcq66ubu/evT169Ni7d29zczPV\nGBERcfbs2ReezPr6+i1btri6uvr4+Pj5+cXFxT1ry+HDh5MkuX379t9++40kyZkzZxYWFubl5c2a\nNeuFvQBAhz1zBN3S0lJWVmZhYfG6ZmNfl+YrV+wrKla99dbpI0fmffAB0dDgVFMzy9V1sZeXqbd3\nRkGBvYXF54GBRHp6RUWFVr9+V7Ozf42O/n3dOj6fv2HDhs2HD89ZsmT54sU1aWnvv/9+zGefTfT3\nv3XrVi97e82MjN9On+bcvOlgZpaWkWFjauouk30wfryOjs6769bl5uYuX7x4Yo8eBEEMCghYU17+\n3nvvnT5+3Ky8nCCIyWvXjlmyxNvb++Svvx7/4ov6+nqpVDrrv/89cOAAWV8f/cknOT/9NHv2bKK2\nltDVJRoaiMxM6u1kSqVfbN16OTnZhc0219TU+PvvhzzeDz/8EB0ZeWDRovPnz89yda2QSLZt2za4\nX7/3R4z49yz07Eno6tZWVHwwatSYMWOG6+lVFxdPmDAh7uhR7p07lsXFgrt3iQEDWvf16NGjHJI0\nsLa+cv68Zm6unZ1dVFTUzfr6oJkzUxITN4aEHN68uadE4jR+fJsIiZ49SZIkGhr6k2Tq4cM/p6QU\n370rFouL8vK8tbSI9HR5PG32UnlLRVHRozNnbGxs9PX16RAPWt7Mlpwcws2N6Jo8qSBBFxUVzZgx\nIzU1lcPhpKSkLFy4MDo6usPfjl87UXV1yvTpTra2DxwdCYIg7t+v/+Yb97Ky/qamVe7uC1essKyp\nyZozJy0tjSCIKB4vQ0MjsEePpGnTevTo4VtS8juP5+PjY9bQYPbXX/91cEhPTyf++cfOzk6/tLR8\n/Xqirs7fwMDW3z8tI2OQpaX3zZvFa9ey2WzDhobcxsYB5ubE/5bO8HV0PPzkiVlDQ+Fnn/39998y\nmayustLX15edm7veyMgnICA5OXnrJ5+cPHfOtLJyRnl5bm7uqdjYMz16vPv11556ei3ffcfhcIqK\nir4pLb0rlZ7ctSt77lxrY2Pu1au3tbUjIyMLExPdHjyYKBAYNTc/njJl0rffzhg8uDk7+9/f5RYs\nINzdY7ZvX6an5yQW9xUIEu/f9/b2PnvwoPfly+Znzrg1NREDBxLu7sT9+8TevWlpaTU1NdcHDtyU\nmDjV3X05h5O0d+9bPJ6dr6+vr6+PufnD1auXDxlS/PXXRK9e8r3+PeMLFgwaNCju228dfvvN6uZN\nCysrzfr6kJAQu8bG1To6V1NTHR0dDT75hNqrdtu2lJSUJ0+exNvZcTw9ty5YoL9vn/w47Y/cdS0R\ncXH3z5wJq6m5WFFhaWnpvmfPv9vMnUsMGECwWEqOBy1vbMu+fcRXXxE6OkRXaD+oDg4ODg8Pr6ur\nMzc3l0qlH3/88YgRI5Q+tFfsyJEjWlpa3bp1MzMzi4qKKikpWbBggaGhoa2trYmJiZGR0aRJk955\n5x0Oh7Nr166UlBSqHsPMzOzTTz/t1q2biYkJi8Xq1avXqlWrnJycBAIBh8NZvXp1WVnZ4MGDp02b\nxmKxNDU1NTQ0evfuLRAIrK2tP/3009OnT+vq6goEgn79+m3fvj0nJ2fRokXm5uZ8Pj8gIGDw4MFn\nzpwZPHiwra0tl8vt1q2bs7Ozqalp9+7ddXR0PDw8fv/9dycnJ01NTT8/v8GDB+vr67u7u48ZM8bX\n19fDw+Onn35atWpVVFSUl5eXQCBwd3cXCARmZmaLFi1ydXV9/PixkZHRBx98kJycvHTp0mvXrrU+\nG4sWLbp9+zb1eN26dX379nV0dAwICLh+/XpxcXH//v0NDQ2NjY1tbGwsLS2dnJwsLS2Tk5PDwsJc\nXFx69eqlqanJ5XLZbDaPx9PR0endu/fChQtJkszJyTl16lReXp68o/r6+tDQUD09PRcXF2tr6ylT\npgwfPtzOzm7Tpk0LFy708PCIjY3ds2dPeHi4qanpjBkzfvzxx+HDh2/YsGHy5MlK+FTcu3dv48aN\nn3/+ubzQ+8yZM++99558g7CwsHPnzv37ZORI8n9zQQA0p2A1O2Nj47y8PAMDAwsLi5KSErFYbGdn\nV1tb2yV/H17R0aNHp06dKn/KYrFIktTT0xMKhXZ2dr///ruRkVFVVZVUKqVqn1taWuQbc7nc1pfe\nsdnsf08Bi8VisUxMTMrLy6mzweFwZDKZlpZWY2OjlZVVdXW1RCKhNuNyuY2NjYaGhiYmJiNGjNi3\nbx/VhUwmY7FYbDZbV1e3pqam9ZGpBwRBaGpqUg/69Omzf//+06dPr1q1is1mU1FRW1JdU2FQG2to\naFBXGFZWVm7ZssXU1JTL5T569CghIeHOnTs6Ojr9+vUTCoX/+c9/jhw50r17d01NzVOnTh0+fNjL\ny8vLyysmJiYnJ4ckye7du+fn57NYLA6Ho6OjU1dXp6GhIZPJZDJZfX09SZJ8Pl8mk+nq6hIE4ePj\nQ5Jkz549v/rqq+rq6sbGxh9//HHHjh0rV66UyWSHDh2qr68vLCz8+OOP9fX1ExMTf//9d19f39TU\n1KamJh0dHVdXVw8PjwsXLlhYWPz000+ampr79++/ceOGra3tggULCgoKtm7deufOHQMDgylTpsyY\nMaMz9X+XL19evnz58uXLBQLBvn373N3dP/jgg9WrVwcEBPj4+FDbnD9//uzZs2vXriUIgggIIOLj\nCVz4DozQPme7urqeOnWKJElzc3OSJE+fPu3s7NzFfyde1pEjR17v239+JS+LxaI2oFJtm5csLCza\ntLR+0H7uvvVBWP/zwgjZbLa2tra1tTX1B6B1wNQR2Gw2VV5NjV579Oih879vW/JIWCyWra2tjo6O\ngYEBn8/X0tKiXqL21dbWZrPZbDbbwsJCU1Nz4sSJZmZm/fr1mzBhgp+fn4+Pz9tvvy0QCIYMGUL9\nK1y8eFFfX79nz57e3t5TpkwxNjZms9nTpk0bOnSovr4+lZ03btyoo6MjFAo//fRTT0/PlStXZmZm\nnjhxwsXFxdfXd+bMmeHh4WvWrOnVq9fQoUPLy8uvXr166tSpR48evepHYsyYMfJLKGUymbe3d0ND\nw9dff33ixAn5NjExMdu2bfv3yaxZZEtLZz+IAEqhYByxY8eOoKCgESNGVFdXi0SilJSU6OjoF+YR\nhnr+engkSVIbkO2+Z5AkWVJSIh8gU2dT/tKzdlH4+PlkMlldXV1dXR21l6zVVXDyjjQ0NEiSFIvF\nR44cMTQ0pAb1fD6/oaFBvplEIqmvr6e+CVHjaBaLRQ3eW1paqG8AJSUlHA7H2dlZT08vNTU1ISFB\nT09v8uTJGRkZZmZmGRkZ1tbW3bp109HRqa6utrOzmzBhQmpqqp6eXnV19W+//aavr9+/f//bt29X\nVVXdvn1bKpWWlZVFR0dXVVV5enrOnz//xx9/tLKyoiLctWsXQRBU6fqgQYMCAwNtbW137NgxZcqU\nOXPmvOTJoY4gv9SIxWI5Ozvfu3dPJBJNnz69X79+tra2hYWF27dvP3To0L874I4qwBwKErSnp+ed\nO3dOnjzp6upqaWn57bffCgQC5UfGCM/Kxa8Rlf3bP+ZwOPIJHCrPkiTZ2NhYUlJCNVLZmRpit7S0\nVFZWkiTJ5XKpLC9P9NRTLpfL4XAaGxtJkqyoqKBmw/Py8t5///3//Oc/CxYsuH//PpfLtba2dnNz\no67MvH///v79+/X19R89eqSlpcVms6uqqq5du1ZbW6upqUkduaGhoV+/fh4eHlVVVfv27fvggw9a\nWloKCwvl1+nY2NhQiw6uWbPGxMTko48+CgwMHDNmzMvfWUZLS6u8vJzK0SRJZmRkCIVCTU3N7du3\nL126tKqqytDQcMeOHTY2Np3+pwBQtqcSNPX/eZ8+ffLy8t59912qUSKRWFpaVldXqyA6umqdKNs/\nfb2eNe5uM5SmLmSiHrT+s0GSZOuJeBaLxefzpVJpY2OjfJvm5mYejye/FOrAgQO9evUSCARUhp02\nbdqRI0eGDRsmkUhCQkKuXLlSX18/aNCg69evP378WCgUhoeH7969W0NDo6mpqampicPhaGpqBgUF\nicXilJSUmzdvWlpaFhUV9ezZs7q62sTEhMPh3L59m+p3w4YNDQ0N3bt3nzp1akhIyNy5c0ePHp2e\nnv7yCXrNmjUikejjjz+m5qBFIhH158Hd3T0mJkbBDrijCjDHUx9TPp/P5/Pv37/Pb8XExCQwMFBV\n8dHTK2XnzhSSt96XmlmWtyiMgcViyX9jbH0E+QZ8Pr++vr5Hjx7y+WuCIDgcDpWyqd2bmpqkUun9\n+/dZLFZkZGRdXV1YWJiWlhaPx7tx40ZSUtKyZct69+49evToqqqqvLy89PR0MzMzas6E6q5v377n\nzp3r1q1bS0sLl8t98OBBaWnp6dOnCwsLy8rKCII4c+bMxIkTe/fu3aNHDw6Hs3Xr1jNnzvz222/3\n7t0rKioyNzd/+VM0aNAgahHXs2fPLly4cMmSJS/YAXdUAeZ4KkFLpVKpVBoQECB9WuvVhTqDJMnq\n6mqZmv3v0ZnBtXxUy2Kx5NPQCrug2jU0NDgcTuvN5DPUBEHweLxx48bx+fyMjAxqrM3lcjU0NFpa\nWmbMmJGenm5vb09NTxsbG1OjXTabXVFR8fjx48LCwtmzZx88eNDJycna2jonJ4f6g1FQUHDz5s3i\n4mINDY3//Oc/vXv35vP5jx498vT0vH37NpXrq6qqli1btmXLlqFDhyYmJsbHx//4448ODg5SqdTL\ny+vgwYNffvnlw4cPx48f/913392+fdvT0/OVzpKtre2KFSsiIiIGDx7c4VMNQEcv80vijz/++LI/\nOipSW1u7fv16R0dHamEdDofTo0ePtWvXNjQ0vOqhXnsVRyex2Wx5pUfHtKn9aN1OjUmJVmla4e5U\nRpZvpqGhQVU3yzeg9O7dmwpVU1OTzWbr6+s7OTnl5OT06tWrdSmInp6elpbW0KFDfX19z507Z2lp\naWxsPGDAAGdn59DQ0J49e/bu3dvf33/Pnj1BQUGLFy92cXGZMmWKhYXFgAEDevbsuWXLFqFQqKGh\nYWRkpKura25uPnbsWOpmBZGRkW2udJ8/f35WVhZJkhcvXpwyZYqDg0NISEhVVVVnPmwvhioOYA4F\nCTo7OzssLGzq/0yYMIGqt+uwmTNnjhw5MikpSSwWUyv+pKSkjB8/fvbs2a96qFdK0M+ZW3jWSwor\n5KhLV7hcbut2FoulpaVFVZgRBEFNvFKVzrADZ4sAACAASURBVFTSZLPZVLk0QRBcLtfY2Fh+cC0t\nLQ0NjW7dumloaFCTvxRbW9tu3boJBAI+n08lTS6Xy+PxQkJCnJyc7OzsqBxqZmbm5OSkoaHh5+fX\nrVs36vc9auPWmZrH42lqagoEguDgYBMTk6SkJC0trQULFuzatUtHR8fd3d3T0zMxMXHo0KF8Pn/J\nkiW//PJLUlKShYWFvb39lStXRo0aFRAQkJKSUlJSEhwcbGRk5OPjs3z58tLSUpIkN27cOHPmzDNn\nzuTl5YWHh+vp6Tk5OVHXhsTHx3t4eCxbtiwqKurJkyfP+de8cuXKmDFjSkpKSJK8cOGCr68vVZEN\nABQFVRwzZszo27evjY1NVlZWSEjIzp079+/f//Jpsb3Y2NisrCz5SvPGxsbe3t5ubm5CobADR6Zq\neK2trR8+fCiVStlsNvV7F/G/GVgLC4vevXtnZmaWlZWZmJg4OztnZ2dTxWoymYzNZjc1NclkMupn\nKIlEoqWlJRAIamtrq6qqqDlTPp9PDQlTU1PLy8vFYrFAIBAKhVS5ArUZn893d3cPDw9/6623AgIC\nCgsLDQwMeDxeQUGBnZ2dubn53LlzJ06cyGaz8/LyQkNDHzx4MHz4cBcXl6tXrw4dOjQjIyMhIWHI\nkCFWVlZffPHFkydP3nrrLQ0NjfLycl9f3zlz5vz222+2tra//PKLiYnJxYsXTU1N5eUZOjo6NTU1\njo6OM2fOPHHixLlz5xoaGgQCAZfLHT9+fGJiYkNDg6ur64ULF3g8no2Nzblz54KCgrp166arq1tX\nV/frr78KBAI7O7uUlJTExMRff/21X79+tbW1v/zyi7m5+aZNm7Zv375161ZqLaQhQ4bU1NTweLzd\nu3e3vkTI1tY2Kyvr7t27/v7+u3btsrOz+/rrr7Ozs+fPn5+Xl3f69OmXmUceMGDAp59+GhYW9uTJ\nk759+x45cqT1bc4BQMEPXHw+v6ioiM/nT5o0KTExsbS0lKqG7nAf7u7uixYtalPc+uuvv27cuPH6\n9euvdKijR49mfPhhZXm5jY1NtrX1t6dPaz5+fPaDD9LS0ubMmXO/Z8+569fX5+ePbmoiCMLAwOC2\nhUXgwoUOWlp/r1/P4XCKi4uTNDTY3bvPHDWKGx9fW1srEAhONTaOXbQo2M9PLzk5KyuroqLC4D//\ncZ04kXj0SPrbb48fP05LS/uprMzIwyPrr79cCwttbGxIkryorX2hqGi8p6fb/fuWlpZ9+vT5i83m\n9+79rr+/XnLyv+EGBBBCIfHoEXHyJNXQMmJE5Jkzd5KSJrLZfn5+LBZLvs3ut94aOXLktWvXph04\nQNrbT/HxiZ4yZc2aNZaWlktOnbpWUVGVnX1tzRo7OztfX1+TkJAbVVXfr1njdv9+VVUVQRC6b7/d\nZ+zYcR4eh6ZOvXfvnpmZ2fANGxL++efnr7+ebmBgbW19586dBJL84fx5S4Io3bfvu+++09HR6blo\n0Z6EhGA/v8ZjxyQSCZvNzra2jvjxR7OWlqbjxxMSEnJzc29bWExYsmTiwIHyd0EEBNRbWIQMG9bv\nwYOAgAAul/v5+fNPjIy++eQT+9u3/73dQbv3TpeW2bOJSZMIOkeIFma1nD5NTJ9OaGoSXUDBCNrC\nwiIrK8vHx6eysrKyslJPTy8rK6szfURGRopEoi1btri4uAgEAolEkpmZWVlZGRcX14GjWTs712Rk\nkAYG88LDdXR0ZBLJtbw8r3Hj/rxy5fc//1y4cOGBTZs0m5sXLFjQq1evv3V0Nvz446m8vI3BwRcu\nXNiybt1tXd3Nhw//+PPPHjU1np6e1dXVRgYGBw4cmDVuHMvIqK+3N0EQhIcHQRAEl8s1NTUzNZ3Q\ns+dwZ+cCqTTHxYV74cKAAQMEAkHanj0rV6604HCm29gkJibmPX784ebNhLU1UV5OyO/IRc1Nc7lU\nS2Nj44yZM93Hjx85ZkzeL7/kx8bOnj1bvk3gO+/s27evvr7eIT1977p1viNGNOvqPmGzTQWCDZs3\nu48fn5+XJ25p6efgYO3iQggEQ3r3HvLTT0RSUmVl5dKlS0tTUgYFB2/etk2LxeJbWlr16nX33r1F\nixYtmjr1QXS0RCLxF4kCLS2nTJkywN7eOjeXb2npOWhQcUPDzp07HQQCQiiUSqUkSWr4+hJmZkR5\nOc/cfBxVrTxkSPv3paWlFX348O/Ll8cmJ8tkMr+xYxd8+aWxTEbIL6Z/+r3TqKWoiNDXJzgcusSD\nFqa3GBh0YdVm+1mPPXv28Hi8wsLCVatW9e/f39fXd9SoUZ2cSWlubj5z5sy+ffv++9//7tmz58yZ\nM80dWrDmyJEj48eP37NnT9++fUeNGrVmzZrhw4cfOHDg9OnTQqGwsrKyurp6/Pjxjx49CggIoHZx\ncXEZMmTIO++8k5CQQLX4+fnp6OgsX76cehoXF+fh4XH9+vUX9i5fZ/n69esffvjhjRs3li5dSpKk\nVCr19/d/4e4bNmz45Zdf5E/DwsIuX77ceoOHDx8GBgb2799/7969f/75p5+f34wZM86fP5+QkLBp\n06Zjx45t3LhR4arNPj4+gwYNCg0NjYmJkUgkXl5eS5YsSUtLa79lS0tLbm5uB66ofnNgsSRgDgUj\n6LCwsLfeesvU1HTNmjW9e/d+/PjxjBkzOvlngMvl+vv7U48bGxup37U6digTE5Pg4OC+fftu27Zt\n2LBhixcvZrFYy5YtGzRokEAgoK5a1tHRMTMze/LkyaZNm8rKyrS1tXk83vbt211dXQ0NDevq6gwM\nDKytrWtqagiCCAwMXLBggZaWFvWUIAhtbW1q0SL5ElFUS8+ePe/cuePp6UmVFaenp/fu3bumpkYq\nlVpZWbW0tCjcS97y6NEj6kxSLSNHjszJyfHw8JBvo6en98svv6Snp8fHx5eVlX3//fcEQaxZs+aT\nTz5ZvHjxuXPnbt26NW/evJqamjZHXrdu3apVq1JTUx0dHY8ePTpixIhHjx71799f3lfreIRCYW1t\nLfVmn/VO3+CWFguLhtpags2mSTxoYXpLXV0dtcRYV+D8u8RXK66ursOHD7e3t2ez2S4uLgMHDtTs\n3PRKVlbWzJkzExMTvb29J02atGDBgo0bN2ZkZAwbNkxbW/uVDpWZmeno6Kivr6+rq6utrf3VV1/d\nvn07Kytr+PDhUqnU09NTW1s7LS2Nw+E0Nzfr6ekdP3581KhRxsbGI0aMYLFYpaWlJ0+etLe3v3fv\n3uDBgy0sLOrq6nJycqKjo9euXVtaWlpbW0tdqczj8RoaGkpKSlq32NnZJScnV1ZWampqHj58mLq1\nilgsPnPmTN++fe3t7RXuJW9pbGyUyWRCoZBqyc3Ntba2Njc3b7OXk5PT2LFjBwwYUFdXx+Fw3N3d\njx49un37di6XO3/+fKlU2v7Ienp6b7/9NovFunnzpoeHh5OTU1BQEPVn6TnxqGfLQw+P2ro6+sSD\nFqa3lJaW6unpddE9CRX8SPjll1/evXt33759VNly53l5efXq1cvCwiIqKmr69OkrVqyoq6tbtWpV\nU1PT4cOHX+lQR48eFYvF4eHh1FOSJKmbcBMEsWHDBrFYvGLFiqqqKpFI1NTUxGaze/bs+euvvzY3\nN2/atCk2NjYnJ8fBwcHBwUEsFlNXUrDZ7JKSknnz5n344YcvE0Bzc/NPP/10+/ZtY2PjtLQ0iUTS\n1NT01ltvffrppy+8YrCoqEgkEkVERPTq1SsxMfHYsWOnTp3q8DcJAHjjKUjQw4YNu3nzZkNDg7W1\ntby0Nicnp8N9aGlp3b9/X0NDw9DQsLa2lho1l5eXOzk5VVZWvtKh2iTo1kiSPHbs2PHjxzkczrRp\n08aOHXv+/PnTp09/9dVX1AZpaWlHjhzZtm0bQRBisXjVqlU5OTkcDic4OHjevHkdfnev5NGjR999\n911hYaG7u7uPj4+lpaW8+hCUBGtxAHMoSNAZGRntt3N2du5wH+bm5klJSb169Wq93H5aWtrcuXMV\n9kU5fvw4NQnbWmlp6ciRI7du3foy/ZaUlPj4+Jiamo4ePXrw4MFffPHFoUOHhEJhh9/I65KQkBAR\nEeHp6VlUVKShoXHw4MFOTiLBK8CC/cAgSvghctWqVba2tleuXKGeFhYWLlmyxMzM7MCBA696qCNH\njuzatetltiwtLR0yZEhsbOzatWuHDBliZ2fX+h5OKvTkyRNvb++6ujrqaXR0dEREhGpDUi+o4gDm\nUMYXvS+++GL//v3URc8EQTQ0NFhZWZ0+fTo0NLTrOt27d+9nn302ceLENWvWXLp0acGCBdQSlyp3\n8+ZNf39/+U2epk6deuHCBdWGpF5sbDC/AUyhjE8qi8UaOXKkfG7Byclp+fLllpaWp06d6rpOCwsL\ne/bsKX/aq1evgoKCruvu5enp6T158kT+ND8/n6pC8fHx+eyzz1rfNRG6RFQUEjQwhYJP6uLFi8+f\nP996lfeukJqaGhwc3HXHd3NzS5Zfck0QSUlJ/fr167ruXp6zs/OtW7fS0tIIgqivr/f393///ffP\nnTuXkpJiZma2fv16VQcIAHSh4EfCNWvWxMbGlpaWTp48WSQS+fr6cmnzi8rRo0e/+eYbHR2dAQMG\nfPzxxwYGBs/asrGxccKECV5eXs7OzklJSQ0NDXv37lVmqM/x6NGjzz77LD8/XyaTSaXSS5cuyV8a\nPnz4uXPnVBjbmw9VHMAcCj6mERERt27dunTpErVqs7W1dVhYWEJCwvPvr/oyyNexYP+oUaNOnz49\nZMiQyZMnNzU1PWszTU3N+Pj4/v37P3z4cOrUqc/KzjExMfPmzVuyZEl6enpnonollpaWUVFRSUlJ\nhw4dcnR0VFq/QBC4owowyTPHEUZGRjY2Nt27d29qarp06dLatWuFQuGJEyc60EddXd2GDRucnJz4\nfL6+vj6Px3N0dIyIiJAvE/pKTE1NNTQ0xo0b5+Pj8/yf19hs9rhx4xYvXuzl5aVwg88///zChQsr\nVqwIDQ1dvXr177//3oF4OoO65/S9e/eopzExMW5ubkqOAQBoS8HcxebNm0+fPn3t2jUvL69x48at\nXr2a+n0vKSkpJCRk4sSJr9pHeHh4UVFRZGSks7Oznp5edXV1dnb2pk2bwsPDO7PStK2tbXFxcYd3\nb2hoSElJSUpKop4ePXp04sSJbe6+2NDQsHz58ps3b3K5XHNz8+3bt7/S7fJexp49e+bNm2dgYFBT\nU2Nubt6+9BteM1RxAIO0r7ybMmXK8ePHq6ur5S0SiYT6b0xMTAdK+fT19YuKito0VldXGxsbv+qh\nWtdBT548+c6dOx2Ih5KbmztnzpzWLcOGDWuzzYcffnjw4EHq8aVLl8aPH9/h7p6voqJCXhkNAEBR\ncNPYtLS0CRMmaGlpUU+rqqosLS0JgtDR0Xn77bc78DdAKBTGx8e3afzjjz9sbW07cLTY2Nhdu3ZN\nmjRpyJAhTk5OHTgCxd7ePjMzUz6x/vDhw/ZLUl2/fn3mzJnU4yFDhnC5XGp1/NfOyMhIXhkNAEB5\naoqDuuEQdT+n1u0ikagzfbzeBfv79etnYWGxbdu2Tl60zeFw3n///YkTJ4aGhlZXVx84cGDPnj0v\n3AV1yoyHKg5gjqcSNJV9Ro0alZCQ8Br78PT0zM3NTU5OLigoEIvFhoaGYWFhfn5+Haves7e3DwoK\nei2BTZs2bdCgQYmJidra2qdOnTKU3yvhf/r06RMXFzdhwgSCIDIzM2tqauTXQwJTUVUcSNDABArq\noOnsOavZdQWJRPLee++VlpZqaGiwWKzdu3d3bFoGaASLJQFzPPUx9fT0/OKLL1avXt1+u2vXrikr\npJdVXV0dHx8vkUj8/f3t7e27ogtdXd3o6Oimpqampqauu2kCKBWqOIA5nkrQ33//vVAoZESl1927\nd2fNmjVt2jSBQPDee+/NmDEjJCSki/ri8Xiv694FoHpRUaqOAOBltR1BEwRhbGycm5tra2srk8li\nYmIaGhqmU3d3ppMVK1ZER0c7ODgQBBESEuLn5ycSiTSo++wCALwRFHzX++KLL5ydnaurqzdv3vzN\nN9988803S5YsUX5kz1dVVUVlZ4IgNDQ03Nzc8vLyVBsSMMPKlbjUG5hCwU8l33zzTVpamrGx8e7d\nuy9fviyVSgcOHLh7927lB/ccXC5XIpHI54Xz8vKsra1VGxIwA6o4gDkUfExbWloMDAyuXr1qbm5u\na2tbV1f3nDWJVOWDDz4ICQnJzMx88ODB559/7ubmhh/xAOANo2AEHRwcPGbMmObm5s8+++zevXuz\nZ88OCAhQfmTPFxgYaGxsvHPnzvr6+jFjxnTp0tLwRkEVBzCHgjpoqVT622+/EQQxefLk/Pz8uLi4\n+fPn02R8quQ6aAAAFXrmhSotLS3l5eUmJib0Wa2fQIIGAHWi4LteeXn51KlT+Xx+jx49tLS0pk6d\nWl5ervzIALoEqjiAORQk6LCwMG1t7eLiYolEUlxczOfzFyxYoPzIALoE7qgCzKFg+uKvv/4qLCyk\nVg4yNTXdunWrvOIYAACURsEI2sLCovUN+m7cuEGtBw3wJkAVBzCHghH0xo0bRSLRpEmT7O3tCwoK\nYmNjo7B8Abwx8GEG5lAwlAgKCrp+/Xq/fv0aGhr69et3/fr1yZMnKz8yAAA1p7iEzsHBYenSpUoO\nBUAZcEcVYA4FCbr92sq6urqmpqZjx44NDw/X1tZWRlwAXQRrcQBzKEjQa9eujYqK+vjjj21sbB48\neLBly5Zp06bZ29vv3Lnzn3/+eeGN+wAA4LVQkKDXrFmTlpZGVW7069fPw8Nj+PDhd+7c8fX1dXR0\nVFpkZ86c+fXXX9s05ufnDxgwQOH2p0+fjouL09DQeOedd4YMGdL1AQIzoYoDmEPxHHRRUZG8tK6o\nqKimpoYgCOq/SuPh4dH+Lq6JiYlaWlrtN960adO9e/c++uijhoaGiIiIqVOnTpkyRSlhAtOgigOY\nQ0GCjoiICAwMDA0NtbOzKywsPHjw4Lp165KTk0NCQpT5y6GhoaGHh0ebxry8PLFY3KZRJpPFxcWl\npKSwWCyCIH7++efRo0cjQQMA0yn4rhcaGvrXX39paGhcvXqVw+HEx8fPnz/f3Nz8+PHjy5cvV36I\nLyQWi7t160ZlZ4IgNDU15Y8B2sJaHMAcCkbQixcvFolEX3zxBYfDkTf26tVLiVG9GjMzswcPHtTX\n11OzH2VlZbg5ITwTqjiAORQkaENDw/fff7+0tHTy5MkikcjX15dWK44qtGLFiokTJ86aNauxsXH/\n/v1ff/21qiMCAOgsBeOIiIiIW7duXbp0qUePHmvXrrW2tg4LC0tISGhublZ+fC9p4sSJkZGRNTU1\nJEnGxMQMHDhQ1REBXaGKA5jjmZ9UIyMjGxub7t27NzU1Xbp0ae3atUKh8MSJE8oM7oVKSkquX78u\nkUgIgrCzswsLC5s1a5apqamq4wIai4pCggamUPBJ3bx587Bhw6ytrfft29e/f//09PSMjIxLly4d\nOnSIPgtDy2SyBQsWhIWF/fTTT2PHjo2OjlZ1RAAAr5mCyeXs7OzFixcHBAQIBILW7QMGDNi9e7ey\nAnuBqKgoBweH77//niAImUw2fvx4Ly8vrFsNL4a1OIA5FHxMJ0yYMHLkyNbZ+auvviIIQkdH5+23\n31ZeaM917ty5d999l3rMZrODgoIuXryo2pCAGXBHFWAOBQn67bff9vDwuHbtmrxl3bp1SgzppRgY\nGFRWVsqfVlZWGhgYqDAeAIDXTkGC1tfX37Zt24QJE7Zt2yaj61hj+vTpn3zyCZWjMzMzY2Njhw0b\npuqggAlQxQHMobjAedy4cVeuXJk2bdqZM2cOHjyo3JBeyuDBgxcvXhwSEtLU1GRpafnDDz+0mTEH\nUAxrcQBzPPMKFGtr67Nnz0ZERPTv37+hoUGZMb2kgICAgIAAVUcBANBVFHzXk5dqcLncdevW/fDD\nD/Kf4wAYD2txAHMoGEFPmzaNIIiWlpby8nITE5MRI0aMGDFC6YEBdA2sxQHMoeBjWl5ePnXqVD6f\n36NHDy0tralTp5aXlys/MgAANacgQYeFhWlraxcXF0skkuLiYj6fT58LCAE6C1UcwBwskiTbNOnp\n6RUWFsrvZlJRUeHg4PDkyROlx6bA0aNHxWJxeHi4qgMBAOhyCoYSFhYW6enp8qc3btyQ3/4KAACU\nRsGPhBs3bhSJRJMmTbK3ty8oKIiNjY16TaWjJEnW1NTo6uqy8R0TVAVrcQBzKPiYBgUFXb9+vV+/\nfg0NDf369bt+/frkyZM700ddXd2GDRucnJz4fL6+vj6Px3N0dIyIiGhsbOzMYQE6AmtxAHMoGEG7\nurpGRka+xvvDhoeHFxUVRUZGOjs76+npVVdXZ2dnb9q0KTw8fP/+/a+rFwCAN4yCBD1lypRdu3a5\nu7vzeLzX0kdsbGxWVpaVlRX11NjY2Nvb283NTSgUIkGDsqGKA5hDQYI+c+bMzZs39fT0rK2t5Xcj\nzMnJ6XAfQqEwPj5+zpw5rRv/+OMPW1vbDh8ToIOwFgcwh4IE/e23377ePiIjI0Ui0ZYtW1xcXAQC\ngUQiyczMrKysjIuLe70dAQC8SRQkaGdnZ4IgWlpaysrKLCwsWCxWJ/vw9PTMzc1NTk4uKCgQi8WG\nhoZhYWF+fn70v1k4vIFQxQHMoSBFFhUVzZgxIzU1lcPhpKSkLFy4MDo6upN3k+Jyuf379x8xYoQ8\n3be0tIjFYhMTk84cFuCVYS0OYA4FH9Nly5b16tWroqJCR0fHxcXFy8tr3rx5nekjMzOzb9++xsbG\nPXr0OHXqFNX44MED3H4bAOA5FCTohISE9evXa2lpEQTB4XCWL1+elpbWmT7mz58fFBTU0NBw4MCB\nBQsWtL6ZFoCyoYoDmEPBJ9Xa2rr1DVivXLnSyfmNGzduLFu2jMfj+fr67tq1a8GCBS0tLZ05IEDH\nRUUhQQNTKJiD3rFjR1BQ0IgRI6qrq0UiUUpKSnR0dGf6cHR0TEhIEIlEBEFMmDDh4MGDq1evfuG0\nyd9//91+5H7t2jUbG5vOBAMAwBQKVrOrra1taGg4efLkw4cPLS0tx44dKxAIdHR0OtxHYmLi5MmT\nnZ2dT5w4YWZmJhaLR40aRZLkzZs32/cud/fu3Vu3brVpTE1NNTEx+fTTTzscDKg7VHEAczw1gpZK\npQRB9OnTJy8vT36bK4lEYmlpWV1d3eE+AgIC7t69e/78eWpe28TEJDU1NTY29vr168/Zy8nJycnJ\nqU0jSZJisbjDkQCgigMY5KkEzefzCYJoaWmhHshRsxOdYWlpOXXqVPlTTU1Nb2/vzozKAQDeeApG\n0KNGjUpISOjqjlNTU0NDQyUSSVd3BPAUVHEAcygusyMIoqWl5dGjR8+ZI+4kkUiE7AwqgCoOYA4F\nn9SioiJ/f3+BQODk5HTr1i0vL6/8/PzX0hlJktXV1TKsxgsA8BKUcSUhFuwHGlm5Egv2A1MoqINO\nSEjIy8trfSWhnZ1dZ/rAgv1AI6jiAOZQkKCpKwnHjh1LPe38lYRYsB8AoAMUjCN27Ngxc+bMKVOm\nUFcSzpo16+uvv+5MH9SC/W0asWA/qAaqOIA5FFxJSBBERUVF6ysJLSwsOtPHtWvXRCKRlpZW+wX7\nPTw8XulQR48eFYvF4eHhnYkHAIARFCfo104qlbZesL979+4dW7AfCRoA1MdTKXLkyJHP2u7MmTOd\n6obL9ff378wRAF4PrMUBzPFUgg4ODpY/fv/999esWYM7nsCbBlUcwBxPJei5c+fKHy9btiw4ONje\n3l7ZEQEAAEEQCqs4AN5kqOIA5sB9tUHNREWpOgKAl/VUgn748KH8sUwma/0Ucx0AAEr2VIJuczcp\nHx8f+WPlVOMBdDlUcQBzPJWgkYXhzYcqDmAOfEwBAGgKCRrUDKo4gDlQxQFqBlUcwBwYSgAA0BQS\nNKgZ3FEFmAMJGtQMVcUBwARI0AAANIUEDWoGVRzAHKjiADWDKg5gDgwlAABoCgka1AyqOIA56DvF\nUV5efv/+/TaN+fn5PB5PJfHAGwJrcQBz0DdBZ2VlxcfHt2nMycnp27evSuIBAFAy+iZoPz8/Pz+/\nNo3UXb1VEg+8IVDFAcxB3wQN0CVQxQHMgaEEAABNIUGDmkEVBzAHEjSoGazFAcyBBA0AQFNI0KBm\nUMUBzIEqDlAzqOIA5sBQAgCAppCgQc2gigOYAwka1AyqOIA5kKABAGgKCRrUDKo4gDlQxQFqBlUc\nwBwYSgAA0BQSNKgZVHEAcyBBg5pBFQcwBxI0AABNIUGDmkEVBzCHUj+pJElWV1fL8AUTVCgqCgka\nmEIZn9S6uroNGzY4OTnx+Xx9fX0ej+fo6BgREdHY2KiE3gEAGEoZCTo8PPzcuXORkZHFxcVNTU2l\npaUHDhxIT08PDw9XQu8AT0EVBzCHMi5UiY2NzcrKsrKyop4aGxt7e3u7ubkJhcL9+/crIQCA/0dV\ncWCWA5hAGR9ToVAYHx/fpvGPP/6wtbVVQu8AAAyljBF0ZGSkSCTasmWLi4uLQCCQSCSZmZmVlZVx\ncXFK6B3gKajiAOZgkSSphG6kUmlycnJBQYFYLDY0NOzevbufnx+X+8p/Ho4ePSoWizF5DQDqQEmL\nJXG5XH9/f+X0BQDwZsB3PVAzqOIA5lDGCDonJ+dZL/Xq1UsJAQD8P1RxAHMoI0GHhoZevnxZW1vb\n0NCwzUsPHz5UQgAAAEykjASdmpo6Z84cbW3tb7/99uX3kkqlNTU1bRpra2tfa2igflDFAcyhjATN\nYrHeeeedGzduvNJecXFx3333XZvG0tLSUaNGvb7QQP3gjirAHEoqs3tdUGYHAOpDZd/1ioqKTp06\npareQX2higOYQ2UJOjU1NTg4WFW9g/rCHVWAOVSWoEUikUQiUVXvAAD0hwX7Qc2gigOYAwv2g5rB\nHVWAObBgPwAATSmjzM7AwKD1gv2U24HI4wAADgpJREFUmpoaoVAoFotf6VAos4POWrmSWL8eg2hg\nBCzYD2oGVRzAHFiwHwCAppSRoD09PXNzc1sv2B8WFtaxBfsBOgtVHMAcWLAf1AzW4gDmwFACAICm\nkKBBzWAtDmAOJGhQM6jiAOZAggYAoCkkaFAzqOIA5kChG6gZVHEAc2AoAQBAU0jQoGZQxQHMgQQN\nagZVHMAcSNAAADSFBA1qBlUcwByo4gA1gyoOYA4MJQAAaAoJGtQMqjiAOZCgQc2gigOYAwkaAICm\nkKBBzaCKA5gDVRygZlDFAcyBoQQAAE3RdwR9/Pjx77//vk1jaWnpyJEjVRIPvCFWriTWr8csBzAC\nfRN0UFBQUFBQm8ajR4+KxWKVxANvCKqKAwkamAAfUwAAmkKCBjWDKg5gDvpOcQB0CVRxAHNgKAEA\nQFNI0KBmsBYHMAcSNKgZrMUBzIEEDQBAU0jQoGZQxQHMgSoOUDOo4gDmwFACAICmkKBBzaCKA5gD\nCRrUDKo4gDmQoAEAaAoJGtQMqjiAOVDFAWoGVRzAHBhKAADQFBI0qBlUcQBzIEGDmkEVBzAHEjQA\nAE0pNUGTJFldXS3D+AVUCFUcwBzK+KTW1dVt2LDBycmJz+fr6+vzeDxHR8eIiIjGxkYl9A7wlKgo\nJGhgCmV8UsPDw8+dOxcZGVlcXNzU1FRaWnrgwIH09PTw8HAl9A4AwFAskiS7ug8DA4OsrCwrK6vW\njTU1NUKhUCwWv9Khjh49KhaLkdmh41auJNavxyAaGEEZH1OhUBgfH9+m8Y8//rC1tVVC7wBPQRUH\nMIcyriSMjIwUiURbtmxxcXERCAQSiSQzM7OysjIuLk4JvQMAMJQyErSnp2dubm5ycnJBQYFYLDY0\nNAwLC/Pz8+NycaE5KB2qOIA5lJQiuVyuv78/SZI1NTW6urps/B8CqoK1OIA5UGYHAEBTKLMDNYO1\nOIA5lDHFERsb27rMztjY2Nvb283NTSgU7t+/XwkBAPw/qooDk2zABMpI0FSZ3Zw5c1o3vrDMLi4u\n7tChQ20ai4qKfHx8Xn+IAAD0o4wLVa5duyYSibS0tNqX2Xl4eDxrr+bmZolE0qZRJpMJBAIej9fF\nIcOba/ZsYt8+jKCBEZSRoAmCkEqlrcvsunfvjjI7AIDnU1KCbq+oqOjGjRvjxo1TSe8AAPSnsi96\nqampwcHBquod1BeqOIA5VDaCBlCNgAAiPp7A9BowARbsBwCgKVxJCGoGa3EAcyhjiiM0NLSoqGjV\nqlXOzs56enrV1dXZ2dmbNm0yNTXFhSoAAM/CsAX7AQDUBxbsBzWDKg5gDvpeSQjQJVDFAcyBKwlB\nzSBBA3O8yXXQnp6ehoaGqo7izVdTU1NbW2thYaHqQF7Ksqysr3v3JlksVQfSEfn5+XZ2dhwOR9WB\nvOFaWlq0tbVPnTql6kCUdUcVlRAIBImJiaqO4s2XmJh45cqVzz77TNWBvKzRqg6gw4KCgiIjI42M\njFQdyBuurKzs/fffV3UUBKHCS70BAOD5kKABAGgKCRoAgKaQoAEAaAoJGgCApt7kBI06a+XgcDgo\n/FIONpvNxkpPXY8+5/lNroOuqakRCASqjuLN19LS0tjYqK2trepA3nz4SCsNTU71m5ygAQAYjRbD\neAAAaA8JGgCAppCgAQBoCgkaAICmkKABAGgKCRoAgKaQoAEAaIp5CfratWv9+/c3NDScNWtWfX39\ny7z6rF1Ikhw8eHBOTo7yomemDpxzAqe3c55/zgmc3q5En3PLsAQtlUpFIlF4eHhGRsaDBw+2bdv2\nwleftctff/01Z86cy5cvq+BtMEoHzjmB09s5zz/nBE5vV6LVuWVYgk5OTtbS0po7d263bt1WrVp1\n8ODBF776rF1u3Lihra2NC5RfqAPnnMDp7Zznn3MCp7cr0ercMmw5oYKCAhcXF+qxi4tLYWEhSZKs\n/91fTuGrz9pl2bJlBEHExsYq/U0wTAfOOU5vJz3/nBMEgdPbdWh1bhk2ghaLxfIVTPT09Jqammpq\nap7/6vN3gRfqwDlXQZRvFpxVoDAsQRsaGkokEupxdXU1l8vV1dV9/qvP3wVeqAPnXAVRvllwVoHC\nsATt4OCQmZlJPc7Ozra3t2+9bKvCV5+/C7xQB865CqJ8s+CsAoVh/+rDhg2rrKw8duyYRCLZvHnz\nu+++S7UfO3asqKhI4avP2gVeUgfOOXTS88+5amMDpSKZ5urVq25ubkZGRqGhoQ0NDVSjjo7OyZMn\nn/WqwkZKt27dsrOzlfwWGKcD55yC09thzz/nFJzerkOTc4sF+wEAaIphUxwAAOoDCRoAgKaQoAEA\naAoJGgCAppCgAQBoCgkaAICmkKABAGgKCRoAgKaQoAEAaAoJGgCAppCgAQBoCgkaAICmkKABAGgK\nCRoAgKaQoAEAaAoJGgCAppCgAQBoCgkaAICmkKABAGgKCVpNWVhYsP7H3t7++++/79LuuFyuVCq9\ndu2ap6fn6z3yxx9/bGhoWFZW9jIbv1IAXRFtJ1GnUdVRgPLgprFqysLC4vDhw+7u7nV1dadOnVq0\naFF6erqLi0sXdcflchsaGp48eXL58uXAwMDXeGRDQ8Ps7GwLC4uX2biiouLlA3iljZWDOo1cLlfV\ngYCSYAStvgQCgYGBgZWVVVhYWO/evW/evJmRkTFs2LAvv/zS1dWVIIiYmJiePXvq6+sHBQWVl5cT\nBHHo0KF58+bNmDHDwMDAy8vrzp071KHab9n6UKNGjWppaenevXtGRsbq1auftUtOTo63t/eWLVu6\ndesmFArPnj3bJuD2u7z99ttPnjwZOHAg9ZQ6iJeX17Jly0xMTLy9vVNTUwcMGCAQCJYuXUoQxL17\n96gApFLpe++9Z2hoaGJism7dOoUt8o2fFdjBgwft7e3t7e1/+OEHe3v7NtEqfDUlJcXd3d3Y2Hja\ntGnFxcXPOXj7Nys/jVOnTt28eTO12Zo1axYvXvysf5f23bV/m0BrJKglc3Pzq1evkiTZ3Nz8559/\n8vn8ixcv3r59W19ff/r06Wlpafn5+fr6+gkJCRUVFaGhoVOnTiVJMjo6msvl7t69u6ysbPny5W5u\nbjKZTOGWrQ9FkiSHw2lubr569aqHhwdJkgp3yc7O1tHR2bhxY21t7fLly4cMGdI6YIW7kCSpr69f\nU1Mj3yw7O5vFYu3fv7+iosLDw8PMzKygoCA1NZUgiLKyMnkAR48e7dmz5717965fv66pqZmbm9u+\nRb6xwsD+/vtvExOTy5cvFxUVeXt729nZtY5W4atisdjY2PjkyZOPHz9+7733RowY8ayDP+vNUqfx\n+PHjw4YNo1pcXV2Tk5MV/rso7K7923ydnyp43ZCg1ZS5ubmOjo6+vr6WlhaLxfr4449Jkrx9+7am\npmZjYyNJktu2bZs5cya1cVlZmYaGhlQqjY6OdnV1pRqbmpqMjIzu3r2rcMvWhyLbJWiFu2RnZ+vp\n6TU3N1OR9OzZs3XACnchFSVoIyMj6iArVqyYP38+1W5nZ/fPP/+0TtAODg6XL1+WyWTl5eWNjY3t\nW1on6PaBffrpp9RJI0nyxIkTbRK0wlcPHjwoEomoxtraWi0trWe962e9Weo0VldX6+joVFVV5eXl\nmZubP+vfRWF37d/m8z8noFqYzFJfP//8MzWVYWJioqurSzVaW1vzeDyCIEpKSuRfzE1NTXk8HvVF\nWygUUo0aGhr29vZFRUXP2lJ+qPaetYuFhQU1wdp+mlXhLgqnnk1NTeUHkW/Q5oCTJ0+urq4OCwsr\nLS1duHDhsmXL2re03r59YEVFRQMGDKAe29ratolB4asPHjw4ffq0PCQOh0P9ttn+4M9/swKBwMvL\nKzEx8f79+5MnT+ZwOISifxeF3T3/bQLdYA5afVlZWVGTpPLsTLTKERYWFoWFhdRjaqhlYmJCEMS9\ne/eoRqlUev/+fUtLy2dt+Zzfsp61C4vFetVdOiY/P3/EiBE3b968fPnyyZMnY2Ji2re03r59YJaW\nlvfv36ceP3jw4GVetbS0DAwMLCkpKSkpKSoqSklJ+b/27d41dSiM4zgI/g++gKWKf4BYkCxSV99A\nFEqWroJCJ10EJ2cndXCxQxYnUUFEsRBw1EkoBUcXQRRBFHEQ7hCQUl+qVui5934/UzgnPMmTAz9j\nSLT03C/+bbN+v7/RaFSr1Ugkoo3sr8vBw51uE6IhoHFYIBCoVCpvb2/z+TyRSIRCIS1wB4NBsVic\nTqfpdNpkMtnt9mN7frFcLr8tfsX5XKderz89Pb2/v2+3281ms16v90dOV4hEIq+vr/1+fzweZ7PZ\nc2Z9Pl+32202m7PZLJVKxePxYz9IJ5rVLqPf76/VasPh0O12a+P763LwcJe2id9FQOMwm81WKpVi\nsdjd3d1iscjn89q41+vtdDpWq1VV1XK5rNPpju35WTgctlgsq9XqdPErzuc60WjUbDa7XK6HhwdJ\nkp6fn/dHTldwOp2ZTCYYDD4+Psqy/PlfyLFZg8GgKEoymbRYLL1eT1GUS5vdXcb7+3uTyRQMBnfB\nvb8uBw93aZv4Zb/9EBx/E0VRdm8U/Oc+Pj5UVdW2W63W7rWKc2ZvQpKkVqulbbMu/yruoIFrzOdz\nWZYnk8l6vc7lcl++Zzk9+0Or1UpV1dFo5PF4blgWAiKggWtIkvTy8uJwOOx2u9FojMfj58/+ULvd\nlmW5UCjo9fobloWA+NQbAATFHTQACIqABgBBEdAAICgCGgAERUADgKAIaAAQFAENAIIioAFAUAQ0\nAAiKgAYAQRHQACAoAhoABEVAA4CgCGgAEBQBDQCC+gNxPPcRiQGNMQAAAABJRU5ErkJggg==\n"
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "'''Plot missingness versus heterozygosity. Use Rmagic.'''\n",
    "\n",
    "%Rpush imiss_file het_file\n",
    "%R imiss <- read.table(imiss_file, header=T)\n",
    "%R het   <- read.table(het_file,   header=T)\n",
    "\n",
    "#Calculate the log10 of the F_MISS and the mean heterozygosity (among non-missing genotypes).\n",
    "%R imiss$logF_MISS <- log10(imiss$F_MISS)\n",
    "%R het$meanHet     <-  (het$N.NM. - het$O.HOM.)/het$N.NM.\n",
    "\n",
    "%R -o summary_logF_MISS summary_logF_MISS <- summary(imiss$logF_MISS)\n",
    "print(\"Summary of log10(missingness): \" + str(summary_logF_MISS))\n",
    "\n",
    "%R -o summary_meanHet summary_meanHet <- summary(het$meanHet)\n",
    "print(\"Summary of mean heterozygosity: \" + str(summary_meanHet))\n",
    "\n",
    "#Plot the logF_MISS against the mean heterozygosity.\n",
    "#You can modify this plotting command to focus on your range of interest.\n",
    "%R plot(imiss$logF_MISS, het$meanHet, xlim=c(-3,0), ylim=c(0,0.5), pch=1, xlab=\"Proportion of missing genotypes\", ylab=\"Heterozygosity rate\",axes=F); axis(2,at=c(0,0.05,0.10,0.15,0.2,0.25,0.3,0.35,0.4,0.45,0.5),tick=T); axis(1,at=c(-3,-2,-1,0),labels=c(0.001,0.01,0.1,1)); abline(h=mean(het$meanHet) - (2 * sd(het$meanHet)), col=\"RED\",  lty=2); abline(h=mean(het$meanHet) - (3 * sd(het$meanHet)), col=\"gray\", lty=2); abline(h=mean(het$meanHet) + (2 * sd(het$meanHet)), col=\"red\",  lty=2); abline(h=mean(het$meanHet) + (3 * sd(het$meanHet)), col=\"gray\",  lty=2); abline(v=-1.522879, col=\"red\", lty=2)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "'''Write file containing FID:IID pairs for individuals with outlying heterozygosity'''\n",
    "%R het.filt <- het[which( het$meanHet > (mean(het$meanHet) + 3 * sd(het$meanHet)) | het$meanHet < (mean(het$meanHet) - 3 * sd(het$meanHet)) ),]\n",
    "%R write.table(het.filt[,c(0,1)], file=\"fail-het-qc.txt\", sep=\"\\t\", row.names=F, quote=F)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of individuals who failed heterozygosity QC:['103']\n"
     ]
    }
   ],
   "source": [
    "fail_het_qc_count = !cat fail-het-qc.txt | wc -l\n",
    "\n",
    "print(\"Number of individuals who failed heterozygosity QC:\" + str(fail_het_qc_count))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "PLINK v1.90b4.1 64-bit (30 Mar 2017)           www.cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2017 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to merged_neighbor.log.\n",
      "Options in effect:\n",
      "  --bfile merged_neighbor\n",
      "  --maf 0.01\n",
      "  --make-bed\n",
      "  --out merged_neighbor\n",
      "\n",
      "64329 MB RAM detected; reserving 32164 MB for main workspace.\n",
      "Note: --make-bed input and output filenames match.  Appending '~' to input\n",
      "filenames.\n",
      "547812 variants loaded from .bim file.\n",
      "5183 people (2299 males, 2752 females, 132 ambiguous) loaded from .fam.\n",
      "Ambiguous sex IDs written to merged_neighbor.nosex .\n",
      "5051 phenotype values loaded from .fam.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 5183 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 0%\b\b1%\b\b2%\b\b3%\b\b4%\b\b5%\b\b6%\b\b7%\b\b8%\b\b9%\b\b10%\b\b\b11%\b\b\b12%\b\b\b13%\b\b\b14%\b\b\b15%\b\b\b16%\b\b\b17%\b\b\b18%\b\b\b19%\b\b\b20%\b\b\b21%\b\b\b22%\b\b\b23%\b\b\b24%\b\b\b25%\b\b\b26%\b\b\b27%\b\b\b28%\b\b\b29%\b\b\b30%\b\b\b31%\b\b\b32%\b\b\b33%\b\b\b34%\b\b\b35%\b\b\b36%\b\b\b37%\b\b\b38%\b\b\b39%\b\b\b40%\b\b\b41%\b\b\b42%\b\b\b43%\b\b\b44%\b\b\b45%\b\b\b46%\b\b\b47%\b\b\b48%\b\b\b49%\b\b\b50%\b\b\b51%\b\b\b52%\b\b\b53%\b\b\b54%\b\b\b55%\b\b\b56%\b\b\b57%\b\b\b58%\b\b\b59%\b\b\b60%\b\b\b61%\b\b\b62%\b\b\b63%\b\b\b64%\b\b\b65%\b\b\b66%\b\b\b67%\b\b\b68%\b\b\b69%\b\b\b70%\b\b\b71%\b\b\b72%\b\b\b73%\b\b\b74%\b\b\b75%\b\b\b76%\b\b\b77%\b\b\b78%\b\b\b79%\b\b\b80%\b\b\b81%\b\b\b82%\b\b\b83%\b\b\b84%\b\b\b85%\b\b\b86%\b\b\b87%\b\b\b88%\b\b\b89%\b\b\b90%\b\b\b91%\b\b\b92%\b\b\b93%\b\b\b94%\b\b\b95%\b\b\b96%\b\b\b97%\b\b\b98%\b\b\b99%\b\b\b\b done.\n",
      "Warning: 11574 het. haploid genotypes present (see merged_neighbor.hh ); many\n",
      "commands treat these as missing.\n",
      "Warning: Nonmissing nonmale Y chromosome genotype(s) present; many commands\n",
      "treat these as missing.\n",
      "Total genotyping rate is 0.999488.\n",
      "19334 variants removed due to minor allele threshold(s)\n",
      "(--maf/--max-maf/--mac/--max-mac).\n",
      "528478 variants and 5183 people pass filters and QC.\n",
      "Phenotype data is quantitative.\n",
      "--make-bed to merged_neighbor.bed + merged_neighbor.bim + merged_neighbor.fam\n",
      "... 0%\b\b1%\b\b2%\b\b3%\b\b4%\b\b5%\b\b6%\b\b7%\b\b8%\b\b9%\b\b10%\b\b\b11%\b\b\b12%\b\b\b13%\b\b\b14%\b\b\b15%\b\b\b16%\b\b\b17%\b\b\b18%\b\b\b19%\b\b\b20%\b\b\b21%\b\b\b22%\b\b\b23%\b\b\b24%\b\b\b25%\b\b\b26%\b\b\b27%\b\b\b28%\b\b\b29%\b\b\b30%\b\b\b31%\b\b\b32%\b\b\b33%\b\b\b34%\b\b\b35%\b\b\b36%\b\b\b37%\b\b\b38%\b\b\b39%\b\b\b40%\b\b\b41%\b\b\b42%\b\b\b43%\b\b\b44%\b\b\b45%\b\b\b46%\b\b\b47%\b\b\b48%\b\b\b49%\b\b\b50%\b\b\b51%\b\b\b52%\b\b\b53%\b\b\b54%\b\b\b55%\b\b\b56%\b\b\b57%\b\b\b58%\b\b\b59%\b\b\b60%\b\b\b61%\b\b\b62%\b\b\b63%\b\b\b64%\b\b\b65%\b\b\b66%\b\b\b67%\b\b\b68%\b\b\b69%\b\b\b70%\b\b\b71%\b\b\b72%\b\b\b73%\b\b\b74%\b\b\b75%\b\b\b76%\b\b\b77%\b\b\b78%\b\b\b79%\b\b\b80%\b\b\b81%\b\b\b82%\b\b\b83%\b\b\b84%\b\b\b85%\b\b\b86%\b\b\b87%\b\b\b88%\b\b\b89%\b\b\b90%\b\b\b91%\b\b\b92%\b\b\b93%\b\b\b94%\b\b\b95%\b\b\b96%\b\b\b97%\b\b\b98%\b\b\b99%\b\b\bdone.\n"
     ]
    }
   ],
   "source": [
    "'''4) MAF'''\n",
    "!plink --bfile $input_filestem --min_maf 0.01 --make-bed --out $input_filestem"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "PLINK v1.90b4.1 64-bit (30 Mar 2017)           www.cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2017 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to merged_neighbor.log.\n",
      "Options in effect:\n",
      "  --bfile merged_neighbor\n",
      "  --indep-pairwise 50 5 0.2\n",
      "  --out merged_neighbor\n",
      "\n",
      "64329 MB RAM detected; reserving 32164 MB for main workspace.\n",
      "528478 variants loaded from .bim file.\n",
      "5183 people (2299 males, 2752 females, 132 ambiguous) loaded from .fam.\n",
      "Ambiguous sex IDs written to merged_neighbor.nosex .\n",
      "5051 phenotype values loaded from .fam.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 5183 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 0%\b\b1%\b\b2%\b\b3%\b\b4%\b\b5%\b\b6%\b\b7%\b\b8%\b\b9%\b\b10%\b\b\b11%\b\b\b12%\b\b\b13%\b\b\b14%\b\b\b15%\b\b\b16%\b\b\b17%\b\b\b18%\b\b\b19%\b\b\b20%\b\b\b21%\b\b\b22%\b\b\b23%\b\b\b24%\b\b\b25%\b\b\b26%\b\b\b27%\b\b\b28%\b\b\b29%\b\b\b30%\b\b\b31%\b\b\b32%\b\b\b33%\b\b\b34%\b\b\b35%\b\b\b36%\b\b\b37%\b\b\b38%\b\b\b39%\b\b\b40%\b\b\b41%\b\b\b42%\b\b\b43%\b\b\b44%\b\b\b45%\b\b\b46%\b\b\b47%\b\b\b48%\b\b\b49%\b\b\b50%\b\b\b51%\b\b\b52%\b\b\b53%\b\b\b54%\b\b\b55%\b\b\b56%\b\b\b57%\b\b\b58%\b\b\b59%\b\b\b60%\b\b\b61%\b\b\b62%\b\b\b63%\b\b\b64%\b\b\b65%\b\b\b66%\b\b\b67%\b\b\b68%\b\b\b69%\b\b\b70%\b\b\b71%\b\b\b72%\b\b\b73%\b\b\b74%\b\b\b75%\b\b\b76%\b\b\b77%\b\b\b78%\b\b\b79%\b\b\b80%\b\b\b81%\b\b\b82%\b\b\b83%\b\b\b84%\b\b\b85%\b\b\b86%\b\b\b87%\b\b\b88%\b\b\b89%\b\b\b90%\b\b\b91%\b\b\b92%\b\b\b93%\b\b\b94%\b\b\b95%\b\b\b96%\b\b\b97%\b\b\b98%\b\b\b99%\b\b\b\b done.\n",
      "Warning: 11557 het. haploid genotypes present (see merged_neighbor.hh ); many\n",
      "commands treat these as missing.\n",
      "Warning: Nonmissing nonmale Y chromosome genotype(s) present; many commands\n",
      "treat these as missing.\n",
      "Total genotyping rate is 0.999482.\n",
      "528478 variants and 5183 people pass filters and QC.\n",
      "Phenotype data is quantitative.\n",
      "Pruned 31470 variants from chromosome 1, leaving 8845.\n",
      "Pruned 34444 variants from chromosome 2, leaving 8441.\n",
      "Pruned 28408 variants from chromosome 3, leaving 7252.\n",
      "Pruned 24781 variants from chromosome 4, leaving 6599.\n",
      "Pruned 25658 variants from chromosome 5, leaving 6629.\n",
      "Pruned 27777 variants from chromosome 6, leaving 6780.\n",
      "Pruned 22702 variants from chromosome 7, leaving 5909.\n",
      "Pruned 23939 variants from chromosome 8, leaving 5526.\n",
      "Pruned 20166 variants from chromosome 9, leaving 5140.\n",
      "Pruned 21921 variants from chromosome 10, leaving 5752.\n",
      "Pruned 20404 variants from chromosome 11, leaving 5252.\n",
      "Pruned 20103 variants from chromosome 12, leaving 5568.\n",
      "Pruned 15704 variants from chromosome 13, leaving 4152.\n",
      "Pruned 13656 variants from chromosome 14, leaving 3748.\n",
      "Pruned 12158 variants from chromosome 15, leaving 3528.\n",
      "Pruned 12136 variants from chromosome 16, leaving 3827.\n",
      "Pruned 10171 variants from chromosome 17, leaving 3601.\n",
      "Pruned 12147 variants from chromosome 18, leaving 3602.\n",
      "Pruned 6322 variants from chromosome 19, leaving 2738.\n",
      "Pruned 10163 variants from chromosome 20, leaving 3238.\n",
      "Pruned 6006 variants from chromosome 21, leaving 1808.\n",
      "Pruned 5867 variants from chromosome 22, leaving 1978.\n",
      "Pruned 9556 variants from chromosome 23, leaving 2823.\n",
      "Pruned 2 variants from chromosome 24, leaving 2.\n",
      "Pruned 0 variants from chromosome 25, leaving 3.\n",
      "Pruned 43 variants from chromosome 26, leaving 33.\n",
      "Pruning complete.  415704 of 528478 variants removed.\n",
      "Marker lists written to merged_neighbor.prune.in and merged_neighbor.prune.out\n",
      ".\n"
     ]
    }
   ],
   "source": [
    "'''5) IBD'''\n",
    "'''Generate LD-pruned variants first.'''\n",
    "!plink --bfile $input_filestem --indep-pairwise 50 5 0.2 --out $input_filestem"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'output_filestem' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-2-b042855a3db4>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      3\u001b[0m \u001b[0mThe\u001b[0m \u001b[0mgoal\u001b[0m \u001b[0;32mis\u001b[0m \u001b[0mto\u001b[0m \u001b[0mend\u001b[0m \u001b[0mup\u001b[0m \u001b[0;32mwith\u001b[0m \u001b[0mabout\u001b[0m \u001b[0;36m100\u001b[0m\u001b[0mk\u001b[0m \u001b[0mSNPs\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0ma\u001b[0m \u001b[0mGWAS\u001b[0m \u001b[0mstudy\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0;34m>=\u001b[0m \u001b[0;36m500\u001b[0m\u001b[0mk\u001b[0m \u001b[0mSNPs\u001b[0m \u001b[0mgenome\u001b[0m\u001b[0;34m-\u001b[0m\u001b[0mwide\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      4\u001b[0m Then proceed with identification of related samples.'''\n\u001b[0;32m----> 5\u001b[0;31m \u001b[0mprune_in_file\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0moutput_filestem\u001b[0m \u001b[0;34m+\u001b[0m \u001b[0;34m'.prune.in'\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      6\u001b[0m \u001b[0munlinked_variant_count\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mget_ipython\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mgetoutput\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34mu'cat $prune_in_file | wc -l'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      7\u001b[0m \u001b[0;32mprint\u001b[0m \u001b[0;34m\"Unlinked (LD-pruned) genetic variant count: \"\u001b[0m \u001b[0;34m+\u001b[0m \u001b[0mstr\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0munlinked_variant_count\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mNameError\u001b[0m: name 'output_filestem' is not defined"
     ]
    }
   ],
   "source": [
    "'''If the above output pruned too few or too many SNPs, adjust the linkage R^2 value \n",
    "Increase R^2 to retain more SNPs, decrease to prune more.\n",
    "The goal is to end up with about 100k SNPs for a GWAS study (>= 500k SNPs genome-wide).\n",
    "Then proceed with identification of related samples.'''\n",
    "prune_in_file = input_filestem + '.prune.in'\n",
    "unlinked_variant_count = !cat $prune_in_file | wc -l\n",
    "print \"Unlinked (LD-pruned) genetic variant count: \" + str(unlinked_variant_count)\n",
    "!plink --bfile $input_filestem --extract $prune_in_file --genome --out $input_filestem\n",
    "#That command should generate a *.genome file."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "genome_file = input_filestem + '.genome'\n",
    "#Clean up the whitespace:\n",
    "!perl -i.bak -lpe 's/^\\s+//; s/\\s+/\\t/g;' $genome_file\n",
    "genome_file_line_count = !wc -l $genome_file\n",
    "print \"Line count in genome file (used for relationship estimation): \" + str(genome_file_line_count)\n",
    "!head $genome_file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "'''Generate a file containing FID:IID pairs of individuals to be excluded on the basis of relatedness.\n",
    "Exclude the individual with greater missingness from any pair of individuals that are related more than first cousins.'''\n",
    "imiss, removed = dict(), set()\n",
    "\n",
    "with open(imiss_file) as imiss_in, open(genome_file) as genome_in, open('fail-IBD-QC.txt', 'w') as fail_IBD_out:\n",
    "    #Read the missingness into a dictionary.\n",
    "    header = imiss_in.readline()\n",
    "    for line in imiss_in:\n",
    "        fid, iid, miss_pheno, n_miss, n_geno, f_miss = line.rstrip().split(\"\\t\") #File was already whitespace-cleaned.\n",
    "        imiss[(fid, iid)] = float(f_miss) #Key by FID:IID tuple, value by F_MISS (missing frequency).\n",
    "        \n",
    "    #Now exclude related samples. Drop whichever one has higher missingness.\n",
    "    header = genome_in.readline()\n",
    "    for line in genome_in:\n",
    "        fid1, iid1, fid2, iid2, rt, ez, z0, z1, z2, pi_hat, phe, dst, ppc, ratio = line.rstrip().split(\"\\t\")\n",
    "        if float(pi_hat) > 0.185:\n",
    "            if imiss[(fid1, iid1)] >= imiss[(fid2, iid2)]: #The first sample has higher missingness.\n",
    "                if (fid1, iid1) not in removed:\n",
    "                    fail_IBD_out.write(\"\\t\".join([fid1, iid1]) + \"\\n\")\n",
    "                    removed.add((fid1, iid1)) #Remember that we already excluded this FID:IID pair.\n",
    "            else: #The second sample has higher missingness.\n",
    "                if (fid2, iid2) not in removed:\n",
    "                    fail_IBD_out.write(\"\\t\".join([fid2, iid2]) + \"\\n\")\n",
    "                    removed.add((fid2, iid2))\n",
    "\n",
    "fail_IBD_count = !cat fail-IBD-QC.txt | wc -l\n",
    "print \"Number of samples that failed IBD QC: \" + str(fail_IBD_count)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "'''Now exclude the related samples from the study.'''\n",
    "!plink --bfile $input_filestem --remove fail-IBD-QC.txt --make-bed --out $input_filestem"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "'''Now that related samples have been excluded, repeat the marker callrate, sample callrate, and MAF filtering.'''\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "'''7) Marker callrate.'''\n",
    "!plink --bfile $input_filestem --geno $max_missing_marker_callrate --make-bed --out $input_filestem"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "'''8) Sample callrate.'''\n",
    "!plink --bfile $input_filestem --mind $max_missing_sample_callrate --out $input_filestem"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "'''9) MAF'''\n",
    "!plink --bfile $input_filestem --min_maf 0.01 --make-bed --out $input_filestem"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "'''10) HWE'''\n",
    "!plink --bfile $input_filestem --hwe $min_hwe_p --midp --include-nonctrl --make-bed --out $input_filestem\n",
    "'''If you wish to base this test only on cases,\n",
    "which requires your PED file to have accurate case/control phenotype encodings,\n",
    "you can omit the --include-nonctrl option.\n",
    "The mid-p modifier is recommended by Plink.'''"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "'''11) Differential missingness (differential sample callrate).\n",
    "This step relies on accurate case/control phenotype encodings in the FAM file.'''\n",
    "!plink --bfile $input_filestem --test-missing --out $input_filestem\n",
    "diffmiss_file = $input_filestem + '.missing'\n",
    "!perl -i.bak -lpe 's/^\\s+//; s/\\s+/\\t/g;' {diffmiss_file} #Clean up whitespace.\n",
    "diffmiss_snp_count = 0\n",
    "with open('fail-diffmis-qc.txt', 'w') as fail_diffmiss, open(diffmiss_file) as diffmiss:\n",
    "    header = diffmiss.readline() #Skip header line.\n",
    "    for line in diffmiss:\n",
    "        CHROM, SNP, F_MISS_A, F_MISS_U, P = line.rstrip().split(\"\\t\")\n",
    "        if P < min_diffmis_p:\n",
    "            diffmiss_snp_count += 1\n",
    "            fail_difmiss.write(SNP + \"\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
