{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# GWAS QC and Analysis\n",
    "## Copyright Brian S. Cole, PhD 2017\n",
    "\n",
    "A Jupyter Notebook for Reproducible Quality Control of Genotypes and Genotyped Samples Followed by Genotype-Phenotype Analysis using Cloud Computing and Open Source Tools.\n",
    "\n",
    "This work presents an example GWAS workthrough.  Users are encouraged to use this example notebook to generate their own notebooks that document and visualize their GWAS analyses.  Commonly used tools are highlighted, including PLINK, R, Python, and Perl, however this notebook is flexible and users are encouraged to customize this workflow to fit the demands of their analysis.\n",
    "\n",
    "This document can therefore serve as a template for customization, then the resulting analysis can be easily shared in notebook form.  Note that this does not hold if the user develops external scripts/programs that are called from this notebook.  Those must be included separately, for example in a hardware virtual machine image (e.g. AMI) or container service."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Table of contents\n",
    "1. Introduction\n",
    "  1. Genotype-phenotype Association Studies (GWAS)\n",
    "  2. GWAS QC Protects Against Spurious Results\n",
    "  3. Individual-level QC Removes Samples\n",
    "  4. Variant-level QC Removes Genetic Variants\n",
    "  5. Harnessing the Cloud for Reproducible GWAS\n",
    "2. Individual-level QC\n",
    "  1. Gender Concordance\n",
    "  2. Missing Genotypes\n",
    "  3. Outlying Heterozygosity\n",
    "  4. Identity-by-Descent Control (IBD)\n",
    "  5. Removing Individuals that Failed QC\n",
    "3. Variant-level QC\n",
    "  1. Callrate Control (genotype missingness)\n",
    "  2. Differential Callrate (for case/control studies)\n",
    "  3. Hardy-Weinberg Equilibrium Control\n",
    "  4. Minor Allele Frequency Control\n",
    "  5. Removing Variants that Failed QC\n",
    "4. (Optional) Population Structure\n",
    "  1. Quantifying Population Structure\n",
    "  2. Estimating Genetic Relationships\n",
    "5. Case/control GWAS\n",
    "  1. Logistic regression\n",
    "  2. Covariate adjustment\n",
    "  3. Interpreting Results\n",
    "6. Quantitative Trait GWAS\n",
    "  1. Linear Regression\n",
    "  2. Covariate Adjustment\n",
    "  3. Interpreting Results\n",
    "7. Survival GWAS\n",
    "  1. Cox Proportional Hazards Regression\n",
    "  2. Covariate Adjustment\n",
    "  3. Interpreting Results\n",
    "8. Data Egress and Instance Termination\n",
    "  1. Moving Data and Results Out\n",
    "  2. Sharing Notebooks for Reproducible Analysis\n",
    "  3. Cleanup and Termination"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Introduction\n",
    "#### Genotype-Phenotype Association Studies (GWAS)\n",
    "The advent of low-cost, high-throughput genotyping technologies has enabled scientific research into the association between variation in phenotypes within a set of unrelated individuals sampled from a population and the variation in genotypes within this set of samples.  This analysis, commonly referred to as Genome-Wide Association Study or GWAS, has the power to identify particular genotypes which are associated with quantitative traits, case/control status, survival outcomes, and other traits.  \n",
    "\n",
    "Over the last decade, studies of thousands of traits have accumulated, ushering in a new era of knowledge integration. However, failure to replicate and reproduce GWAS results has led to a complicated situation in which the generalizability and utility of genotype-phenotype associations is not guaranteed by current analysis methodologies.  We propose that cloud computing and open source tools can provide enhanced reproducibility by coupling stateful, virtualized hardware with open-source software.  This document outlines a GWAS QC and analysis workflow which users can recreate using their own data, including modifications of the workflow to fit their needs.  The resulting notebook can be shared, version-controlled, and reproduced.\n",
    "\n",
    "#### GWAS QC Protects Against Spurious Results\n",
    "GWAS analyses typically require that several assumptions are met, otherwise the analysis is not valid.  The primary goal of GWAS Quality Control (QC) is to remove sources of false positive genotype-phenotype associations.  These sources of spurious results are the result of the statistical analysis, the technologies used to obtain genotypes, and the properties of genotypes within populations.\n",
    "\n",
    "This notebook will use a QC approach tooled around the PLINK software (GPLv3) by Shaun Purcell and Christopher Chang, which is available at http://pngu.mgh.harvard.edu/purcell/plink/ and is described in the following work:\n",
    "\n",
    "Purcell S, Neale B, Todd-Brown K, Thomas L, Ferreira MAR, Bender D, Maller J, Sklar P, de Bakker PIW, Daly MJ & Sham PC (2007) PLINK: a toolset for whole-genome association and population-based linkage analysis. American Journal of Human Genetics, 81.\n",
    "\n",
    "The methodology behind the QC approach in this notebook is described in this work:\n",
    "\n",
    "Anderson CA, Pettersson FH, Clarke GM, Cardon LR, Morris AP, Zondervan KT. Data quality control in genetic case-control association studies. Nature protocols. 2010;5(9):1564-1573. doi:10.1038/nprot.2010.116.\n",
    "\n",
    "\n",
    "#### Individual-level QC Removes Samples\n",
    "The goal of individual-level QC phase is to remove samples (sometimes called \"subjects\") which should not be included in the analysis.  These reasons include gender discordance between phenotype and genotype, relatedness to other samples which would violate the assumption of unrelated samples that underlies GWAS hypothesis testing, potential contamination of multiple samples which would yield unreliable genotyping, and low genotyping coverage which might indicate degraded DNA and/or technical failure of DNA preparation.  \n",
    "\n",
    "#### Variant-level QC Removes Genetic Variants\n",
    "The goal of the latter phase is to remove genotypes of poor quality.  Reasons for exclusion of individual genotypes include departure from Hardy Weinberg Equilibrium which might denote clustering failures during genotype calling (preventing one of two homozygous genotypes at a biallelic locus from being called, for example), absence of genotype calls from a subset of samples, significantly different genotyping rate between cases and controls, and (optionally) removal of rare variants which are unlikely to give rise to statistically significant association results for complex traits.\n",
    "\n",
    "#### Harnessing the Cloud for Reproducible GWAS\n",
    "Cloud computing provides abstractions for scalable, elastic, durable, safe, affordable, on-demand, reproducible computation.  Importantly, this notebook can be run on an Amazon Web Services (AWS) Elastic Compute Cloud (EC2) instance coupled with encrypted Elastic Block Storage (EBS) volume to hold the user's data during analyiss.  The resulting notebook can be used to create a git repository for version control, cloning off the cloud, and sharing with others.  Importantly, the notebook can be customized and shared, documenting a user's analysis in a reproducible way.\n",
    "The goal is to utilize notebooks as a framework for analysis in the cloud, decoupling the analyst from the computational infrastructure as entirely as is possible.  The cloud then provides mechanisms for snapshotting data (EBS snapshots), storing notebooks for distribution and reuse (S3), containerizing workflows (ECS), collaboration and resource management, and much more.\n",
    "Users may also connect to their EC2 instances and run all commands documented here individually.  This is recommended for users who wish to develop customized tools (for example, scripts) to supplement this notebook.  It is important to note that any external tools that this notebook uses will then become reproducibility dependencies.  Users are therefore encouraged to use open source licenses and create their own virtual appliances which include their external tools (e.g. custom scripts) by using Amazon Machine Images or other technology.\n",
    "\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The rpy2.ipython extension is already loaded. To reload it, use:\n",
      "  %reload_ext rpy2.ipython\n"
     ]
    }
   ],
   "source": [
    "'''Prepare the notebook. Run this cell first.'''\n",
    "import os, re, sys\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "%load_ext rpy2.ipython"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/clink/NEIGHBOR\n"
     ]
    }
   ],
   "source": [
    "'''Make analysis directory and upload your data'''\n",
    "\n",
    "import os\n",
    "study_dir = '/home/brian/clink/NEIGHBOR/' #Can change this variable to suit your study.\n",
    "if not os.path.isdir(study_dir): os.mkdir(study_dir)\n",
    "%cd {study_dir}\n",
    "\n",
    "#You can either upload data to the EC2 instance using the key pair you specified at instance launch,\n",
    "#  for instance using scp from a system that has your data,\n",
    "#  or you can fetch the data from this instance using SFTP/rsync/curl/wget/s3\n",
    "\n",
    "#Remember that if you want to pull data from a private S3 bucket, the EC2 instance needs to assume an IAM role,\n",
    "#  and that bucket must be configured for the EC2 instance's IAM role.\n",
    "\n",
    "#In this example, I used scp from a firewalled and secured workstation to securely upload a Plink study to the EC2 instance.\n",
    "#I used a command like this from my workstation:\n",
    "#scp -i my_key_pair.pem my_study.* ec2-user@<IPv4_public_IP>:/gwas/my_study/\n",
    "#  ..where <IPv4_public_IP> is the public IPv4 address of the EC2 instance running this notebook.\n",
    "\n",
    "input_filestem  = 'merged_neighbor' #Set this variable to the Plink filestem of your study.\n",
    "output_filestem = 'clean-' + input_filestem #Generate a new filestem for post-QC data."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2: Individual-level QC\n",
    "Our strategy here will be to identify individuals with discordant genders, too many missing genotypes, and outlying heterozygosity.  We'll save the IDs of these individuals to text files, then at the end we'll combine these files and remove the individuals who failed individual-level QC.  We'll generate a new file with clean individuals, and this file will be used in variant-level QC."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 2A: Gender Concordance\n",
    "Gender concordance serves as a check of your phenotype and genotype data.  You should see very few, if any, instances in which the phenotypic gender, which is provided in the Plink \\*.fam file, does not match Plink's guess from looking at the genotypes.  All individuals who fail this step will be excluded from further analysis."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "PLINK v1.90b4.1 64-bit (30 Mar 2017)           www.cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2017 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to merged_neighbor.log.\n",
      "Options in effect:\n",
      "  --bfile merged_neighbor\n",
      "  --check-sex\n",
      "  --out merged_neighbor\n",
      "\n",
      "64329 MB RAM detected; reserving 32164 MB for main workspace.\n",
      "655216 variants loaded from .bim file.\n",
      "5183 people (2299 males, 2752 females, 132 ambiguous) loaded from .fam.\n",
      "Ambiguous sex IDs written to merged_neighbor.nosex .\n",
      "5051 phenotype values loaded from .fam.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 5183 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 0%\b\b1%\b\b2%\b\b3%\b\b4%\b\b5%\b\b6%\b\b7%\b\b8%\b\b9%\b\b10%\b\b\b11%\b\b\b12%\b\b\b13%\b\b\b14%\b\b\b15%\b\b\b16%\b\b\b17%\b\b\b18%\b\b\b19%\b\b\b20%\b\b\b21%\b\b\b22%\b\b\b23%\b\b\b24%\b\b\b25%\b\b\b26%\b\b\b27%\b\b\b28%\b\b\b29%\b\b\b30%\b\b\b31%\b\b\b32%\b\b\b33%\b\b\b34%\b\b\b35%\b\b\b36%\b\b\b37%\b\b\b38%\b\b\b39%\b\b\b40%\b\b\b41%\b\b\b42%\b\b\b43%\b\b\b44%\b\b\b45%\b\b\b46%\b\b\b47%\b\b\b48%\b\b\b49%\b\b\b50%\b\b\b51%\b\b\b52%\b\b\b53%\b\b\b54%\b\b\b55%\b\b\b56%\b\b\b57%\b\b\b58%\b\b\b59%\b\b\b60%\b\b\b61%\b\b\b62%\b\b\b63%\b\b\b64%\b\b\b65%\b\b\b66%\b\b\b67%\b\b\b68%\b\b\b69%\b\b\b70%\b\b\b71%\b\b\b72%\b\b\b73%\b\b\b74%\b\b\b75%\b\b\b76%\b\b\b77%\b\b\b78%\b\b\b79%\b\b\b80%\b\b\b81%\b\b\b82%\b\b\b83%\b\b\b84%\b\b\b85%\b\b\b86%\b\b\b87%\b\b\b88%\b\b\b89%\b\b\b90%\b\b\b91%\b\b\b92%\b\b\b93%\b\b\b94%\b\b\b95%\b\b\b96%\b\b\b97%\b\b\b98%\b\b\b99%\b\b\b\b done.\n",
      "Warning: 11950 het. haploid genotypes present (see merged_neighbor.hh ); many\n",
      "commands treat these as missing.\n",
      "Warning: Nonmissing nonmale Y chromosome genotype(s) present; many commands\n",
      "treat these as missing.\n",
      "Total genotyping rate is 0.849377.\n",
      "655216 variants and 5183 people pass filters and QC.\n",
      "Phenotype data is quantitative.\n",
      "--check-sex: 13098 Xchr and 0 Ychr variant(s) scanned, 151 problems detected.\n",
      "Report written to merged_neighbor.sexcheck .\n"
     ]
    }
   ],
   "source": [
    "'''Run sex check using plink.'''\n",
    "\n",
    "#This should generate a *.sexcheck file for our study\n",
    "!plink --bfile $input_filestem --check-sex --out $input_filestem"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of gender discordances detected: ['151']\n",
      "         FID          IID       PEDSEX       SNPSEX       STATUS            F\n",
      "      ME8001       ME8001            0            2      PROBLEM     -0.03585\n",
      "      ME8002       ME8002            0            1      PROBLEM            1\n",
      "      ME8003       ME8003            0            1      PROBLEM            1\n",
      "      ME8004       ME8004            0            2      PROBLEM    -0.004742\n",
      "      ME8005       ME8005            0            2      PROBLEM     -0.04665\n"
     ]
    }
   ],
   "source": [
    "sexcheck_file = input_filestem + '.sexcheck'\n",
    "sexprobs_file = input_filestem + '.sexprobs'\n",
    "\n",
    "!fgrep PROBLEM $sexcheck_file > $sexprobs_file\n",
    "sexproblems = !cat $sexprobs_file | wc -l\n",
    "print(\"Number of gender discordances detected: \" + str(sexproblems))\n",
    "\n",
    "#Save the FID and IID of the individuals that failed qc, if any.\n",
    "#It's okay to generate an empty file here - we'll count the lines of each fail-* file later for a sumamry.\n",
    "!perl -lne 's/^\\s+//; @f = split; print join \"\\t\", (@f[0..1])' $sexprobs_file >fail-sexcheck-qc.txt\n",
    "\n",
    "#At this point, inspect the reasons for any gender discordances.  In this case, some genders are missing in the ped file:\n",
    "!head -1 $sexcheck_file #For the header.\n",
    "!head -5 $sexprobs_file #Check out the first 5 discordances."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "PLINK v1.90b4.1 64-bit (30 Mar 2017)           www.cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2017 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to clean-merged_neighbor.log.\n",
      "Options in effect:\n",
      "  --bfile merged_neighbor\n",
      "  --geno 0.01\n",
      "  --make-bed\n",
      "  --out clean-merged_neighbor\n",
      "\n",
      "64329 MB RAM detected; reserving 32164 MB for main workspace.\n",
      "655216 variants loaded from .bim file.\n",
      "5183 people (2299 males, 2752 females, 132 ambiguous) loaded from .fam.\n",
      "Ambiguous sex IDs written to clean-merged_neighbor.nosex .\n",
      "5051 phenotype values loaded from .fam.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 5183 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 0%\b\b1%\b\b2%\b\b3%\b\b4%\b\b5%\b\b6%\b\b7%\b\b8%\b\b9%\b\b10%\b\b\b11%\b\b\b12%\b\b\b13%\b\b\b14%\b\b\b15%\b\b\b16%\b\b\b17%\b\b\b18%\b\b\b19%\b\b\b20%\b\b\b21%\b\b\b22%\b\b\b23%\b\b\b24%\b\b\b25%\b\b\b26%\b\b\b27%\b\b\b28%\b\b\b29%\b\b\b30%\b\b\b31%\b\b\b32%\b\b\b33%\b\b\b34%\b\b\b35%\b\b\b36%\b\b\b37%\b\b\b38%\b\b\b39%\b\b\b40%\b\b\b41%\b\b\b42%\b\b\b43%\b\b\b44%\b\b\b45%\b\b\b46%\b\b\b47%\b\b\b48%\b\b\b49%\b\b\b50%\b\b\b51%\b\b\b52%\b\b\b53%\b\b\b54%\b\b\b55%\b\b\b56%\b\b\b57%\b\b\b58%\b\b\b59%\b\b\b60%\b\b\b61%\b\b\b62%\b\b\b63%\b\b\b64%\b\b\b65%\b\b\b66%\b\b\b67%\b\b\b68%\b\b\b69%\b\b\b70%\b\b\b71%\b\b\b72%\b\b\b73%\b\b\b74%\b\b\b75%\b\b\b76%\b\b\b77%\b\b\b78%\b\b\b79%\b\b\b80%\b\b\b81%\b\b\b82%\b\b\b83%\b\b\b84%\b\b\b85%\b\b\b86%\b\b\b87%\b\b\b88%\b\b\b89%\b\b\b90%\b\b\b91%\b\b\b92%\b\b\b93%\b\b\b94%\b\b\b95%\b\b\b96%\b\b\b97%\b\b\b98%\b\b\b99%\b\b\b\b done.\n",
      "Warning: 11950 het. haploid genotypes present (see clean-merged_neighbor.hh );\n",
      "many commands treat these as missing.\n",
      "Warning: Nonmissing nonmale Y chromosome genotype(s) present; many commands\n",
      "treat these as missing.\n",
      "Total genotyping rate is 0.849377.\n",
      "107404 variants removed due to missing genotype data (--geno).\n",
      "547812 variants and 5183 people pass filters and QC.\n",
      "Phenotype data is quantitative.\n",
      "--make-bed to clean-merged_neighbor.bed + clean-merged_neighbor.bim +\n",
      "clean-merged_neighbor.fam ... 0%\b\b1%\b\b2%\b\b3%\b\b4%\b\b5%\b\b6%\b\b7%\b\b8%\b\b9%\b\b10%\b\b\b11%\b\b\b12%\b\b\b13%\b\b\b14%\b\b\b15%\b\b\b16%\b\b\b17%\b\b\b18%\b\b\b19%\b\b\b20%\b\b\b21%\b\b\b22%\b\b\b23%\b\b\b24%\b\b\b25%\b\b\b26%\b\b\b27%\b\b\b28%\b\b\b29%\b\b\b30%\b\b\b31%\b\b\b32%\b\b\b33%\b\b\b34%\b\b\b35%\b\b\b36%\b\b\b37%\b\b\b38%\b\b\b39%\b\b\b40%\b\b\b41%\b\b\b42%\b\b\b43%\b\b\b44%\b\b\b45%\b\b\b46%\b\b\b47%\b\b\b48%\b\b\b49%\b\b\b50%\b\b\b51%\b\b\b52%\b\b\b53%\b\b\b54%\b\b\b55%\b\b\b56%\b\b\b57%\b\b\b58%\b\b\b59%\b\b\b60%\b\b\b61%\b\b\b62%\b\b\b63%\b\b\b64%\b\b\b65%\b\b\b66%\b\b\b67%\b\b\b68%\b\b\b69%\b\b\b70%\b\b\b71%\b\b\b72%\b\b\b73%\b\b\b74%\b\b\b75%\b\b\b76%\b\b\b77%\b\b\b78%\b\b\b79%\b\b\b80%\b\b\b81%\b\b\b82%\b\b\b83%\b\b\b84%\b\b\b85%\b\b\b86%\b\b\b87%\b\b\b88%\b\b\b89%\b\b\b90%\b\b\b91%\b\b\b92%\b\b\b93%\b\b\b94%\b\b\b95%\b\b\b96%\b\b\b97%\b\b\b98%\b\b\b99%\b\b\bdone.\n"
     ]
    }
   ],
   "source": [
    "'''Remove SNPs that are missing in too many samples.'''\n",
    "!plink --bfile $input_filestem --geno 0.01 --make-bed --out $output_filestem"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "PLINK v1.90b4.1 64-bit (30 Mar 2017)           www.cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2017 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to merged_neighbor.log.\n",
      "Options in effect:\n",
      "  --bfile merged_neighbor\n",
      "  --missing\n",
      "  --out merged_neighbor\n",
      "\n",
      "64329 MB RAM detected; reserving 32164 MB for main workspace.\n",
      "655216 variants loaded from .bim file.\n",
      "5183 people (2299 males, 2752 females, 132 ambiguous) loaded from .fam.\n",
      "Ambiguous sex IDs written to merged_neighbor.nosex .\n",
      "5051 phenotype values loaded from .fam.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 5183 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 0%\b\b1%\b\b2%\b\b3%\b\b4%\b\b5%\b\b6%\b\b7%\b\b8%\b\b9%\b\b10%\b\b\b11%\b\b\b12%\b\b\b13%\b\b\b14%\b\b\b15%\b\b\b16%\b\b\b17%\b\b\b18%\b\b\b19%\b\b\b20%\b\b\b21%\b\b\b22%\b\b\b23%\b\b\b24%\b\b\b25%\b\b\b26%\b\b\b27%\b\b\b28%\b\b\b29%\b\b\b30%\b\b\b31%\b\b\b32%\b\b\b33%\b\b\b34%\b\b\b35%\b\b\b36%\b\b\b37%\b\b\b38%\b\b\b39%\b\b\b40%\b\b\b41%\b\b\b42%\b\b\b43%\b\b\b44%\b\b\b45%\b\b\b46%\b\b\b47%\b\b\b48%\b\b\b49%\b\b\b50%\b\b\b51%\b\b\b52%\b\b\b53%\b\b\b54%\b\b\b55%\b\b\b56%\b\b\b57%\b\b\b58%\b\b\b59%\b\b\b60%\b\b\b61%\b\b\b62%\b\b\b63%\b\b\b64%\b\b\b65%\b\b\b66%\b\b\b67%\b\b\b68%\b\b\b69%\b\b\b70%\b\b\b71%\b\b\b72%\b\b\b73%\b\b\b74%\b\b\b75%\b\b\b76%\b\b\b77%\b\b\b78%\b\b\b79%\b\b\b80%\b\b\b81%\b\b\b82%\b\b\b83%\b\b\b84%\b\b\b85%\b\b\b86%\b\b\b87%\b\b\b88%\b\b\b89%\b\b\b90%\b\b\b91%\b\b\b92%\b\b\b93%\b\b\b94%\b\b\b95%\b\b\b96%\b\b\b97%\b\b\b98%\b\b\b99%\b\b\b\b done.\n",
      "Warning: 11950 het. haploid genotypes present (see merged_neighbor.hh ); many\n",
      "commands treat these as missing.\n",
      "Warning: Nonmissing nonmale Y chromosome genotype(s) present; many commands\n",
      "treat these as missing.\n",
      "Total genotyping rate is 0.849377.\n",
      "--missing: Sample missing data report written to merged_neighbor.imiss, and\n",
      "variant-based missing data report written to merged_neighbor.lmiss.\n",
      "PLINK v1.90b4.1 64-bit (30 Mar 2017)           www.cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2017 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to merged_neighbor.log.\n",
      "Options in effect:\n",
      "  --bfile merged_neighbor\n",
      "  --het\n",
      "  --out merged_neighbor\n",
      "\n",
      "64329 MB RAM detected; reserving 32164 MB for main workspace.\n",
      "655216 variants loaded from .bim file.\n",
      "5183 people (2299 males, 2752 females, 132 ambiguous) loaded from .fam.\n",
      "Ambiguous sex IDs written to merged_neighbor.nosex .\n",
      "5051 phenotype values loaded from .fam.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 5183 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 0%\b\b1%\b\b2%\b\b3%\b\b4%\b\b5%\b\b6%\b\b7%\b\b8%\b\b9%\b\b10%\b\b\b11%\b\b\b12%\b\b\b13%\b\b\b14%\b\b\b15%\b\b\b16%\b\b\b17%\b\b\b18%\b\b\b19%\b\b\b20%\b\b\b21%\b\b\b22%\b\b\b23%\b\b\b24%\b\b\b25%\b\b\b26%\b\b\b27%\b\b\b28%\b\b\b29%\b\b\b30%\b\b\b31%\b\b\b32%\b\b\b33%\b\b\b34%\b\b\b35%\b\b\b36%\b\b\b37%\b\b\b38%\b\b\b39%\b\b\b40%\b\b\b41%\b\b\b42%\b\b\b43%\b\b\b44%\b\b\b45%\b\b\b46%\b\b\b47%\b\b\b48%\b\b\b49%\b\b\b50%\b\b\b51%\b\b\b52%\b\b\b53%\b\b\b54%\b\b\b55%\b\b\b56%\b\b\b57%\b\b\b58%\b\b\b59%\b\b\b60%\b\b\b61%\b\b\b62%\b\b\b63%\b\b\b64%\b\b\b65%\b\b\b66%\b\b\b67%\b\b\b68%\b\b\b69%\b\b\b70%\b\b\b71%\b\b\b72%\b\b\b73%\b\b\b74%\b\b\b75%\b\b\b76%\b\b\b77%\b\b\b78%\b\b\b79%\b\b\b80%\b\b\b81%\b\b\b82%\b\b\b83%\b\b\b84%\b\b\b85%\b\b\b86%\b\b\b87%\b\b\b88%\b\b\b89%\b\b\b90%\b\b\b91%\b\b\b92%\b\b\b93%\b\b\b94%\b\b\b95%\b\b\b96%\b\b\b97%\b\b\b98%\b\b\b99%\b\b\b\b done.\n",
      "Warning: 11950 het. haploid genotypes present (see merged_neighbor.hh ); many\n",
      "commands treat these as missing.\n",
      "Warning: Nonmissing nonmale Y chromosome genotype(s) present; many commands\n",
      "treat these as missing.\n",
      "Total genotyping rate is 0.849377.\n",
      "655216 variants and 5183 people pass filters and QC.\n",
      "Phenotype data is quantitative.\n",
      "--het: 543257 variants scanned, report written to merged_neighbor.het .\n"
     ]
    }
   ],
   "source": [
    "'''Remove samples that have too many missing SNPs.\n",
    "Generate missingness and heterozygosity reports using Plink.'''\n",
    "!plink --bfile $input_filestem --missing --out $input_filestem\n",
    "!plink --bfile $input_filestem --het     --out $input_filestem"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>N_MISS</th>\n",
       "      <th>N_GENO</th>\n",
       "      <th>F_MISS</th>\n",
       "      <th>logF_MISS</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>count</th>\n",
       "      <td>5183.000000</td>\n",
       "      <td>5183.000000</td>\n",
       "      <td>5183.000000</td>\n",
       "      <td>5183.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mean</th>\n",
       "      <td>98672.371214</td>\n",
       "      <td>655193.186186</td>\n",
       "      <td>0.150601</td>\n",
       "      <td>-0.822191</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>std</th>\n",
       "      <td>894.183136</td>\n",
       "      <td>20.370968</td>\n",
       "      <td>0.001365</td>\n",
       "      <td>0.003837</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>min</th>\n",
       "      <td>98194.000000</td>\n",
       "      <td>655175.000000</td>\n",
       "      <td>0.149900</td>\n",
       "      <td>-0.824198</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25%</th>\n",
       "      <td>98287.000000</td>\n",
       "      <td>655175.000000</td>\n",
       "      <td>0.150000</td>\n",
       "      <td>-0.823909</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50%</th>\n",
       "      <td>98387.000000</td>\n",
       "      <td>655175.000000</td>\n",
       "      <td>0.150200</td>\n",
       "      <td>-0.823330</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75%</th>\n",
       "      <td>98652.000000</td>\n",
       "      <td>655216.000000</td>\n",
       "      <td>0.150600</td>\n",
       "      <td>-0.822175</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>max</th>\n",
       "      <td>115081.000000</td>\n",
       "      <td>655216.000000</td>\n",
       "      <td>0.175600</td>\n",
       "      <td>-0.755475</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              N_MISS         N_GENO       F_MISS    logF_MISS\n",
       "count    5183.000000    5183.000000  5183.000000  5183.000000\n",
       "mean    98672.371214  655193.186186     0.150601    -0.822191\n",
       "std       894.183136      20.370968     0.001365     0.003837\n",
       "min     98194.000000  655175.000000     0.149900    -0.824198\n",
       "25%     98287.000000  655175.000000     0.150000    -0.823909\n",
       "50%     98387.000000  655175.000000     0.150200    -0.823330\n",
       "75%     98652.000000  655216.000000     0.150600    -0.822175\n",
       "max    115081.000000  655216.000000     0.175600    -0.755475"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "'''Visualize missingness and heterozygosity from reports.'''\n",
    "\n",
    "imiss_file = input_filestem + '.imiss'\n",
    "het_file   = input_filestem + '.het'\n",
    "\n",
    "\n",
    "#!head {imiss_file} #See how Plink generates some extra whitespace...\n",
    "!perl -i.bak -lpe 's/^\\s+//; s/\\s+/\\t/g;' {imiss_file} {het_file} \n",
    "#!head {imiss_file} #Uncomment to see how this Perl one-liner cleaned up these two files for easy reading.\n",
    "#Now the whitespace is cleaned up, and the original het and imiss files are backed up as *.het.bak and *.imiss.bak\n",
    "\n",
    "imiss = pd.read_table(imiss_file, sep=\"\\t\")\n",
    "het   = pd.read_table(het_file,   sep=\"\\t\")\n",
    "\n",
    "het['meanHet'] = (het['N(NM)'] - het['O(HOM)']) / het['N(NM)'] #Compute mean heterozygosity.\n",
    "imiss['logF_MISS'] = np.log10(imiss['F_MISS']) #This might generate a RuntimeWarning: \"divide by zero encountered in log10\" for genotypes that weren't missing in anybody.\n",
    "\n",
    "#imiss.head()\n",
    "\n",
    "imiss.describe()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Summary of log10(missingness): [-0.8242 -0.8239 -0.8233 -0.8222 -0.8222 -0.7555]\n",
      "Summary of mean heterozygosity: [ 0.2349  0.3192  0.3207  0.3207  0.3222  0.3608]\n"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAeAAAAHgCAIAAADytinCAAAgAElEQVR4nO3daVgUV7oH8NN0gzR7\nAyogAs0mJiIq6MQBxeASxy0qPQGdaIwmmLTRSZ5rzJBkEnHLjaI+YzQxwS0X7+gl0SCaoLgFmYgo\nKBoWNYoINgo0ok2DgN3U/VAZhqVYZKmF+v8+zAOnl/NS9vxzKN46JaEoigAAAP+YcF0AAAAwQ0AD\nAPAUAhoAgKcQ0AAAPIWABgDgKQQ0AABPIaABAHgKAQ0AwFMIaAAAnkJAAwDwFAIaAICnENAAADyF\ngAYA4CkENAAATyGgAQB4CgENAMBTCGgAAJ5CQAMA8BQCGgCApxDQAAA8hYAGAOApBDQAAE8hoAEA\neAoBDQDAUwhoAACeQkADAPAUAhoAgKcQ0AAAPIWABgDgKQQ0AABPIaABAHgKAQ0AwFMIaAAAnkJA\nAwDwFAIaAICnENAAADyFgAYA4CkENAAATyGgAQB4CgENAMBTCGgAAJ5CQAMA8BQCGgCApxDQAAA8\nhYAGAOApBDQAAE8hoAEAeAoBDQDAUwhoAACeQkADAPAUAhoAgKcQ0AAAPIWABgDgKQQ0AABPIaAB\nAHgKAQ0AwFMIaAAAnkJAAwDwFAIaAICnWA1oiqJ0Ol1DQwObkwIACBQbAV1TU7NhwwZfX19zc3Nb\nW1szMzMfH5+YmJi6ujoWZgdoJjqaYIkAAsFGQKvV6rNnz8bFxZWUlNTX15eWlu7duzcrK0utVrMw\nO0AzmZkIaBAKCUVRvT2HnZ1dXl6ei4tL08GqqiqlUqnVant7doBmJk8myclEJuO6DoCOsbGCViqV\nycnJLQaPHz/u5ubGwuwAzQweTEzwt3EQBjZW0JmZmSqVSi6X+/v7W1tb6/X63NzcysrKpKSkwMDA\n3p4dAECg2AhoQojBYEhNTS0sLNRqtQqFwsvLKzQ0VIZfMwEA2sZSQDdVXl5uampqZ2fH8rwAhBAS\nHU3Wr8dZDhAENj6mU6dOLS0tJYRoNJrg4GAnJ6cBAwZMmjTp/v37LMwO0Ay6OEA42AjoEydOPHny\nhBDyX//1X56enjqdTq/XBwQEvPPOOyzMDgAgUKyeBb506dJPP/1kaWlJCImOjvb29mZzdgBC0MUB\nQsJSQJeUlLi6uj7//POFhYVDhgwhhOTm5g4cOLCdl1y8ePHs2bMtBisrK8ePHz9t2rRerBX6tj17\nuK4AoLPYCOjx48dHRkaWlZXJ5fKioqKXXnrp3Llzs2fP/vzzz9t5lbOzc+smvNTU1JycHAQ0AIgB\nGwGdmppKCKmvry8qKiovLyeEyOXyo0ePhoSEtPOqwYMHDx48uMXgw4cPcfEhdAu6OEA42PuYmpmZ\neXt7jx07lhAyevRopVJ57Ngx1mYH+B26OEA4OFtHpKenR0ZGcjU7AAD/cRbQKpVKr9dzNTuIF7o4\nQDiwYT+IzJ49CGgQCmzYDwDAU9iwH0QGd1QB4WCjzS4xMbHphv0ODg4hISEBAQFKpXL37t0sFADw\nH3QXB85ygBBgw34AAJ5iYwUdFxenUqliY2Nbb9jPwuwAzaCLA4RDYBv2JyQkaLVanLwGADFgabMk\nmUw2ceJEduYCAOgb8LseiAy6OEA4ENAgMtiLA4QDAQ0AwFMIaBAZdHGAcLB6yysA7uGOKiAcWEoA\nAPAUAhpEBl0cIBwIaBAZdHGAcCCgAQB4CgENIoMuDhAOdHGAyKCLA4QDSwkAAJ5CQIPIoIsDhAMB\nDSKDLg4QDgQ0AABPIaBBZNDFAcKBLg4QGXRxgHBgKQEAwFP8XUEXFxffuHGjxeCvv/5qbW3NST3Q\nR0RHk/XrcZYDBIG/AX3//v2srKwWgwUFBd7e3pzUA30E3cWBgAYh4G9AjxkzZsyYMS0G6bt6c1IP\nAADLsI4AkUEXBwgHf1fQAL0CXRwgHFhKAADwFAIaRAZ7cYBwIKBBZLAXBwgHAhoAgKcQ0CAy6OIA\n4UAXB4gMujhAOLCUAADgKQQ0iAy6OEA4ENAgMujiAOFAQAMA8BQCGkQGXRwgHOjiAJFBFwcIB5YS\nAAA8hYAGkUEXBwgHAhpEBl0cIBwIaAAAnmI1oCmK0ul0DVi/AIfQxQHCwcYntaamZsOGDb6+vubm\n5ra2tmZmZj4+PjExMXV1dSzMDtDMnj0IaBAKNj6parX67NmzcXFxJSUl9fX1paWle/fuzcrKUqvV\nLMwOACBQEoqiensOOzu7vLw8FxeXpoNVVVVKpfJZb9FN39UbyQ5dFx1N1q/HIhoEgY2PqVKpTE5O\nbjF4/PhxNzc3FmYHaAZdHCAcbFxJGBcXp1KpYmNj/f39ra2t9Xp9bm5uZWVlUlISC7MDAAgUGwEd\nFBR069at1NTUwsJCrVarUCiioqJCQ0NlMlxoDqxDFwcIB0sRKZPJJk6c2PjthQsXjEYjAho4gL04\nQDi4WUrMmDGjvLyck6kBAISCjYC2srKSNVdRUeHu7o4VNHAAe3GAcLAR0JcuXRozZszcuXNv3rz5\n4MGDBw8eKBSKK1euPHjwgIXZAZpBFwcIBxtr2KFDh6alpX3xxRfTpk3bsmXLtGnTTExM7O3tHR0d\n23lVTU1N6wQvLS2VSCS9WSwAAF+wdJJBKpW+++67M2fOfOONNw4cOFBfX9/hS86cOfM///M/LQaL\ni4vHjRvXOzWCOKCLA4SDjSsJm2poaNi1a9eZM2d27Njh4ODwrC/HlYQAIB5sLyVMTEyioqIOHjxY\nW1t77NgxlmcHABAQzn7XS09Pj4yM5Gp2EC90cYBwcBbQKpVKr9dzNTuIF7o4QDiwYT8AAE9hw34Q\nGXRxgHBgw34QGdxRBYQDG/YDAPAUNuwHkUEXBwgHNuwHkaG7OHCWA4QAG/YDAPAUNxv2A3AGXRwg\nHFjDgsjgjiogHFhKAADwFAIaRAZdHCAcCGgQGezFAcKBgAYA4CkENIgMujhAONDFASKDLg4QDiwl\nAAB4CgENIoMuDhAOnOIAkRHrXhwZGRmJiYlSqfSVV14ZPnw41+VAp4juYwogQnv27Nm8efOsWbNe\neumlDz/88IcffuC6IugUrKBBZETZxbFr165z587R25MFBQVNnTp1zpw5XBcFHRPdJxXETnx3VHn8\n+PGAAQMaN4+Uy+UymQy3BhUEcX1SAUTI1ta2vLy88RagVVVVDQ0NJiL7r5RA4RQHiEx0NFm/XmyL\n6BUrVowePdrExKShoUGn023bto3riqBTENAgMqLs4qioqBg5cuSQIUPMzMwsLS3j4+NnzZrFdVHQ\nMQQ0QN/3/fffnzx5UiqV0t+ePn1aq9U6OjpyWxV0SFzrCABxdnFQFNWYzoQQe3v7yspKDuuBTsIK\nGkRGlHtxeHl5paWlZWdnHzp0yMLCorCwcOfOnVwXBR3jb0AfOnSo9WeotLR00qRJnNQDIFwbN270\n9fWtr68fMmTItWvXbGxs9u7du2TJEq7rgg5IKIriuoZnkJCQoNVq1Wo114WAYImyiyMrK+tPf/rT\nv/71r5MnT965cyc7O7uoqOjmzZtc1wUd4O8KGqBXiLKL4/jx40+fPvXz8yOEWFlZLV26NC0t7erV\nqwEBAVyXBu0R18cUQJy2b9/+6NEjiqIoitLr9Zs3b7a3t1+9ejXXdUEHENAgMuLr4qAo6sGDBy1G\nnjx5gkYO/sMpDhAZ8XVx3L59u+m39J+d9Hp9UVERRxVBZ4lrKQEgQgaDofVgQ0PD4MGDi4uL2a8H\nOg8BDSIjvjuqtJXCoaGhBQUFLBcDzwQBDSJDd3GISX19PeP4P/7xj4qKCpaLgWcivHPQLi4uubm5\nhBBXV1dbW9vq6urCwkL6IYxgpMMRUl19Ny+Pkkp5Ug8LI48fPyaEKJXKhQsX0iPffffdjRs3AgMD\nCwsLMzMz5XI532oW1oifn1/TK+l7kPAuVKmsrKSvgJJKpRKJhKIoo9FIP4oRjHQ4Qr3+esM33xAT\nE57Uw8JIQkLCX/7yF4lEYmVlRY9UV1ebmJicPHlyxYoVH3zwQUREBN9qFtZI480QepzwVtAtDodE\nImlxdDCCkXZGJPv2mTz7qwQ98ttvvxFCKIqqqqpqfEJDQ8OOHTtsbGzKy8sNBsOjR48cHR35U7Nw\nR3pWm29tNBrLysqcnJwkEknvTd8FAadPE/pz9vLLxM+P3L1LDh78/TGMYAQjrUYeZ2QQQtwJifz9\nGeQIIdcJyTx0aKGZ2YCqqv07duR4eV2tq/vqb3/zy87mQ81CGklIIO+8Q+Ry0gsYAlqj0SxcuDA9\nPV0qlaalpS1btiw+Pt7T07M3pu8C7eDBJDCQEEIcHAghxM7u928xgpHOjGRnk8WLiUTCl3p6f+TS\n3r2EkEeEZP3+DEL/ZbCSoi4ajX5BQcEjRsyfNeuRqeni8PCfYmJ+vxsWz34K/o6MGkXMzEjvYDgH\nPW/ePHt7+9jYWKVSqdFooqOjs7KyTp8+3UsVPBNslgTdNXkySU4mvflrKd/069evrUYOQsjs2bOH\nDh2alpb2ySefHDly5O23337++efZLA/awfAxTUlJuX37tlwuJ4RIpdJVq1a5u7uzXhgA9Iz2GwHO\nnTv31VdfffTRR1OmTPHz8zPrtcUgdAFDH7Srq+svv/zS+O3Fixf5c34DoLvEtxdHQ7t93w8fPly/\nfr2lpaW3t/e1a9e8vb1ZKww6xLCC3rZtW3h4eFhYmE6nU6lUaWlp8fHx7FcG0CvEtxdHh3bt2pWb\nm1tQUHDgwAG+NQWIHENABwUF3bhx4+jRo8OHD3d2dt6+fbu1tTX7lQEAO/r16/fee+/FxcWNHTuW\n61qgmWYBTW+q8txzz92+ffvVV1+lB/V6vbOzs06n46A6gB4nyjuqtK++vv7cuXP4RZmHmgW0ubk5\nIcRoNNJfNFKpVKwWBdB7RHlHlfa99NJLmzZt4roKYMCwgp4yZUpKSkpvTEZfy2RlZWWC/3sA8IO5\nufno0aO5rgKYMQRl63Tu5u8+NTU1GzZs8PX1NTc3t7W1NTMz8/HxiYmJqaur687bAnSF+Lo42lkP\nSSSS4ODgMWPGsFkPdB7DHwmvX7++detWegcsQsiTJ08yMjIWLFjQ5TnUarVGo4mLixs2bJiNjY1O\np8vPz9+4caNard69e3eX3xagK8TXxTFkyJCcnJy2HnV2dp40aRKb9UDnMfyndeHChfX19YMHD66q\nqpoxY8aDBw+6GaOJiYnffvttaGiog4ODqampg4NDSEjI//7v/x45cqQ7bwsAneHv79/WQ5aWlvjb\nIJ8xrKCvXbuWnJxsbm4+e/bsV199dfLkySqVavr06V2eQ6lUJicn03uENjp+/Libm1sX3m3St9+S\nn38mhJD33iNjx5KcHLJmze+PYQQjHY78+c9k2DAikfClnt4fCaivP0DIMEI++f0ZZCsh6YQMNzFJ\n9PMjr7zCeYXCHlm3juzaRf69lWvPYtiLw8PDIz4+fty4cUFBQSdPnjQ3N3d1de3OnRcyMzNVKpVc\nLvf397e2ttbr9bm5uZWVlUlJSYGNO490TkJCQlVx8ZLFiwkhxNqayGSkoYH8+2wMRjDS8UhYGPnu\nOyKT8aWe3h/5+8aN6/77v00Isf39GaSKEAMhfxg9+sKJE3yoUNgjOh2xsyO9gyGgv/nmm+XLl//2\n229xcXE//fSTlZWVubn5icZ/yC4xGAypqamFhYVarVahUHh5eYWGhnZhH1VslgTdJb7NktauXfvJ\nJ5+0Hl+3bt1HH33Efj3Qecx3VCkuLu7fv79MJktISHj48OHChQttbGx6asq6ujqZTNa1O8QkJCQ0\nNDT88Y9/JIQ4OjpaWFjU1taWlZXRj2IEIx2OGF97reLzz4mJCU/qYWGkqKho3LhxAwYMmDhxIj3y\nr3/9q7i4+MmTJzypUOgjrq6uvdQ6zLCOGD58eFxc3ODBgwkh8+fP7/4ceXl5K1euHDBgQGxs7Pz5\n81NTU01MTF5++eXt27c7Ojo+67vV1tbSl56bmpoSQmQyWeOV6BjBSIcjT3butOZTPSyMPHz4kBBS\nU1Nz8+ZNeqSqqsrCwoI/FQp9pPc2MGFYQa9bt+7mzZu7du3qqY0Hg4OD/fz8nJyc9uzZs2DBgg8+\n+KCmpubjjz+ur68/cODAM70VTnEAPKtdu3a9+eabLQbffvvtL7/8kpN6oPMYVtCnTp3Kzs62sbFx\ndXVtPE18/fr1Ls9x+fLlxMREU1PTDRs2rF692sLCwsHBITY21tfXt8vvCdBF4tuLw8LCovUgNnkX\nBIaA3r59e8/OYWNjo9Vq/fz8Dh482PhZuX379qBBg9p51aFDh3bu3NlisLS0FE310C3i24tj5syZ\nrQc9PDxYLwSeGUNADxs2rGfniIqKmjp16vfffx8REUEIKSoq2rJly4EDBz7//PN2XhUeHh4eHt5i\nkD7F0bPlAfRtjPe7+uqrr+j/PwKfsdFstGbNmtDQ0Ma/B9bW1rq4uPz4449BQUEszA7QjPj24qit\nrW09eP/+ffYrgWfFxidVIpFMmjRJqVTS3/r6+q5atcrZ2fnYsWMszA7QzJ49YgvoyspKQohEIjEz\nM5PL5XTLwZMnT7iuCzrG8EldsWLFuXPnjEZjr06cnp4eGRnZq1MAACHExcWFEEJRVH19fWMu49ZW\ngsAQ0AqFYvny5YMGDVKr1WfOnKE3ie5xKpVKr9f3xjsDtCc6mrR7E9W+p0UrLf1tcHBwcXExRxVB\nZzEEdExMzNWrV8+fP+/t7b169WpXV9eoqKiUlJSnT592czKKonQ6Xfv3GAboXXQXh5jQvw1LJBKp\nVCqVSukrLCorK+3t7bkuDTrQ5sk4e3v7wYMHe3l51dfXnz9/fvXq1UqlsmsbhGLDfgAOZWdnE0Io\nipJIJHK5nF4hURRlaWnJdWnQAYaA3rRp04QJE1xdXXft2jVq1KisrKycnJzz58/v37//rbfe6sIc\narX67NmzcXFxJSUl9fX1paWle/fuzcrKwgWBwAHxdXEEBwcTQhwdHaVSaWPL3YsvvshpUdApDG12\nmZmZK1asmDx5cuNV59XV1ZaWlqNHj+7ataGJiYl5eXn0XyoIIfSG/QEBAUqlEndUAbaJ744q9N8D\ntVqthYUF/adC0r1rg4E1zZYSBoPBYDBcuHBh1qxZcrmc/vbRo0fOzs6EEEtLyzlz5nRhDnrD/haD\nXd6wHwCeyZMnT+gGu7q6uqdPn9rb20ulUsbmaOCbZitoc3NzQojRaKS/aKRSqbozR1xcnEqlio2N\nbb1hf3feFqArxLcXh4ODg0KhqK+vf/nllwcOHHjo0CGJRIJLvQWhWUDTHXVTpkxpfWPv7ggKCrp1\n61bTDfujoqK6tmE/QHeJby8OQkhCQsKUKVN++OGHhoYGGxubYcOG9chOwtDbGCKyZ9P592lkssbN\nwgGAZYWFhaampgMHDjQajeXl5TU1Ne3cSRb4o1lABwUFrVmzhvHuOJmZmWyVBNCbxNfFQQhZtWpV\nSkpKTk7OvXv3xowZs2DBAo1G0/52ksAHzQJ6586dSqWy9SafAH2H+Lo4CCG1tbUrV65cvHhxWFjY\niRMnJBLJ+fPn//znP3NdF3Sg5QqaEOLg4HDr1i03N7eGhobDhw/X1tYuWLCAo/IAoAcYDIbIyMgl\nS5YQQsaNG7dlyxYrKyuui4KOMZyDXrNmzYYNG+7du/fVV18dO3astrY2MzMTd8eBPkJ8XRyEkOee\ne+6jjz4qKSlxdnbetm3b0KFDS0tLuS4KOsYQ0P/4xz8uXLjg4ODw5ZdfZmRkGAyGMWPGIKChjxBl\nF0dYWNiiRYuSk5PT09M/+OCDnJycIUOGcF0UdIzhY2o0Gu3s7C5dujRw4EA3N7eamhrGOzIAgFC8\n99578fHxCxYs2LJli1arLS0tHTt2LNdFQccYAjoyMnLq1Knz5s1bsWLFnTt3Fi9ePHnyZPYrA+gV\nouziGDhw4E8//aTRaPbv3+/p6fnPf/6T64qgUyQt9oolhBgMhh9++IEQMnfu3IKCgqSkpKVLl/Lk\nTwr0PQmxyxIAiAFDQNPohnZHR0deXe+HgAYA8WD4Xa+8vDwiIsLc3Nzb21sul0dERJSXl7NfGUCv\nEN8dVUC4GAI6KirKwsKipKREr9eXlJSYm5t3bRtoAD4S3x1VQLgYTl+cPn367t27CoWCENK/f/8t\nW7Z4enqyXhgAgNgxrKCdnJyysrIav71y5Qq9HzRAXyDKLg4QKIYV9GeffaZSqWbPnu3h4VFYWJiY\nmLhHlNsXQN+EDzMIB8NSIjw8/PLlyyNGjKitrR0xYsTly5fnzp3LfmUA0INu3bq1YcOGdevW3b59\nm+taoLPabLPjJ7TZQXeJci+OgwcP/vWvf3V3dzc1Nb1x48bGjRsXL17MdVHQMYZTHK3vhWNlZdW/\nf//p06er1WoLCws26gLoJaLci+Ovf/3rvn37/vSnPxFCLl++PGnSJAS0IDB8TFevXu3m5vbFF18k\nJiZ+8cUX7u7uy5cv/9vf/vbzzz+/99577JcIAN1RV1dnMBjodCaEjBo1yszM7MGDB9xWBZ3BsIL+\n9NNPL1y4QHdujBgxIjAw8MUXX7xx48b48eN9fHxYq+zUqVPfffddi8GCgoLRo0ezVgP0QeLr4ujX\nr5/RaLx69eratWvLy8s9PT1ra2utra25rgs6xnwZt0ajaWyt02g0VVVVhBD6f1kTGBhI92I3dfLk\nSblczmYZ0NeIsotj6NChQUFBo0aNkslkiYmJRqOxrq7O0tKS67qgAwwBHRMTM23atEWLFrm7u9+9\ne3ffvn1r165NTU2dN2/eu+++y1plCoUiMDCwxeDt27e1Wi1rNQD0DVqtNjAwsKSkRCKRPPfccxRF\nHTly5PXXX+e6LugAw+96ixYtOn36tKmp6aVLl6RSaXJy8tKlSwcOHHjo0KFVq1axXyJATxLfXhyP\nHj2iKGrnzp1FRUV379795ZdfqqurscGOIDCsoFesWKFSqdasWSOVShsH/fz8WKwKoNeIr4vD1NTU\nzs7u8OHDI0aMoEfKy8tDQkK4rQo6g+FjqlAoli9fPmjQILVafebMGYPBwH5ZANBTLC0tPTw8rl27\nFhERsX79+pEjR3p4ePzxj3/kui7oGENAx8TEXL169fz5897e3qtXr3Z1dY2KikpJSXn69Cn79QH0\nMPF1cRBCvv76a4VCUVRU9H//939/+MMfzpw5w3VF0CltflLt7e0HDx7s5eVVX19//vz51atXK5XK\nI0eOsFkcQM/bs0eEAe3g4PD1118vW7bslVdemTlzZr9+/biuCDqF4ZO6adOmCRMmuLq67tq1a9So\nUVlZWTk5OefPn9+/fz82hgYQourq6qlTp1ZUVLzwwgsZGRnz588X1h4PosXwR8L8/PwVK1ZMnjy5\nRSv76NGjv/zyS7YKA+gdotyL45tvvlm8ePGrr75KCJk0aVJ0dPSpU6dwM2j+Y/iYzpo1a9KkSU3T\n+fPPPyeEWFpazpkzh73SAHqDKO+okpeX98ILLzR+O3bs2JycHA7rgU5iCOg5c+YEBgZmZmY2jqxd\nu5bFkgCgh3l7e1+7dq3x2+zsbDa3bYAuYwhoW1vbrVu3zpo1a+vWrQ3iW2tAHyfKLo6lS5du3rz5\n4MGDubm527dvz8jIaNw7CfiM+ZM6Y8aMixcv/vDDDzNnzsQVR9CniLKLw87OLjk5+f79+zt37rSw\nsEhMTGx6GRrwFvNmSYQQV1fXM2fOxMTEjBo1qra2ls2aAKDH2djYYLtgwWFYSjS2ashksrVr1377\n7bf0H38B+gLx7cUBwsWwgp4/fz4hxGg0lpeXOzo6hoWFhYWFsV4YQO8Q314cIFwMH9Py8vKIiAhz\nc3Nvb2+5XB4REYHT0AAA7GMI6KioKAsLi5KSEr1eX1JSYm5ujgsIoe8QZRcHCBTDXb1tbGzu3r3b\neDeTiooKT0/Px48fs14bA9zVGwDEg2Ep4eTklJWV1fjtlStXGm9/BQAArGH4I+Fnn32mUqlmz57t\n4eFRWFiYmJi4p4du40ZRVFVVlZWVlQl+xwSuiHIvDhAoho9peHj45cuXR4wYUVtbO2LEiMuXL8+d\nO7c7c9TU1GzYsMHX19fc3NzW1tbMzMzHxycmJqaurq47bwvQFaLciwMEimEFPXz48Li4uB68P6xa\nrdZoNHFxccOGDbOxsdHpdPn5+Rs3blSr1bt37+6pWQAA+hiGgH7llVd27NgxcuRIMzOzHpkjMTEx\nLy/PxcWF/tbBwSEkJCQgIECpVCKggW3o4gDhYAjoU6dOZWdn29jYuLq6ymS/P+H69etdnkOpVCYn\nJy9ZsqTp4PHjx93c3Lr8ngBd1EN/UAFgAUNAb9++vWfniIuLU6lUsbGx/v7+1tbWer0+Nze3srIy\nKSmpZycCAOhLGAJ62LBhhBCj0VhWVubk5CSRSLo5R1BQ0K1bt1JTUwsLC7VarUKhiIqKCg0NbVye\nA7AHXRwgHAwRqdFoFi5cmJ6eLpVK09LSli1bFh8f7+np2a1pZLJRo0aFhYU1xr3RaNRqtY6Ojt15\nW4Bnhr04QDgYPqYrV6708/OrqKiwtLT09/cPDg5+8803uzNHbm7u888/7+Dg4O3tfezYMXqwuLi4\nf//+3XlbAIC+jSGgU1JS1q9fL5fLCSFSqXTVqlUXLlzozhxLly4NDw+vra3du3fvW2+91fRmWgBs\nQxcHCAfDJ9XV1fWXX35p/PbixYvdPL9x5cqVlStXmpmZjR8/fseOHW+99ZbRaOzOGwJ0nSjvqAIC\nxXAOetu2beHh4WFhYTqdTqVSpaWlxcfHd2cOHx+flJQUlUpFCJk1a9a+ffs++eSTDk+bXLt2rfXK\nPTMzc/Dgwd0pBgBAKBh2s6uurq6trT169Oi9e/ecnZ2nT59ubW1taWnZ5TlOnjw5d+7cYcOGHTly\nZMCAAVqtdsqUKRRFZWdnt5690c2bN69evdpiMD093dHR8cMPP+xyMSB26OIA4Wi2gjYYDISQ5557\n7vbt2423udLr9c7OzjqdrstzTJ48+ebNm4PjO7kAABL6SURBVOfOnaPPazs6OqanpycmJl6+fLmd\nV/n6+vr6+rYYpChKq9V2uRIAdHGAgDQLaHNzc0KI0Wikv2hEn53oDmdn54iIiMZv+/XrFxIS0p1V\nOQBAn8ewgp4yZUpKSkpvT5yenr5o0SK9Xt/bEwE0gy4OEA7mNjtCiNFovH//fjvniLtJpVIhnYED\n6OIA4WD4pGo0mokTJ1pbW/v6+l69ejU4OLigoKBHJqMoSqfTNWA3XgCATmDjSkJs2A88Eh2NDftB\nKBj6oFNSUm7fvt30SkJ3d/fuzIEN+4FH0MUBwsEQ0PSVhNOnT6e/7f6VhNiwHwCgCxjWEdu2bXvt\ntddeeeUV+krC119/ffPmzd2Zg96wv8UgNuwHbqCLA4SD4UpCQkhFRUXTKwmdnJy6M0dmZqZKpZLL\n5a037A8MDHymt0pISNBqtWq1ujv1AAAIAnNA9ziDwdB0w34vL6+ubdiPgAYA8WgWkZMmTWrreadO\nnerWNDLZxIkTu/MOAD0De3GAcDQL6MjIyMavly9f/umnn+KOJ9DXoIsDhKNZQL/xxhuNX69cuTIy\nMtLDw4PtigAAgBDC2MUB0JehiwOEA/fVBpHZs4frCgA6q1lA37t3r/HrhoaGpt/iXAcAAMuaBXSL\nu0mNGzeu8Wt2uvEAeh26OEA4mgU0Uhj6PnRxgHDgYwoAwFMIaBAZdHGAcKCLA0QGXRwgHFhKAADw\nFAIaRAZ3VAHhQECDyNBdHABCgIAGAOApBDSIDLo4QDjQxQEigy4OEA4sJQAAeAoBDSKDLg4QDv6e\n4igvLy8qKmoxWFBQYGZmxkk90EdgLw4QDv4GdF5eXnJycovB69evP//885zUAwDAMv4GdGhoaGho\naItB+q7enNQDfQS6OEA4+BvQAL0CXRwgHFhKAADwFAIaRAZdHCAcCGgQGezFAcKBgAYA4CkENIgM\nujhAONDFASKDLg4QDiwlAAB4CgENIoMuDhAOBDSIDLo4QDgQ0AAAPIWABpFBFwcIB7o4QGTQxQHC\ngaUEAABPIaBBZNDFAcKBgAaRQRcHCAcCGgCApxDQIDLo4gDhYPWTSlGUTqdrwC+YwKE9exDQIBRs\nfFJramo2bNjg6+trbm5ua2trZmbm4+MTExNTV1fHwuwAAALFRkCr1eqzZ8/GxcWVlJTU19eXlpbu\n3bs3KytLrVazMDtAM+jiAOFg40KVxMTEvLw8FxcX+lsHB4eQkJCAgAClUrl7924WCgD4D7qLA2c5\nQAjY+Jgqlcrk5OQWg8ePH3dzc2NhdgAAgWJjBR0XF6dSqWJjY/39/a2trfV6fW5ubmVlZVJSEguz\nAzSDLg4QDglFUSxMYzAYUlNTCwsLtVqtQqHw8vIKDQ2VyZ75Pw8JCQlarRYnrwFADFjaLEkmk02c\nOJGduQAA+gb8rgcigy4OEA42VtDXr19v6yE/Pz8WCgD4D3RxgHCwEdCLFi3KyMiwsLBQKBQtHrp3\n7x4LBQAACBEbAZ2enr5kyRILC4vt27d3/lUGg6GqqqrFYHV1dY+WBuKDLg4QDjYCWiKR/OUvf7ly\n5cozvSopKemrr75qMVhaWjplypSeKw3EB3dUAeFgqc2up6DNDgDEg7Pf9TQazbFjx7iaHcQLXRwg\nHJwFdHp6emRkJFezg3jhjiogHJwFtEql0uv1XM0OAMB/2LAfRAZdHCAc2LAfRAZ3VAHhwIb9AAA8\nxUabnZ2dXdMN+2lVVVVKpVKr1T7TW6HNDrorOpqsX49FNAgCNuwHkUEXBwgHNuwHAOApNgI6KCjo\n1q1bTTfsj4qK6tqG/QDdhS4OEA5s2A8ig704QDiwlAAA4CkENIgM9uIA4UBAg8igiwOEAwENAMBT\nCGgQGXRxgHCg0Q1EBl0cIBxYSgAA8BQCGkQGXRwgHAhoEBl0cYBwIKABAHgKAQ0igy4OEA50cYDI\noIsDhANLCQAAnkJAg8igiwOEAwENIoMuDhAOBDQAAE8hoEFk0MUBwoEuDhAZdHGAcGApAQDAU/xd\nQR86dGjnzp0tBktLSydNmsRJPdBHREeT9etxlgMEgb8BHR4eHh4e3mIwISFBq9VyUg/0EXQXBwIa\nhAAfUwAAnkJAg8igiwOEg7+nOAB6Bbo4QDiwlAAA4CkENIgM9uIA4UBAg8hgLw4QDgQ0AABPIaBB\nZNDFAcKBLg4QGXRxgHBgKQEAwFMIaBAZdHGAcCCgQWTQxQHCgYAGAOApBDSIDLo4QDjQxQEigy4O\nEA4sJQAAeAoBDSKDLg4QDgQ0iAy6OEA4ENAAADzFakBTFKXT6RqwfgEOoYsDhIONT2pNTc2GDRt8\nfX3Nzc1tbW3NzMx8fHxiYmLq6upYmB2gmT17ENAgFGx8UtVq9dmzZ+Pi4kpKSurr60tLS/fu3ZuV\nlaVWq1mYHQBAoCQURfX2HHZ2dnl5eS4uLk0Hq6qqlEqlVqt9prdKSEjQarVIdui66Giyfj0W0SAI\nbHxMlUplcnJyi8Hjx4+7ubmxMDtAM+jiAOFg40rCuLg4lUoVGxvr7+9vbW2t1+tzc3MrKyuTkpJY\nmB0AQKDYCOigoKBbt26lpqYWFhZqtVqFQhEVFRUaGiqT4UJzYB26OEA4WIpImUw2ceJEiqKqqqqs\nrKxM8P8Q4Ar24gDhQJsdAABPoc0ORAZ7cYBwsHGKIzExsWmbnYODQ0hISEBAgFKp3L17NwsFAPwH\n3cWBk2wgBGwENN1mt2TJkqaDHbbZJSUl7d+/v8WgRqMZN25cz5cIAMA/bFyokpmZqVKp5HJ56za7\nwMDAtl719OlTvV7fYrChocHa2trMzKyXS4a+a/FismsXVtAgCGwENCHEYDA0bbPz8vJCmx0AQPtY\nCujWNBrNlStXZsyYwcnsAAD8x9kveunp6ZGRkVzNDuKFLg4QDs5W0ADcmDyZJCcTnF4DIcCG/QAA\nPIUrCUFksBcHCAcbpzgWLVqk0Wg+/vjjYcOG2djY6HS6/Pz8jRs39u/fHxeqAAC0RWAb9gMAiAc2\n7AeRQRcHCAd/ryQE6BXo4gDhwJWEIDIIaBCOvtwHHRQUpFAouK6i76uqqqqurnZycuK6kE5ZmZe3\neehQSiLhupCuKCgocHd3l0qlXBfSxxmNRgsLi2PHjnFdCFt3VOGEtbX1yZMnua6i7zt58uTFixc/\n+ugjrgvprJe4LqDLwsPD4+Li7O3tuS6kjysrK1u+fDnXVRDC4aXeAADQPgQ0AABPIaABAHgKAQ0A\nwFMIaAAAnurLAY0+a3ZIpVI0frHDxMTEBDs99T7+HOe+3AddVVVlbW3NdRV9n9ForKurs7Cw4LqQ\nvg8fadbw5FD35YAGABA0XizjAQCgNQQ0AABPIaABAHgKAQ0AwFMIaAAAnkJAAwDwFAIaAICnhBfQ\nmZmZo0aNUigUr7/++pMnTzrzaFsvoSjqhRdeuH79OnvVC1MXjjnB4e2e9o85weHtTfw5tgILaIPB\noFKp1Gp1Tk5OcXHx1q1bO3y0rZecPn16yZIlGRkZHPwYgtKFY05weLun/WNOcHh7E6+OrcACOjU1\nVS6Xv/HGG4MGDfr444/37dvX4aNtveTKlSsWFha4QLlDXTjmBIe3e9o/5gSHtzfx6tgKbDuhwsJC\nf39/+mt/f/+7d+9SFCX59/3lGB9t6yUrV64khCQmJrL+QwhMF445Dm83tX/MCSE4vL2HV8dWYCto\nrVbbuIOJjY1NfX19VVVV+4+2/xLoUBeOOQdV9i04qkATWEArFAq9Xk9/rdPpZDKZlZVV+4+2/xLo\nUBeOOQdV9i04qkATWEB7enrm5ubSX+fn53t4eDTdtpXx0fZfAh3qwjHnoMq+BUcVaAL7V58wYUJl\nZeX333+v1+s3bdr06quv0uPff/+9RqNhfLStl0AndeGYQze1f8y5rQ1YRQnNpUuXAgIC7O3tFy1a\nVFtbSw9aWloePXq0rUcZB2mDBg3Kz89n+UcQnC4ccxoOb5e1f8xpOLy9hyfHFhv2AwDwlMBOcQAA\niAcCGgCApxDQAAA8hYAGAOApBDQAAE8hoAEAeAoBDQDAUwhoAACeQkADAPAUAhoAgKcQ0AAAPIWA\nBgDgKQQ0AABPIaABAHgKAQ0AwFMIaAAAnkJAAwDwFAIaAICnENAAADyFgBYpJycnyb95eHjs3Lmz\nV6eTyWQGgyEzMzMoKKhn3/n9999XKBRlZWWdefIzFdAb1XYTfRi5rgLYg5vGipSTk9OBAwdGjhxZ\nU1Nz7Nixd955Jysry9/fv5emk8lktbW1jx8/zsjImDZtWg++s0KhyM/Pd3Jy6syTKyoqOl/AMz2Z\nHfRhlMlkXBcCLMEKWrysra3t7OxcXFyioqKGDh2anZ2dk5MzYcKEdevWDR8+nBBy+PDhIUOG2Nra\nhoeHl5eXE0L279//5ptvLly40M7OLjg4+MaNG/RbtX5m07eaMmWK0Wj08vLKycn55JNP2nrJ9evX\nQ0JCYmNjBw0apFQqz5w506Lg1i+ZM2fO48ePx4wZQ39Lv0lwcPDKlSsdHR1DQkLS09NHjx5tbW39\n7rvvEkLu3LlDF2AwGN5++22FQuHo6Lh27VrGkcYnt1XYvn37PDw8PDw8vv32Ww8PjxbVMj6alpY2\ncuRIBweH+fPnl5SUtPPmrX/YxsMYERGxadMm+mmffvrpihUr2vp3aT1d6x8TeI0CURo4cOClS5co\ninr69OmJEyfMzc1/+eWXX3/91dbWdsGCBRcuXCgoKLC1tU1JSamoqFi0aFFERARFUfHx8TKZ7Msv\nvywrK1u1alVAQEBDQwPjM5u+FUVRUqn06dOnly5dCgwMpCiK8SX5+fmWlpafffZZdXX1qlWrxo4d\n27RgxpdQFGVra1tVVdX4tPz8fIlEsnv37oqKisDAwAEDBhQWFqanpxNCysrKGgtISEgYMmTInTt3\nLl++3K9fv1u3brUeaXwyY2HXrl1zdHTMyMjQaDQhISHu7u5Nq2V8VKvVOjg4HD169OHDh2+//XZY\nWFhbb97WD0sfxkOHDk2YMIEeGT58eGpqKuO/C+N0rX/MnvxUQU9DQIvUwIEDLS0tbW1t5XK5RCJ5\n//33KYr69ddf+/XrV1dXR1HU1q1bX3vtNfrJZWVlpqamBoMhPj5++PDh9GB9fb29vf3NmzcZn9n0\nrahWAc34kvz8fBsbm6dPn9KVDBkypGnBjC+hmALa3t6efpMPPvhg6dKl9Li7u/tvv/3WNKA9PT0z\nMjIaGhrKy8vr6upajzQN6NaFffjhh/RBoyjqyJEjLQKa8dF9+/apVCp6sLq6Wi6Xt/VTt/XD0odR\np9NZWlo+evTo9u3bAwcObOvfhXG61j9m+58T4BZOZonXP//5T/pUhqOjo5WVFT3o6upqZmZGCHnw\n4EHjL+b9+/c3MzOjf9FWKpX0oKmpqYeHh0ajaeuZjW/VWlsvcXJyok+wtj7NyvgSxlPP/fv3b3yT\nxie0eMO5c+fqdLqoqKjS0tJly5atXLmy9UjT57cuTKPRjB49mv7azc2tRQ2MjxYXF//444+NJUml\nUvpvm63fvP0f1traOjg4+OTJk0VFRXPnzpVKpYTp34VxuvZ/TOAbnIMWLxcXF/okaWM6kyYZ4eTk\ndPfuXfpreqnl6OhICLlz5w49aDAYioqKnJ2d23pmO3/LauslEonkWV/SNQUFBWFhYdnZ2RkZGUeP\nHj18+HDrkabPb12Ys7NzUVER/XVxcXFnHnV2dp42bdqDBw8ePHig0WjS0tLo9Gz95h3+sDNmzPjx\nxx8TExNVKhU90vrfhXG69n9M4BsENDCbOXPm4cOHT58+XVlZuXLlyjlz5tCBe+3ata+//lqr1f79\n7393cXHx8fFp65kt6PX6Dt+8C/V0TVJSUkRERG5urtForKure/LkSeuR9t9BpVLt3bs3MzPz/v37\nmzdv7syj06dPT0tLS05Orqio+PDDD5ctW9bWf5Da+WHpwzhjxowjR47cvHlz/Pjx9HjrfxfG6Z71\nxwRuIaCBmZeX1549e9Rqtbu7u06n2759Oz0+bdq0U6dOeXp6/vzzzwcPHjQxMWnrmU2Fh4e7ublV\nV1e3/+ZdqKdrli5dOmjQoD/84Q9BQUFjx45duHBh65H23yEwMHDt2rWzZs2aMGHCvHnzmv4W0taj\nTk5O8fHx77//vpub26VLl+Lj45/1h208jEql0sXFZdasWY3B3frfhXG6Z/0xgWNcnwQHIYmPj2/s\nKBC5/Pz8n3/+mf76xIkTjW0VnXm0R4wdO/bEiRP01/h36auwggboisrKynnz5pWVlT158uSLL75o\ncT1L+492U3V19c8//1xUVPTiiy/24NsCDyGgAbpi7NixK1asGDlypI+Pj7Oz87Jlyzr/aDelpKTM\nmzdvx44dpqamPfi2wEO41BsAgKewggYA4CkENAAATyGgAQB4CgENAMBTCGgAAJ5CQAMA8BQCGgCA\npxDQAAA8hYAGAOApBDQAAE8hoAEAeAoBDQDAUwhoAACeQkADAPAUAhoAgKf+H3mEn2ptwLYgAAAA\nAElFTkSuQmCC\n"
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "'''Plot missingness versus heterozygosity. Use Rmagic.'''\n",
    "\n",
    "%Rpush imiss_file het_file\n",
    "%R imiss <- read.table(imiss_file, header=T)\n",
    "%R het   <- read.table(het_file,   header=T)\n",
    "\n",
    "#Calculate the log10 of the F_MISS and the mean heterozygosity (among non-missing genotypes).\n",
    "%R imiss$logF_MISS <- log10(imiss$F_MISS)\n",
    "%R het$meanHet     <-  (het$N.NM. - het$O.HOM.)/het$N.NM.\n",
    "\n",
    "%R -o summary_logF_MISS summary_logF_MISS <- summary(imiss$logF_MISS)\n",
    "print(\"Summary of log10(missingness): \" + str(summary_logF_MISS))\n",
    "\n",
    "%R -o summary_meanHet summary_meanHet <- summary(het$meanHet)\n",
    "print(\"Summary of mean heterozygosity: \" + str(summary_meanHet))\n",
    "\n",
    "#Plot the logF_MISS against the mean heterozygosity.\n",
    "#You can modify this plotting command to focus on your range of interest.\n",
    "%R plot(imiss$logF_MISS, het$meanHet, xlim=c(-3,0), ylim=c(0,0.5), pch=1, xlab=\"Proportion of missing genotypes\", ylab=\"Heterozygosity rate\",axes=F); axis(2,at=c(0,0.05,0.10,0.15,0.2,0.25,0.3,0.35,0.4,0.45,0.5),tick=T); axis(1,at=c(-3,-2,-1,0),labels=c(0.001,0.01,0.1,1)); abline(h=mean(het$meanHet) - (2 * sd(het$meanHet)), col=\"RED\",  lty=2); abline(h=mean(het$meanHet) - (3 * sd(het$meanHet)), col=\"gray\", lty=2); abline(h=mean(het$meanHet) + (2 * sd(het$meanHet)), col=\"red\",  lty=2); abline(h=mean(het$meanHet) + (3 * sd(het$meanHet)), col=\"gray\",  lty=2); abline(v=-1.522879, col=\"red\", lty=2)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "'''Write file containing FID:IID pairs for individuals with outlying heterozygosity'''\n",
    "%R het.filt <- het[which( het$meanHet > (mean(het$meanHet) + 3 * sd(het$meanHet)) | het$meanHet < (mean(het$meanHet) - 3 * sd(het$meanHet)) ),]\n",
    "%R write.table(het.filt[,c(0,1)], file=\"fail-het-qc.txt\", sep=\"\\t\", row.names=F, quote=F)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of individuals who failed heterozygosity QC:['103']\n"
     ]
    }
   ],
   "source": [
    "fail_het_qc_count = !cat fail-het-qc.txt | wc -l\n",
    "\n",
    "print(\"Number of individuals who failed heterozygosity QC:\" + str(fail_het_qc_count))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "'''Exclude genetic variants that are not present in at least 99% of samples.'''\n",
    "!plink --bfile $input_filestem --geno 0.01 --make-bed --out $output_filestem"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 2B: Missing Genotypes\n",
    "The purpose of this step is to identify samples that have too many missing genotypes.  This may indicate a technical failure for the sample because near-complete genotyping is expected.  If this is true, then few or no samples should fail this step.  Any samples that do are excluded from further analysis.\n",
    "Critical note: if a subset of genotypes are missing in all or nearly all samples, this step, which excludes samples that have too many missing genotypes, could exclude all or nearly all samples.  In this case, the variants in the study can be pruned first (see 3A: Callrate Control) to remove variants that are not present in at least 97% of samples, say."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "PLINK v1.90b3.45 64-bit (13 Jan 2017)      https://www.cog-genomics.org/plink2\n",
      "(C) 2005-2017 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to merged_neighbor.log.\n",
      "Options in effect:\n",
      "  --bfile merged_neighbor\n",
      "  --missing\n",
      "  --out merged_neighbor\n",
      "\n",
      "995 MB RAM detected; reserving 497 MB for main workspace.\n",
      "655216 variants loaded from .bim file.\n",
      "5183 people (2299 males, 2752 females, 132 ambiguous) loaded from .fam.\n",
      "Ambiguous sex IDs written to merged_neighbor.nosex .\n",
      "5051 phenotype values loaded from .fam.\n",
      "Using 1 thread.\n",
      "Before main variant filters, 5183 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 0%\b\b1%\b\b2%\b\b3%\b\b4%\b\b5%\b\b6%\b\b7%\b\b8%\b\b9%\b\b10%\b\b\b11%\b\b\b12%\b\b\b13%\b\b\b14%\b\b\b15%\b\b\b16%\b\b\b17%\b\b\b18%\b\b\b19%\b\b\b20%\b\b\b21%\b\b\b22%\b\b\b23%\b\b\b24%\b\b\b25%\b\b\b26%\b\b\b27%\b\b\b28%\b\b\b29%\b\b\b30%\b\b\b31%\b\b\b32%\b\b\b33%\b\b\b34%\b\b\b35%\b\b\b36%\b\b\b37%\b\b\b38%\b\b\b39%\b\b\b40%\b\b\b41%\b\b\b42%\b\b\b43%\b\b\b44%\b\b\b45%\b\b\b46%\b\b\b47%\b\b\b48%\b\b\b49%\b\b\b50%\b\b\b51%\b\b\b52%\b\b\b53%\b\b\b54%\b\b\b55%\b\b\b56%\b\b\b57%\b\b\b58%\b\b\b59%\b\b\b60%\b\b\b61%\b\b\b62%\b\b\b63%\b\b\b64%\b\b\b65%\b\b\b66%\b\b\b67%\b\b\b68%\b\b\b69%\b\b\b70%\b\b\b71%\b\b\b72%\b\b\b73%\b\b\b74%\b\b\b75%\b\b\b76%\b\b\b77%\b\b\b78%\b\b\b79%\b\b\b80%\b\b\b81%\b\b\b82%\b\b\b83%\b\b\b84%\b\b\b85%\b\b\b86%\b\b\b87%\b\b\b88%\b\b\b89%\b\b\b90%\b\b\b91%\b\b\b92%\b\b\b93%\b\b\b94%\b\b\b95%\b\b\b96%\b\b\b97%\b\b\b98%\b\b\b99%\b\b\b\b done.\n",
      "Warning: 11950 het. haploid genotypes present (see merged_neighbor.hh ); many\n",
      "commands treat these as missing.\n",
      "Warning: Nonmissing nonmale Y chromosome genotype(s) present; many commands\n",
      "treat these as missing.\n",
      "Total genotyping rate is 0.849377.\n",
      "--missing: Sample missing data report written to merged_neighbor.imiss, and\n",
      "variant-based missing data report written to merged_neighbor.lmiss.\n",
      "PLINK v1.90b3.45 64-bit (13 Jan 2017)      https://www.cog-genomics.org/plink2\n",
      "(C) 2005-2017 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to merged_neighbor.log.\n",
      "Options in effect:\n",
      "  --bfile merged_neighbor\n",
      "  --het\n",
      "  --out merged_neighbor\n",
      "\n",
      "995 MB RAM detected; reserving 497 MB for main workspace.\n",
      "655216 variants loaded from .bim file.\n",
      "5183 people (2299 males, 2752 females, 132 ambiguous) loaded from .fam.\n",
      "Ambiguous sex IDs written to merged_neighbor.nosex .\n",
      "5051 phenotype values loaded from .fam.\n",
      "Using 1 thread.\n",
      "Before main variant filters, 5183 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 0%\b\b1%\b\b2%\b\b3%\b\b4%\b\b5%\b\b6%\b\b7%\b\b8%\b\b9%\b\b10%\b\b\b11%\b\b\b12%\b\b\b13%\b\b\b14%\b\b\b15%\b\b\b16%\b\b\b17%\b\b\b18%\b\b\b19%\b\b\b20%\b\b\b21%\b\b\b22%\b\b\b23%\b\b\b24%\b\b\b25%\b\b\b26%\b\b\b27%\b\b\b28%\b\b\b29%\b\b\b30%\b\b\b31%\b\b\b32%\b\b\b33%\b\b\b34%\b\b\b35%\b\b\b36%\b\b\b37%\b\b\b38%\b\b\b39%\b\b\b40%\b\b\b41%\b\b\b42%\b\b\b43%\b\b\b44%\b\b\b45%\b\b\b46%\b\b\b47%\b\b\b48%\b\b\b49%\b\b\b50%\b\b\b51%\b\b\b52%\b\b\b53%\b\b\b54%\b\b\b55%\b\b\b56%\b\b\b57%\b\b\b58%\b\b\b59%\b\b\b60%\b\b\b61%\b\b\b62%\b\b\b63%\b\b\b64%\b\b\b65%\b\b\b66%\b\b\b67%\b\b\b68%\b\b\b69%\b\b\b70%\b\b\b71%\b\b\b72%\b\b\b73%\b\b\b74%\b\b\b75%\b\b\b76%\b\b\b77%\b\b\b78%\b\b\b79%\b\b\b80%\b\b\b81%\b\b\b82%\b\b\b83%\b\b\b84%\b\b\b85%\b\b\b86%\b\b\b87%\b\b\b88%\b\b\b89%\b\b\b90%\b\b\b91%\b\b\b92%\b\b\b93%\b\b\b94%\b\b\b95%\b\b\b96%\b\b\b97%\b\b\b98%\b\b\b99%\b\b\b\b done.\n",
      "Warning: 11950 het. haploid genotypes present (see merged_neighbor.hh ); many\n",
      "commands treat these as missing.\n",
      "Warning: Nonmissing nonmale Y chromosome genotype(s) present; many commands\n",
      "treat these as missing.\n",
      "Total genotyping rate is 0.849377.\n",
      "655216 variants and 5183 people pass filters and QC.\n",
      "Phenotype data is quantitative.\n",
      "--het: 543257 variants scanned, report written to merged_neighbor.het .\n"
     ]
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>N_MISS</th>\n",
       "      <th>N_GENO</th>\n",
       "      <th>F_MISS</th>\n",
       "      <th>logF_MISS</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>count</th>\n",
       "      <td>5183.000000</td>\n",
       "      <td>5183.000000</td>\n",
       "      <td>5183.000000</td>\n",
       "      <td>5183.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mean</th>\n",
       "      <td>98672.371214</td>\n",
       "      <td>655193.186186</td>\n",
       "      <td>0.150601</td>\n",
       "      <td>-0.822191</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>std</th>\n",
       "      <td>894.183136</td>\n",
       "      <td>20.370968</td>\n",
       "      <td>0.001365</td>\n",
       "      <td>0.003837</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>min</th>\n",
       "      <td>98194.000000</td>\n",
       "      <td>655175.000000</td>\n",
       "      <td>0.149900</td>\n",
       "      <td>-0.824198</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25%</th>\n",
       "      <td>98287.000000</td>\n",
       "      <td>655175.000000</td>\n",
       "      <td>0.150000</td>\n",
       "      <td>-0.823909</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50%</th>\n",
       "      <td>98387.000000</td>\n",
       "      <td>655175.000000</td>\n",
       "      <td>0.150200</td>\n",
       "      <td>-0.823330</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75%</th>\n",
       "      <td>98652.000000</td>\n",
       "      <td>655216.000000</td>\n",
       "      <td>0.150600</td>\n",
       "      <td>-0.822175</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>max</th>\n",
       "      <td>115081.000000</td>\n",
       "      <td>655216.000000</td>\n",
       "      <td>0.175600</td>\n",
       "      <td>-0.755475</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              N_MISS         N_GENO       F_MISS    logF_MISS\n",
       "count    5183.000000    5183.000000  5183.000000  5183.000000\n",
       "mean    98672.371214  655193.186186     0.150601    -0.822191\n",
       "std       894.183136      20.370968     0.001365     0.003837\n",
       "min     98194.000000  655175.000000     0.149900    -0.824198\n",
       "25%     98287.000000  655175.000000     0.150000    -0.823909\n",
       "50%     98387.000000  655175.000000     0.150200    -0.823330\n",
       "75%     98652.000000  655216.000000     0.150600    -0.822175\n",
       "max    115081.000000  655216.000000     0.175600    -0.755475"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "collapsed": false,
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Summary of log10(missingness): [-0.8242 -0.8239 -0.8233 -0.8222 -0.8222 -0.7555]\n",
      "Summary of mean heterozygosity: [ 0.2349  0.3192  0.3207  0.3207  0.3222  0.3608]\n"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAeAAAAHgCAIAAADytinCAAAgAElEQVR4nO3deUBU5d4H8GeGRYEZ\nNpVN1PEqOjoWKooXE+mW2ShipA6iVBRSmTf1Wr2XTG8aiZWVWykU89p1eV2mXPJ6EaOyHNI06gqC\nkQshm2wGMoDs5/3jdCcYDuswZ+F8P3/NPJw5z4/j6dfDM7/zPBKKoggAAPCPlOsAAACAGRI0AABP\nIUEDAPAUEjQAAE8hQQMA8BQSNAAATyFBAwDwFBI0AABPIUEDAPAUEjQAAE8hQQMA8BQSNAAATyFB\nAwDwFBI0AABPIUEDAPAUEjQAAE8hQQMA8BQSNAAATyFBAwDwFBI0AABPIUEDAPAUEjQAAE8hQQMA\n8BQSNAAATyFBAwDwFBI0AABPIUEDAPAUEjQAAE8hQQMA8BQSNAAATyFBAwDwFBI0AABPIUEDAPAU\nEjQAAE8hQQMA8BQSNAAATyFBAwDwFBI0AABPIUEDAPAUEjQAAE8hQQMA8BQSNAAATyFBAwDwFBI0\nAABPIUEDAPAUEjQAAE8hQQMA8BQSNAAATyFBAwDwFBI0AABPIUEDAPAUEjQAAE8hQQMA8BQSNAAA\nTyFBAwDwFBI0AABPIUEDAPAUEjQAAE9Zs9ONXq/fu3dvVlaWwWCQy+UqlSoyMjIwMJCd3gEAhIiN\nEXRCQkJwcDAhJCIiIiYmJiIiQiKRBAcHJyQksNA7AIBASSiKsnQfXl5eOp1uxowZrRsvXLig0WgK\nCgos3TtAG2vXkrg4IsXkHggAG7dpZWWlUqk0aVQqlRUVFSz0DtBGWhppaeE6CIBuYSNBh4SEhIeH\n6/X6qqoqiqKqqqpSU1M1Gk1ISAgLvQMACBQbCVqr1SoUCrVa7eTkJJVKnZyc1Gr1yJEjExMTWegd\noI1hwzC/AULBxhw0rbGxMTc3t7q6WiaTKRQKGxsbdvoFABAo9hI0AAD0CEt10EZXrlw5duxYc3Pz\nvHnz/P39We4dAFUcICBs3KbOzs719fWEkLNnz06dOvXSpUtXrlwJDAw8efIkC70DtIEqDhAONqY4\nJBLJvXv3Bg4c+PDDD8+ePTsmJoYQotVqExIS0tLSLN07QBuPPEJOnybWbP/tCNALrP6hd+3atdDQ\nUPp1aGhodnY2m70DEIIqDhASlsYR+/fvt7GxkUgkdXV1dMu9e/esOx3FXLp06ezZsyaNFRUVM2fO\nnDt3rqUChX5vzx6uIwDoLjYSdHBw8Oeff04Iuf/++4uKinx9fQkh33zzDf2iI56enn5+fiaN3377\nbWZmJhI0AIgBZ2V2NTU1FEXJZLIefUqn05WXl69YscJCUUH/hyoOEA7OvipxcHDgqmsQNbqKAwka\nhICz2zQ5OXn9+vVc9Q4AwH+cJeiCggLU2AEHUMUBwiGwR70xBw0A4oGhBAAAT7GUoPV6fXR0dEBA\nwIQJEwICAqKjo/V6PTtdA7Sxdi0e9QahwJ6EIDJYiwOEg40yu9jY2KSkJJM9CaOiojQazfLly1kI\nAABAiLAnIYgMqjhAOLAnIYjMnj1I0CAU2JMQAICn2EjQcrlcq9VWVlZeu3btp59+unbtWkVFRWJi\nolwuZ6F3gDZQxQHCwd5aHDY2Nj4+Pqx1B8AMa3GAcOA2BQDgKSRoEBlUcYBwYGc2EBnsqALCgaEE\nAABPIUGDyKCKA4QDCRpEBmtxgHAgQQMA8BQSNIgMqjhAOFDFASKDKg4QDgwlAAB4CgkaRAZVHCAc\nSNAgMqjiAOFAggYA4CkkaBAZVHGAcKCKA0QGVRwgHBhKAADwFH9H0Pn5+b/88otJ45UrV7APC5hl\n7VoSF4dZDhAE/ibo27dv//jjjyaNOTk5o0eP5iQe6CewowoIB38TtL+/v7+/v0mjTqcrLy/nJB4A\nAJZhHAEigyoOEA7+jqABLAJVHCAcGEoAAPAUEjSIDNbiAOFAggaRwVocIBxI0AAAPIUEDSKDKg4Q\nDlRxgMigigOEA0MJAACeQoIGkUEVBwgHEjSIDKo4QDiQoAEAeAoJGkQGVRwgHKjiAJFBFQcIB4YS\nAAA8hQQNIoMqDhAOJGgQGVRxgHCwNAet1+v37t2blZVlMBjkcrlKpYqMjAwMDGSndwAAIWJjBJ2Q\nkBAcHEwIiYiIiImJiYiIkEgkwcHBCQkJLPQO0AaqOEA4JBRFWboPLy8vnU43Y8aM1o0XLlzQaDQF\nBQU9OhW9J+GKFSv6NEAAAD5iYyhRWVmpVCpNGpVKZUVFBQu9AwAIFBsJOiQkJDw8XK/XV1VVURRV\nVVWVmpqq0WhCQkJY6B2gDVRxgHCwkaC1Wq1CoVCr1U5OTlKp1MnJSa1Wjxw5MjExkYXeAdpAFQcI\nBxtVHHK5XKvVxsfH5+bmVldXy2QyhUJhY2PDQtcAAMLF3qPeNjY2Pj4+rHUHwAxVHCAc3NypCoWi\npKSEk65B7PbsQYIGoWBjBK1Wq01abt++vWTJEltb2+TkZBYCAAAQIjYS9Llz5/z8/OhnVWjffffd\n9OnTZTIZC70DtLF2LYmLwyAaBIGNBH316tWVK1eeP39+x44dI0eOJIRs3779xRdf9PDwYKF3gDbo\nKg4kaBACNhK0QqH417/+deLECbVavXTp0piYmO58qra2tri42KSxpKREIpFYIEYAAN5hr4ojNDT0\nkUceeeONNyZOnGgwGLo8/uuvv963b59JY35+PpZYArOgigOEg421OExcvXr13LlzkZGRdnZ2Pf0s\n1uIAAPHgYMur8ePHjx8/nv1+AQCEhbO/9ZKTk9evX89V7yBeWIsDhIOzBF1QUJCWlsZV7yBeWIsD\nhIOzBB0dHY2nVAAAOoGvs0FkUMUBwoE9CUFk9uzhOgKA7sKehAAAPMXGCDo2NjYpKclkT8KoqCiN\nRrN8+XIWAgD4A9biAOHAnoQgMqjiAOHAnoQAADyFPQlBZFDFAcLB3locjY2N5u9JiLU4AEA8sCch\nAABP4W89EBmsxQHCgQQNIoMqDhAOJGgAAJ5CggaRQRUHCAcHC/YDcAlrcYBwYCgBAMBTSNAgMqji\nAOHAFAeIDF3FIb5p6IsXL544ccLKyiosLOz+++/nOhzoFtHdpgAitGfPnvfff3/+/PmPPvroa6+9\ndvz4ca4jgm7BCBpERpRVHFqt9ty5c9bW1oSQKVOmqNXqxx9/nOugoGuiu1NB7PbsEVuCvnv3rpub\nG52dCSF2dnbW1tYtmIgXAnHdqQAi5OTkVFZWVl9fT781GAwtLS1Skf1fSqAwxQEiI8odVVatWjV1\n6lSpVNrS0lJVVbVz506uI4JuQYIGkRFlFcedO3cmTZo0duxYW1tbBweH/fv3z58/n+ugoGtI0AD9\n32effZaSkmJlZUW//eqrr8rLywcPHsxtVNAlcY0jAMRZxUFRlDE7E0JcXV2xI6ggYAQNIiPKtThG\njRql1+svX7589OhRe3v73NzchIQEroOCrvE3QR89erT9PVRSUjJr1ixO4gEQri1btowZM6ahoWHs\n2LEZGRmOjo6ffPLJsmXLuI4LusDenoR9AnsSgrlEWcXx448/zpkzJzU1NSUl5ddff718+XJeXt61\na9e4jgu6wN8RNIBFiLKKIzk5ubGxUalUEkJkMtnzzz+v1+vT09N9fX25Dg06I67bFECcPvzww8rK\nSoqiKIqqrq5+//33XV1dN27cyHVc0AUkaBAZ8VVxUBRVXFxs0nLv3j0UcvAfpjhAZMRXxXHz5s3W\nb+mvnaqrq/Py8jiKCLqLeSjx73//++mnn16wYAEh5NKlS1988QW7UQFAn2lqamrf2NLSMmzYsPz8\nfPbjge5jSNDbtm176aWX/Pz8Ll++TAhxcHB4/fXXWQ8MwDLEt6NKR1k4KCgoJyeH5WCgRxgS9Pbt\n21NSUlauXGlra0sIUSqVv/zyC+uBAVgGXcUhJg0NDYztO3bsuHPnDsvBQI8wzEHX1ta6ubm1fmtn\nZ8diSF3w8vLKysoihHh7ezs5OdXU1OTm5tI/QgtaumwhNTW3rl6lrKx4Eg8LLXfv3iWEjBw58qmn\nnqJbPv30019++cXPzy83NzctLc3Ozo5vMQurRalUtn6Svg8xPKiyaNGiUaNGbd68WaVS/fzzz6+8\n8kpZWdm+ffss0X1P6XS6iooK+gkoKysriURCUVRzczP9U7SgpcsW6plnWj7+mEilPImHhRadThcR\nESGRSGQyGd1SU1MjlUpTUlJWrVoVExOzePFivsUsrBbjZgh9juG8H3zwwYIFC1xdXevr6z09PUeN\nGnXixAkLdd8LJpdDIpGYXB20oKWTFsk//ynt+acE3XL9+nVCCEVRBoPBeEBLS8uuXbscHR3Lysqa\nmpoqKysHDx7Mn5iF29K3GE7t6el5/vz5jIyMW7duDR8+3NfXVyKRWC6CnvL96itC32ePPUaUSnLr\nFjl8+PefoQUtaGnXcvfiRULICELCfz+CfE5INiFpR48+ZWvrZjAc2LUrc9So9Pr6+FdfVV6+zIeY\nhdSi05EXXySWmQdmSND79u176qmnfH19jY+BJiYmPvvss5bovhfKhw0jfn6EEDJoECGEODv//hYt\naOlOy+XLJCqKSCR8icfyLT988gkhpJKQH38/gtDfDFZQ1KXmZuWUKQ9MnLh0/vxKG5uohQuT3njj\n992wePZb8Ldl8mRia0ssg2EOWqFQGCfCaS4uLjx56AiLJYG5HnmEnD5NLPlnKd8MGDCgo0IOQkho\naOi4ceP0ev3rr7/++eefv/DCCyqVis3woBNtbtO0tDRCSENDA/2CduPGDblcznZcANBHOl+x8ty5\nc/Hx8evWrZs9e7ZSqbS12GAQeqFNgg4NDSWElJaW0i8IIVKp1M3Nbfv27RyEBmAJ4luLo6XTuu/f\nfvstLi7ugw8+GD16dEZGxujRo1kLDLrUJkEXFBQQQlasWLF7926O4gGwMPGtxdElrVablZWVk5Nz\n6NAhXlUEAMNM3O7du+/du3ft2rXy8nLjH0fYxwSgvxowYMCaNWsSExMDAgK4jgXaYEjQ3377bVhY\nmJ2dXVFRkbe396+//qpSqTIzM9kPDqDviXJHlc41NDScO3du//79XAcCphhu07/97W+bNm3Kzc31\n8vLKyck5fvz4X/7yFzO70ev10dHRAQEBEyZMCAgIiI6O1uv1Zp4ToDfEtxZHlx599NF3333XycmJ\n60DAFEOCvnHjRlRUFCGEfqgxNDT0srF2vVcSEhKCg4MJIRERETExMfRTp8HBwdhXGIBzAwcOnDp1\nKtdRADOGKQ4nJ6c7d+64ubm5u7tnZmaOGjWqtLTUnD5iY2OTkpJmzJjRujEqKkqj0SxfvtycMwP0\nmPiqOKRSqXEFCRMSieSBBx7w9/dnOSToJoY79emnn/7qq68IIStXrpwxY8aoUaNCQkLM6aOyspLe\nrbI1pVLJk4dfQFz27BFbgh47dmwnP/X09EQJAG8xPEnY2vXr16uqqvyMTzf2yuLFi+/cubNhwwZf\nX1+5XG4wGDIyMjZu3Dh48ODDxgfbuwdPEgL01NKlSw8dOsT4I5lM1noFJeAb0ymOxsbGiRMnZmZm\n0uWQPj4+5veh1WrXrFmjVqtra2vpFgcHhyVLlmzdurUXZ5u1dy/55htCCFmzhgQEkMxMEhv7+8/Q\ngpYuWzQaMmECkUj4Eo/lW3wbGg4RMoEQ48ZI2wi5QMj9UukJpZKEhXEeobBbNm0iWi3571KufYth\nBD1p0qSLFy/2+ROfjY2Nubm51dXVMplMoVDY2Nj04iQ6nc6Qn78sKooQQuRyYm1NWlrI3bu//xgt\naOmy5aGHyKefEmtrvsRj+ZZ/bNmy6e23pYQYqzQMhDQRMm3q1O/PnOFDhMJuqaoizs7EMhgS9Ftv\nvXX+/PlVq1Z5eHgYHyuaMGGChSLoEUxxgLnEt1jSm2++ybit6KZNm9atW8d+PNB9DAnaw8Oj/XHF\nxcXmdHPmzJmbN29qNBq5XP7+++/n5OTMmjVryZIlPT2PTqdraWmZPn06IWTw4MH29vZ1dXXGIhO0\noKXLlubIyDvvvEOkUp7Ew0JLXl5eYGCgm5vbww8/TLekpqbm5+ffu3ePJxEKvcXb21tqmW+eGcYR\nZubi9rZu3fruu+86ODjs2LFj3rx5tbW11tbWy5Ytq6io6MVYuK6ujl5dj54ksba2Ni62hxa0dNly\nLyFBzqd4WGj57bffCCG1tbXXrl2jWwwGg729PX8iFHqLBRcwoSxv5MiRFy5cKC0tlUqlhw4dohuP\nHDmiUql6eqojR47s2rWrrwME6M8SExPb/4f/wgsvcB0XdI2NgtDCwkKlUjlkyBBXV9cHHniAbnz4\n4YdzcnJY6B2gjbVrxfaot729ffvGESNGsB8J9BQbX5UMHjy4srLS2dn52Wefdf7v1513795lvG+M\njh492v5Z8JKSEhTVg1notTjE9KwK44NmCoWC9UCgx9hI0KGhofn5+QqFYvPmzcbGU6dOdf6A6cKF\nCxcuXGjSSFdxWCRKgH6Kcb+r+Pj4xYsXsx8M9AjDOGL16tU//vhj+/Ze27VrV2BgoEnjM8888+mn\nn/ZhLwDdIr61OOrq6to33r59m/1IoKcY7lSKoubMmTN+/Pi33norLy/PQh3L5XIHBwcLnRygQ+Jb\ni4Ne9EYikdja2trZ2dElB/fu3eM6Lugaw526c+fOoqKiLVu2pKenjx8//qGHHvrkk0+qqqr6tuPk\n5OT169f37TkBoD0vLy9CCEVRDQ0NxryMra0EgXkoYW1tPW/evMOHD3///ffl5eVRUVEeHh5RUVH5\n+fl91XFBQUHrvcMBWCK+Kg6q7cNo9NsHHnigD/9zBgthTtAVFRWJiYlBQUGBgYHTpk3T6/XZ2dmO\njo5qtbqvOo6Ojk5OTu6rswF0l/h2VKEXg5ZIJFZWVlZWVvQTFhUVFa6urlyHBl1gqOJYsGDBmTNn\nZs6c+cILL4SGhg4cOJBu37p1q6OjI7vhAYC56B2RKIqSSCQDBw6kZzkoisKXQPzHkKADAgJ27drl\n6elp0i6VSktKSnrXjV6v37t3b1ZWlsFgkMvlKpUqMjKyfWkHgMWJr4qDfjps8ODBBoPBWHJn/kaj\nwAKGO9Xd3d0kOxsfFe3d/3KxJyHwiPiqOOjvA8vLy+kpDnrGIzs7m+u4oGsMq9kpFIrc3NzWLS4u\nLuZsT+Xl5aXT6Uz2JLxw4YJGoykoKOjRqbDcKEBP3blzx8vLy8rKqqGhQSKRODo63r17V6PRdLTN\nCvBHmykOuqyioaGhdX3FjRs3jAs49Q72JAQeWbuWxMWJahA9aNAgFxeXhoaGxx57zN3d/ejRoxKJ\nBI96C0KbBB0aGkoIKS0tpV8QQqRSqZub2/bt283pIyQkJDw8vP2ehGbuRQvQG+Jbi4MQotPpZs+e\nffz48ZaWFkdHxwkTJixdupTroKBrbRI0PeGwYsWK3bt392EffbsnIQD0VG5uro2Njbu7e3Nzc1lZ\nWW1t7X333cd1UNC1P+ag09LSRowYMWTIEMbnR6ZMmWJmT321JyHmoMEsUVFEqxXbCNrDw+P48eOZ\nmZkFBQX+/v5PPvnklStXhg4dynVc0IU/RtCLFi2KjY196qmnFi1a1P44k68Ne8HGxqZP9ggHMMue\nPVxHwIG6urpXXnklKirqoYceOnPmjEQiOX/+vEaj4Tou6MIfCdqYgs3PxQDAK01NTeHh4cuWLSOE\nBAYGbt26VSaTcR0UdI3hD71Lly4VFhYSQsrKylavXv3KK69UVlayHhiAZYhvLQ5CyPjx49etW7d2\n7dqdO3eOGTNm3LhxvX7oDNjEkKCfeOKJlpYWQsjf/va3X3/9tbCw8Pnnn2c9MADLEN9aHISQhx56\n6O23387MzDx27FhMTMzMmTPHjh3LdVDQNYZHvUtLS4cNG9bQ0HDq1KmbN286OjoOGzaM/cgAoK+s\nWbMmNDR0zZo1o0ePPnPmTElJSUBAANdBQdcYRtADBw4sKys7d+7c2LFjBw8eLJFI6uvr2Y8MwCLE\ntxYHIcTd3T0pKamwsPDAgQN/+tOfDh48yHVE0C0MI+jo6OipU6fW1ta+/fbbhJAffvhh9OjRrAcG\nYBmirOIghLi4uKxZs4brKKBnGBL0pk2bHn/8cWtra19fX0KIk5NT3z63AgAA3cH8t55EItm/f/+S\nJUteeeWVhoaGzrffBhASUVZxgEAxJOhDhw4FBQVVVlaqVKrKysqZM2fqdDr2IwOwCFFWcYBAMUxx\nvPHGG0lJScbV9CMjI5977rmwsDB2AwMAEDuGEXRxcfGf//xn49tp06YVFxezGBKAJYmyigMEiuFO\n9fPz27x5c1NTEyGkqakpLi7O/JWSAPhCfDuqgHAxTHHEx8c/9thjO3bsGDp0aGFhoYeHx8mTJ9mP\nDAD60I0bN3Q6XUtLy5IlS0aNGsV1ONAtDEOJMWPGZGZmnjlzZtOmTWfOnLly5QrqoKH/EGUVx+HD\nhx944IETJ06cPn162rRpe8RaDC44DCNoQoiVldXUqVNZDgWADaLcUWX16tX//Oc/58yZQwj56aef\nZs2aFRUVxXVQ0DWGBP3qq6+atAwYMMDb23vu3LlY4RtAcOrr65uamujsTAiZPHmyra1tcXGxh4cH\nt4FBlxgSdHl5+cGDB2fPnj1s2LD8/PwvvvgiLCzsP//5z5o1a44cORIcHMxOZF9++eWnn35q0piT\nk4OhPZhFfFUcAwYMaG5uTk9Pf/PNN8vKyv70pz/V1dWZuRM0sIMhQRcXFyclJT344IP022+++ea9\n9947derU6dOnY2JiWEvQfn5+Li4uJo0pKSl2dnbsBAD9kyinX8eNGzdlypTJkydbW1ufOHGiubm5\nvr7ewcGB67igC3/sSWjk6OhYUVFhZWVFv21ubnZ1db179y5FUYMGDfrtt99YD/IP2JMQoBd8fHwG\nDRpUVFQkkUi8vb0pinr22WefeeYZruOCLjD8rTd27Njt27fTa/a3tLRs376dXtu7tLTU3t6e7QAB\n+pb4qjgqKyspikpISMjLy7t169Z3331XU1NTVlbGdVzQNYYpDq1Wu3Dhwi1btnh5eRUVFTk6Oh4/\nfpwQcv369Y0bN7IdIEDfEl8Vh42NjbOz87FjxyZOnEi3lJWVzZgxg9uooDsYErSvr292dvaPP/54\n+/ZtLy8vPz8/erpjxowZ+EcFEBwHBweFQpGRkbF48eL777//s88+UygU06dP5zou6BrDOGL16tXp\n6enTpk0LDQ319/c3TkYD9Afiq+IghHz00UcuLi55eXlHjhyZNm3a119/zXVE0C0MI2iKoubMmTN4\n8OAnn3wyIiJi+PDh7IcFYCmirOIYNGjQRx99pNPpcnNzJ02aNGDAAK4jgm5hGErs3LmzqKhoy5Yt\n6enp48ePf+ihhz755JOqqir2gwOAPlFTU6NWq+/cufPnP//54sWLS5cubV++BTzE/LeetbX1vHnz\nDh8+/P3335eXl0dFRXl4eERFReXn57McH0AfE18VByHk448/joqKWr169axZs2JjYxUKxZdffsl1\nUNA15gRdUVGRmJgYFBQUGBg4bdo0vV6fnZ3t6OioVqtZjg+gj4lyR5WrV6+2XuQ9ICAgMzOTw3ig\nmxjmoBcsWHDmzJmZM2e+8MILoaGhAwcOpNu3bt3q6OjIbngA0AdGjx6dkZFhXJby8uXLkydP5jYk\n6A6GBB0QELBr1y5PT0+TdqlUWlJSwkpUABYjyiqO559/Pjg4uKGh4b777jt79uzFixfXrVvHdVDQ\nNYY79ejRoyYr9M+fP59+gYf3QfBEuaOKs7Pz6dOnb9++nZCQYG9vf+LECZTPCgLDWhxyudzb23vG\njBkffvghXY7j7e1dUFDARXimsBYHAIgHw1DCwcHh4sWLZWVlQUFBhYWF7McEYEGirOIAgWL+W49e\nf2POnDlTp05NTU1lOSYACxJlFQcIFPOWV4QQiUSyYcOGyZMnh4aGGgwGNmMCAADCmKA/+ugj4+uQ\nkJDvvvuu/c4mAEIlyioOECiGLwkJIT/99NPBgwcLCwuHDh0aERExadIk9iNjhC8JAUA8GIYShw4d\nCgoKqqysVKlUlZWVM2fO1Ol0Znaj1+ujo6MDAgImTJgQEBAQHR2t1+vNPCcAQP/GkKDfeOONpKQk\nrVa7fv16rVablJS0YcMGc/pISEigdzKMiIiIiYmJiIiQSCTBwcEJCQnmnBagN1DFAcLBMMXh7Oxc\nVlZmY2NDv21oaHB3d6+oqOh1H15eXjqdzmSx/wsXLmg0mp6WV2OKA8z1yCPk9Gli3eHX4wD8wTCC\n9vPz27x5c1NTEyGkqakpLi5uypQp5vRRWVmpVCpNGpVKpTlJHwCg32NI0PHx8YcPH3Zzc7vvvvvc\n3Nw+/fTT+Ph4c/oICQkJDw/X6/VVVVUURVVVVaWmpmo0mpCQEHNOC9AbqOIA4WD4Q2/MmDGZmZk/\n/fRTUVGRl5fX5MmTzXxsX6vVrlmzRq1W19bW0i0ODg5LlizZunWrOacF6A1R7qgCAmU6B93Y2Dhx\n4sTMzEyJRNK3PTU2Nubm5lZXV8tkMoVCYZzj7hHMQQOAeJiOoG1sbGxtbRsbG21tbfu2JxsbGx8f\nn749J0CPrV1L4uIwywGCwHCbhoWFLVy4MCUl5cqVK5n/ZWY3N27c0Ol0P//8s7Glubl5+/btZp4W\noMewFgcIB0OZnYeHR/vjiouLe93H559/HhYWNnz48Fu3bj333HM7duywsrKqq6uzs7Pr6c6VmOIA\nc6HMDoSDYQRdzMScPl5//fWdO3dev3791q1bWVlZS5YsaWxsNOeEAL2HKg4QDuY79d///vfTTz+9\nYMECQsilS5e++OILc/q4fv06fSpPT8/k5OR79+4tWrSovr7enHMC9JIod1QBgWL4Q2/btm0JCQkv\nvvjitm3bCCEODg6rVq2aPXt2r/vw8PDIy8sbMmQIIWTAgAFHjx4NCwt7/PHHO/9URkbG999/b9KY\nlpY2bNiwXkcCACAgDEOJ7du3p6SkrFy5kmH1E78AABQASURBVC7kUCqVv/zyizl9zJkzZ+/evca3\ntra2n332GZ2vOzFw4ECXdmQyGfZSA7NgLQ4QDoYRdG1trZubW+u3dnZ25vSxbds2kwkNa2vrgwcP\nvvPOO518asyYMWPGjDFppCiqvLzcnGBA7OgqDsxygBAw3KZBQUEbNmxobm4mhFAUtXHjxlmzZpnT\nh62trVwuN2m0srJSKBTmnBYAoH9jGEF/8MEHCxYscHV1ra+v9/T0HDVq1IkTJ/q84+Tk5NTU1E2b\nNvX5mQE6gyoOEA6GBO3p6Xn+/PmMjIxbt24NHz7c19e3zx/7JoQUFBSkpaX1+WkBuoC1OEA4GIYS\n+/btk0gkvr6+8+fPnzhxokQiSUxM7POOo6Ojk5OT+/y0AAD9BkOCfv31101a/v73v7MSDIDloYoD\nhKPNFAc959DQ0NB68uHGjRvtv+LrKb1ev3fv3qysLIPBIJfLVSpVZGRkYGCgmacF6DFUcYBwtEnQ\noaGhhJDS0lL6BSFEKpW6ubmZuapRQkLC3//+97CwsIiICCcnp7t376anpwcHB2/ZsmX58uXmnBkA\noB9rk6DpHQJXrFixe/fuPuwjNjY2KSnJZE/CqKgojUaDBA1sQxUHCAfDnbp79+6+XYsDexICj2At\nDhAOhjt127ZtL730kp+f3+XLlwkhDg4O7b827BHsSQgA0AtsrMWh1WoVCoVarXZycpJKpU5OTmq1\neuTIkZao3gPoAqo4QDjYWItDLpdrtdr4+Hjz9yQEMBeqOEA42FiLg0bvSThp0iQfHx9kZwCALnG2\nFgcAN1DFAcLB2VocANzAWhwgHMxbZ9Jrcfj6+rIcDQAAGLVJ0O2rlY2ys7MtHwyA5a1dS+LiMMsB\ngtAmQW/cuNH4+rnnnnv77bddXV3ZjgjAolDFAcLRJkGHh4cbX7/44ovz58/39vZmPSQAACCEscwO\noD9DFQcIB/OXhAD9Fqo4QDjaJOgDBw4YX9fX1x8/ftzFxYV++8QTT7AaFwCA6EkoijK+mTBhQkfH\nZWZmshJPF3Q6XXl5+YoVK7gOBAQLVRwgHG1G0DzJwgAWhCoOEA7cpgAAPIUEDSKDKg4QDlRxgMig\nigOEA0MJAACeQoIGkcGOKiAc/J3iKCsry8vLM2nMycmhN+IC6CVUcYBw8DdBX7169fTp0yaN2dnZ\nKpWKk3gAAFjG3wQdFBQUFBRk0kg/qMJJPNBPoIoDhIO/CRrAIlDFAcKBoQQAAE8hQYPIoIoDhAMJ\nGkSGruIAEAIkaAAAnkKCBpFBFQcIB6o4QGRQxQHCgaEEAABPIUGDyKCKA4QDCRpEBlUcIBxI0AAA\nPIUEDSKDKg4QDlRxgMigigOEA0MJAACeQoIGkUEVBwgHEjSIDKo4QDiQoAEAeIqlLwn1ev3evXuz\nsrIMBoNcLlepVJGRkYGBgez0DvAHVHGAcLBxpyYkJAQHBxNCIiIiYmJiIiIiJBJJcHBwQkICC70D\ntLFnDxI0CAUbI+jY2NikpKQZM2a0boyKitJoNMuXL2chAAAAIWJjKFFZWalUKk0alUplRUUFC70D\ntIEqDhAONhJ0SEhIeHi4Xq+vqqqiKKqqqio1NVWj0YSEhLDQO0AbqOIA4WAjQWu1WoVCoVarnZyc\npFKpk5OTWq0eOXJkYmIiC70DAAgUG3PQcrlcq9XGx8fn5uZWV1fLZDKFQmFjY8NC1wCmUMUBwsHe\nWhw2NjY+Pj6sdQfADGtxgHBwNpRoamp69dVXueodAID/uEzQ77zzDle9g3ihigOEg40pDsZi5+bm\nZha6BjBFV3FgGhqEgI3b9H//93+rq6ut22GhawAA4WIjS6pUqvDw8Hnz5rVurKur6/xR76amJoPB\nYNJYU1PT9/GBqKCKA4SDjQT9zDPP1NfXm3Zsbb1u3bpOPnXy5Mn4+HiTxpKSktmzZ/dxfCAqqOIA\n4ZBQFMV1DD2g0+nKy8tXrFjBdSAAABaHv/VAZFDFAcLBWYJOTk5ev349V72DeGEtDhAOzhJ0QUFB\nWloaV70DAPAfZwk6Ojo6OTmZq95BvFDFAcKBYmQQGVRxgHBgT0IAAJ7CnoQgMqjiAOHAnoQgMliL\nA4QDexICAPAU9iQEkUEVBwgH9iQEkdmzBwkahIK9tTgaGxvN35MQa3EAgHhgT0IQmbVrSVwcBtEg\nCLhNQWSwFgcIBxI0AABPIUGDyKCKA4QDa3GAyGAtDhAODCUAAHgKCRpEBmtxgHAgQYPIoIoDhAMJ\nGgCAp5CgQWRQxQHCgSoOEBlUcYBwYCgBAMBTSNAgMqjiAOFAggaRQRUHCAcSNAAATyFBg8igigOE\nA1UcIDKo4gDhwFACAICn+DuCPnr0aEJCgkljSUnJrFmzOIkH+gnsqALCwd8EvXDhwoULF5o00nsS\nchIP9BN0FQcSNAgBblMAAJ5CggaRQRUHCAd/pzgALAJVHCAcGEoAAPAUEjSIDNbiAOFAggaRwVoc\nIBxI0AAAPIUEDSKDKg4QDlRxgMigigOEA0MJAACeQoIGkUEVBwgHEjSIDKo4QDiQoAEAeAoJGkQG\nVRwgHKjiAJFBFQcIB4YSAAA8hQQNIoMqDhAOJGgQGVRxgHCwNAet1+v37t2blZVlMBjkcrlKpYqM\njAwMDGSndwAAIWJjBJ2QkBAcHEwIiYiIiImJiYiIkEgkwcHB7feEBbA4VHGAcEgoirJ0H15eXjqd\nbsaMGa0bL1y4oNFoCgoKenQqetPYFStW9GmAAAB8xMZQorKyUqlUmjQqlcqKigoWegcAECg2EnRI\nSEh4eLher6+qqqIoqqqqKjU1VaPRhISEsNA7QBuo4gDhYCNBa7VahUKhVqudnJykUqmTk5NarR45\ncmRiYiILvQO0gSoOEA42qjjkcrlWq42Pj8/Nza2urpbJZAqFwsbGhoWuAQCEi71HvW1sbHx8fFjr\nDoAZqjhAOFAHDSKDtThAOFAHDQDAU2yMoGNjY5OSkkzqoKOiojQazfLly1kIAOAPa9eSuDjMcoAg\noA4aRAZVHCAcbIyg6TroDRs2+Pr6yuVyg8GQkZGxcePGzuugT548eeDAAZPGwsJCzFwDgEiw8ai3\nwWBYs2bNoUOHamtr6RYHB4clS5Zs3bpVLpd39KnGxsbq6mqTxpaWFrlcbmtra8FwoX+LiiJaLaY4\nQBDYSNC0xsZG1EEDAHQfewkaAAB6hLM/9JKTk9evX89V7yBeWIsDhIOzBF1QUJCWlsZV7yBeqOIA\n4cAUB4jMI4+Q06eJNfazBwHAd9kgMliLA4SDpRE01uIAAOgprMUBAMBTAtuTEMBcWIsDhANrcYDI\noIoDhAN7EgIA8BR/1+Iw35QpU1xcXCx3fqAZDIaamhoPDw+uA+mWV65efX/cOEoi4TqQ3sjJyRkx\nYoSVlRXXgfRzzc3N9vb2p06d4jqQfr0Wx1/+8pezZ89auhdISUm5dOnSunXruA6k/1u4cGFiYqKr\nqyvXgfRzpaWlK1euPHLkCNeBYE9CAAC+wnfZAAA8hQQNAMBTSNAAADyFBA0AwFP9OUFbY8UyVlhZ\nWaHwix1SqVSKZyAtjz/XuT8vN0ovzMR1FP1fc3NzfX29vb0914H0f7ilWcOTS92fEzQAgKDxYhgP\nAADtIUEDAPAUEjQAAE8hQQMA8BQSNAAATyFBAwDwFBI0AABPCSlBGwyG8PBwBwcHb2/vjjac7eiY\njtp37NgxadIkGxub5cuXW/wXECBzrjmuba9157Lj8loC366qkB6Gfumll0pLS3Nzc7Ozs+fOnatS\nqQIDA7t5TEft3t7esbGxfFiZm5/Muea4tr3WncuOy2sJvLuqlEA0NDQ4ODh8++239Ntly5YtW7as\nm8d0+dm//vWvzz//vGV/AQEy55ob4dr2VHcuuxEuryXw56oKZoojNze3pqZm4sSJ9NuJEydmZWV1\n85jufBbaM+eaQ6/hkoKRYBJ0dXU1IcS4fImTk5PBYOjmMd35LLRnzjWHXsMlBSPBJGiZTEYIMd6p\nd+/ebb/WVEfHdOez0J451xx6DZcUjASToBUKhb29fUZGBv02PT1dpVJ185jufBbaM+eaQ6/hkoKR\nYBK0jY3NkiVLYmNjKyoqvv/+e51OFxkZSf9Iq9UmJyd3ckwnn21qaqqrq2tubm5ubq6rq2tqauLq\nF+Qhc645wbXtre5cdoLLaxm8u6pcf0vZA1VVVWFhYfb29p6envHx8cb2Rx99dN26dZ0f01H7unXr\nWl+NmJgY1n4dQTDnmuPa9lp3LjsuryXw7apiwX4AAJ4SzBQHAIDYIEEDAPAUEjQAAE8hQQMA8BQS\nNAAATyFBAwDwFBI0AABPIUEDAPAUEjQAAE8hQQMA8BQSNAAATyFBAwDwFBI0AABPIUEDAPAUEjQA\nAE8hQQMA8BQSNAAATyFBAwDwFBI0WJBCoUhLS+vz0+bm5k6dOtXFxeXDDz+0aEgWih+gm5Cg+60J\nEyZIJBKJRDJkyJClS5feuXOHhU69vb1bZ7TY2NgRI0b0eS87d+6cOXNmRUXFiy++2NPP9igkC8Xf\nV0yuNvQ/SND92d69exsbG8+fP5+dnf3aa6+1/lGf7yfPeMKnnnpqyJAhfdsRISQvL2/cuHG9+2yP\nQrJQ/ADdhATdn0mlUmtrax8fn8WLF6enpxNCPDw83nrrrYkTJ953332EkBs3bjzyyCMuLi7jxo3T\n6XT0pzw8PN58882AgACVSrV8+fKGhga6vaODjSdcsWJFcXHx/PnzFQrFvn37SKspgo4++8477/j7\n+/v4+Dz99NOMKb79ByMjI5OSkl599VWFQnHt2rXWB3t4eOzevXvMmDEymezll1/Oz88PCgqSyWQh\nISE1NTX0McaQNmzY4Onp6ezsPHbs2EuXLjG2GA/uKNTMzEx/f38nJ6d58+ZFR0e/+uqr7X+Fjo4p\nKSkJCwtzc3NTKBTbtm3r/Jq0vw4mV/vdd99duHChsdNVq1atXr26k39Nxt7bXwHgGAX9lEql2r9/\nP0VRBoPhwQcffPLJJymKcnd3nzt3bl1dXUtLS1NT07hx4/7xj3/U19efO3dOJpP95z//MR7T2NjY\n2Ng4d+7cN998k6Kozg+mT0hR1NChQ3/44QdjDCNGjPjhhx86+eyCBQsaGhoaGhr8/f3/7//+z+RX\n6OiDjz32WGJiYvtf2d3d/eGHHy4rK7t586azs3NAQEB6errBYJg+ffr27dtbh5SWljZixIjS0lKK\nonJycvLz89u3GA/uKNTGxsZRo0a99957TU1NZ8+etbOzi4mJMQmpo2NaWloCAgJefvnle/fu5ebm\njhkz5tSpUx111NF1aH21i4qK7O3tKyoq6E6HDBmSlpbW0b8mY++MVwC4hQTdb6lUKicnJ3d39wED\nBjz44IMFBQUURbm7u585c4Y+4NKlSy4uLo2NjfTbZcuWvfzyy/QxSUlJdGNycvL48eM7P9h4QqqD\nBN3JZ7/55hu68bXXXqMbW+vog50k6JSUFPp1SEjIunXr6Nfvv/9+ZGRk65AuX748ZMiQlJSU+vp6\nur19C9U2QbcP9eLFi25ubs3NzXT7Y4891j5Bd3TM5cuXHR0dm5qa6PadO3fSETJ21NF1MLnaarX6\n448/pijqX//617hx44zXpP2/JmPvjFcAuIUpjv4sLi4uPT29qqrq7NmzQ4cOpRs9PT3pF0VFRUOH\nDrW2tqbfKhSKwsJC+rW7uzv9wsPDo6ioqPODjSfsSCefHTRoEP3Czs6uurq6+x/siIeHB/3C3t6+\n9WuTk/v6+r799ttr164dMmTIE088UVZW1r7F5MztQy0qKvLy8pJKf/+PaNiwYYy/O+Mxt27dqq+v\nV6lUSqVSqVRu27atqqqqk466cx0iIyMPHDhACDlw4MCTTz5pbG//r8nYe5dXANiHBN2f0SNoW1vb\n1o0SiYR+4eXlVVhYaJzlzM3NNSbxX3/9lX6Rk5Pj5eXV+cHGExJCjJmotU4+27lef7A7oqKifvjh\nh5s3b969e3fTpk2MLZ3z9PQsKipqaWmh3+bn53f/mGHDhjk7O//888/Z2dnZ2dk5OTnHjh3rqKOO\nroPJ1Q4NDc3IyMjMzDx16lRERISxvf2/Zke99/QKgKUhQYvX5MmT3dzc4uLiGhsbv/vuO51Ot3Tp\nUvpH7733XllZ2Z07dzZv3rx48eLOD27Nzc3txo0b3e+o1xGaKSsr6/z5801NTY6OjjKZzNraun1L\nlyfx8/OTyWTbtm1rbm7+5ptvvvjii+4f4+vrO3LkyJiYmOrq6ubm5qtXr3bypVxH18Hkag8cOHDR\nokVLly719/cfPny4sb39vyZj7724AmBpSNDiZWVldfLkyXPnzg0ZMmTZsmUfffTR5MmT6R+Fh4dP\nnz59zJgxkyZNiomJ6fzg1l577bX/+Z//cXFxSUxM7E5HvY7QTNXV1X/9619dXV2HDh3a1NS0fv36\n9i1dnsTa2vrYsWOHDx92cXF57733wsLCBgwY0M1jpFLpiRMnioqKRo8ePXjw4KioqMrKyo466ug6\ntL/akZGRV65caT2/QZj+NRl778UVAEuTUBTFdQzALx4eHl9++eWECRO4DkRg5s+fHxIS8uyzz5p5\njDny8vKUSmVxcbGjoyPdgn9NQcMIGqD3UlNT6bq006dPf/3113Pnzu3dMX2ipaVl69at4eHhxuwM\nQodpJoDey87OXrRoUX19/dChQw8ePMj4HWZ3jjFfTU2Nu7v7iBEjkpOTLXF+4ASmOAAAeApTHAAA\nPIUEDQDAU0jQAAA8hQQNAMBTSNAAADyFBA0AwFNI0AAAPIUEDQDAU0jQAAA8hQQNAMBTSNAAADyF\nBA0AwFNI0AAAPIUEDQDAU0jQAAA89f9nXX5C9klUyQAAAABJRU5ErkJggg==\n"
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "PLINK v1.90b4.1 64-bit (30 Mar 2017)           www.cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2017 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to clean-merged_neighbor.log.\n",
      "Options in effect:\n",
      "  --bfile clean-merged_neighbor\n",
      "  --missing\n",
      "  --out clean-merged_neighbor\n",
      "\n",
      "64329 MB RAM detected; reserving 32164 MB for main workspace.\n",
      "547812 variants loaded from .bim file.\n",
      "5183 people (2299 males, 2752 females, 132 ambiguous) loaded from .fam.\n",
      "Ambiguous sex IDs written to clean-merged_neighbor.nosex .\n",
      "5051 phenotype values loaded from .fam.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 5183 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 0%\b\b1%\b\b2%\b\b3%\b\b4%\b\b5%\b\b6%\b\b7%\b\b8%\b\b9%\b\b10%\b\b\b11%\b\b\b12%\b\b\b13%\b\b\b14%\b\b\b15%\b\b\b16%\b\b\b17%\b\b\b18%\b\b\b19%\b\b\b20%\b\b\b21%\b\b\b22%\b\b\b23%\b\b\b24%\b\b\b25%\b\b\b26%\b\b\b27%\b\b\b28%\b\b\b29%\b\b\b30%\b\b\b31%\b\b\b32%\b\b\b33%\b\b\b34%\b\b\b35%\b\b\b36%\b\b\b37%\b\b\b38%\b\b\b39%\b\b\b40%\b\b\b41%\b\b\b42%\b\b\b43%\b\b\b44%\b\b\b45%\b\b\b46%\b\b\b47%\b\b\b48%\b\b\b49%\b\b\b50%\b\b\b51%\b\b\b52%\b\b\b53%\b\b\b54%\b\b\b55%\b\b\b56%\b\b\b57%\b\b\b58%\b\b\b59%\b\b\b60%\b\b\b61%\b\b\b62%\b\b\b63%\b\b\b64%\b\b\b65%\b\b\b66%\b\b\b67%\b\b\b68%\b\b\b69%\b\b\b70%\b\b\b71%\b\b\b72%\b\b\b73%\b\b\b74%\b\b\b75%\b\b\b76%\b\b\b77%\b\b\b78%\b\b\b79%\b\b\b80%\b\b\b81%\b\b\b82%\b\b\b83%\b\b\b84%\b\b\b85%\b\b\b86%\b\b\b87%\b\b\b88%\b\b\b89%\b\b\b90%\b\b\b91%\b\b\b92%\b\b\b93%\b\b\b94%\b\b\b95%\b\b\b96%\b\b\b97%\b\b\b98%\b\b\b99%\b\b\b\b done.\n",
      "Warning: 11574 het. haploid genotypes present (see clean-merged_neighbor.hh );\n",
      "many commands treat these as missing.\n",
      "Warning: Nonmissing nonmale Y chromosome genotype(s) present; many commands\n",
      "treat these as missing.\n",
      "Total genotyping rate is 0.999488.\n",
      "--missing: Sample missing data report written to clean-merged_neighbor.imiss,\n",
      "and variant-based missing data report written to clean-merged_neighbor.lmiss.\n"
     ]
    }
   ],
   "source": [
    "'''Exclude individuals with less than 99% callrate (1% or more of genetic variants are missing in that sample).'''\n",
    "#First, generate missingness reports:\n",
    "!plink --bfile $output_filestem --missing --out $output_filestem\n",
    "imiss_file = output_filestem + '.imiss'\n",
    "#Clean up the whitespace:\n",
    "!perl -i.bak -lpe 's/^\\s+//; s/\\s+/\\t/g;' $imiss_file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of individuals who failed heterozygosity QC:['103']\n"
     ]
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "PLINK v1.90b3.45 64-bit (13 Jan 2017)      https://www.cog-genomics.org/plink2\n",
      "(C) 2005-2017 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to clean-merged_neighbor.log.\n",
      "Options in effect:\n",
      "  --bfile merged_neighbor\n",
      "  --geno 0.01\n",
      "  --make-bed\n",
      "  --out clean-merged_neighbor\n",
      "\n",
      "995 MB RAM detected; reserving 497 MB for main workspace.\n",
      "655216 variants loaded from .bim file.\n",
      "5183 people (2299 males, 2752 females, 132 ambiguous) loaded from .fam.\n",
      "Ambiguous sex IDs written to clean-merged_neighbor.nosex .\n",
      "5051 phenotype values loaded from .fam.\n",
      "Using 1 thread.\n",
      "Before main variant filters, 5183 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 0%\b\b1%\b\b2%\b\b3%\b\b4%\b\b5%\b\b6%\b\b7%\b\b8%\b\b9%\b\b10%\b\b\b11%\b\b\b12%\b\b\b13%\b\b\b14%\b\b\b15%\b\b\b16%\b\b\b17%\b\b\b18%\b\b\b19%\b\b\b20%\b\b\b21%\b\b\b22%\b\b\b23%\b\b\b24%\b\b\b25%\b\b\b26%\b\b\b27%\b\b\b28%\b\b\b29%\b\b\b30%\b\b\b31%\b\b\b32%\b\b\b33%\b\b\b34%\b\b\b35%\b\b\b36%\b\b\b37%\b\b\b38%\b\b\b39%\b\b\b40%\b\b\b41%\b\b\b42%\b\b\b43%\b\b\b44%\b\b\b45%\b\b\b46%\b\b\b47%\b\b\b48%\b\b\b49%\b\b\b50%\b\b\b51%\b\b\b52%\b\b\b53%\b\b\b54%\b\b\b55%\b\b\b56%\b\b\b57%\b\b\b58%\b\b\b59%\b\b\b60%\b\b\b61%\b\b\b62%\b\b\b63%\b\b\b64%\b\b\b65%\b\b\b66%\b\b\b67%\b\b\b68%\b\b\b69%\b\b\b70%\b\b\b71%\b\b\b72%\b\b\b73%\b\b\b74%\b\b\b75%\b\b\b76%\b\b\b77%\b\b\b78%\b\b\b79%\b\b\b80%\b\b\b81%\b\b\b82%\b\b\b83%\b\b\b84%\b\b\b85%\b\b\b86%\b\b\b87%\b\b\b88%\b\b\b89%\b\b\b90%\b\b\b91%\b\b\b92%\b\b\b93%\b\b\b94%\b\b\b95%\b\b\b96%\b\b\b97%\b\b\b98%\b\b\b99%\b\b\b\b done.\n",
      "Warning: 11950 het. haploid genotypes present (see clean-merged_neighbor.hh );\n",
      "many commands treat these as missing.\n",
      "Warning: Nonmissing nonmale Y chromosome genotype(s) present; many commands\n",
      "treat these as missing.\n",
      "Total genotyping rate is 0.849377.\n",
      "107404 variants removed due to missing genotype data (--geno).\n",
      "547812 variants and 5183 people pass filters and QC.\n",
      "Phenotype data is quantitative.\n",
      "--make-bed to clean-merged_neighbor.bed + clean-merged_neighbor.bim +\n",
      "clean-merged_neighbor.fam ... 0%\b\b1%\b\b2%\b\b3%\b\b4%\b\b5%\b\b6%\b\b7%\b\b8%\b\b9%\b\b10%\b\b\b11%\b\b\b12%\b\b\b13%\b\b\b14%\b\b\b15%\b\b\b16%\b\b\b17%\b\b\b18%\b\b\b19%\b\b\b20%\b\b\b21%\b\b\b22%\b\b\b23%\b\b\b24%\b\b\b25%\b\b\b26%\b\b\b27%\b\b\b28%\b\b\b29%\b\b\b30%\b\b\b31%\b\b\b32%\b\b\b33%\b\b\b34%\b\b\b35%\b\b\b36%\b\b\b37%\b\b\b38%\b\b\b39%\b\b\b40%\b\b\b41%\b\b\b42%\b\b\b43%\b\b\b44%\b\b\b45%\b\b\b46%\b\b\b47%\b\b\b48%\b\b\b49%\b\b\b50%\b\b\b51%\b\b\b52%\b\b\b53%\b\b\b54%\b\b\b55%\b\b\b56%\b\b\b57%\b\b\b58%\b\b\b59%\b\b\b60%\b\b\b61%\b\b\b62%\b\b\b63%\b\b\b64%\b\b\b65%\b\b\b66%\b\b\b67%\b\b\b68%\b\b\b69%\b\b\b70%\b\b\b71%\b\b\b72%\b\b\b73%\b\b\b74%\b\b\b75%\b\b\b76%\b\b\b77%\b\b\b78%\b\b\b79%\b\b\b80%\b\b\b81%\b\b\b82%\b\b\b83%\b\b\b84%\b\b\b85%\b\b\b86%\b\b\b87%\b\b\b88%\b\b\b89%\b\b\b90%\b\b\b91%\b\b\b92%\b\b\b93%\b\b\b94%\b\b\b95%\b\b\b96%\b\b\b97%\b\b\b98%\b\b\b99%\b\b\bdone.\n"
     ]
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "PLINK v1.90b3.45 64-bit (13 Jan 2017)      https://www.cog-genomics.org/plink2\n",
      "(C) 2005-2017 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to clean-merged_neighbor.log.\n",
      "Options in effect:\n",
      "  --bfile clean-merged_neighbor\n",
      "  --missing\n",
      "  --out clean-merged_neighbor\n",
      "\n",
      "995 MB RAM detected; reserving 497 MB for main workspace.\n",
      "547812 variants loaded from .bim file.\n",
      "5183 people (2299 males, 2752 females, 132 ambiguous) loaded from .fam.\n",
      "Ambiguous sex IDs written to clean-merged_neighbor.nosex .\n",
      "5051 phenotype values loaded from .fam.\n",
      "Using 1 thread.\n",
      "Before main variant filters, 5183 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 0%\b\b1%\b\b2%\b\b3%\b\b4%\b\b5%\b\b6%\b\b7%\b\b8%\b\b9%\b\b10%\b\b\b11%\b\b\b12%\b\b\b13%\b\b\b14%\b\b\b15%\b\b\b16%\b\b\b17%\b\b\b18%\b\b\b19%\b\b\b20%\b\b\b21%\b\b\b22%\b\b\b23%\b\b\b24%\b\b\b25%\b\b\b26%\b\b\b27%\b\b\b28%\b\b\b29%\b\b\b30%\b\b\b31%\b\b\b32%\b\b\b33%\b\b\b34%\b\b\b35%\b\b\b36%\b\b\b37%\b\b\b38%\b\b\b39%\b\b\b40%\b\b\b41%\b\b\b42%\b\b\b43%\b\b\b44%\b\b\b45%\b\b\b46%\b\b\b47%\b\b\b48%\b\b\b49%\b\b\b50%\b\b\b51%\b\b\b52%\b\b\b53%\b\b\b54%\b\b\b55%\b\b\b56%\b\b\b57%\b\b\b58%\b\b\b59%\b\b\b60%\b\b\b61%\b\b\b62%\b\b\b63%\b\b\b64%\b\b\b65%\b\b\b66%\b\b\b67%\b\b\b68%\b\b\b69%\b\b\b70%\b\b\b71%\b\b\b72%\b\b\b73%\b\b\b74%\b\b\b75%\b\b\b76%\b\b\b77%\b\b\b78%\b\b\b79%\b\b\b80%\b\b\b81%\b\b\b82%\b\b\b83%\b\b\b84%\b\b\b85%\b\b\b86%\b\b\b87%\b\b\b88%\b\b\b89%\b\b\b90%\b\b\b91%\b\b\b92%\b\b\b93%\b\b\b94%\b\b\b95%\b\b\b96%\b\b\b97%\b\b\b98%\b\b\b99%\b\b\b\b done.\n",
      "Warning: 11574 het. haploid genotypes present (see clean-merged_neighbor.hh );\n",
      "many commands treat these as missing.\n",
      "Warning: Nonmissing nonmale Y chromosome genotype(s) present; many commands\n",
      "treat these as missing.\n",
      "Total genotyping rate is 0.999488.\n",
      "--missing: Sample missing data report written to clean-merged_neighbor.imiss,\n",
      "and variant-based missing data report written to clean-merged_neighbor.lmiss.\n"
     ]
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'data.frame':\t5183 obs. of  6 variables:\n",
       " $ FID       : Factor w/ 5073 levels \"ME8001\",\"ME8002\",..: 1 2 3 4 5 6 7 8 9 10 ...\n",
       " $ IID       : Factor w/ 5183 levels \"DU9001\",\"DU9002\",..: 51 52 53 54 55 56 57 58 59 60 ...\n",
       " $ MISS_PHENO: Factor w/ 2 levels \"N\",\"Y\": 2 2 2 2 2 2 2 2 2 2 ...\n",
       " $ N_MISS    : int  46 37 34 27 27 77 634 32 27 27 ...\n",
       " $ N_GENO    : int  547804 547804 547804 547804 547804 547804 547804 547804 547804 547804 ...\n",
       " $ F_MISS    : num  8.40e-05 6.75e-05 6.21e-05 4.93e-05 4.93e-05 ...\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of samples that failed callrate QC: ['20']\n"
     ]
    }
   ],
   "source": [
    "'''Isolate the FID:IID pairs for individuals that failed callrate QC.'''\n",
    "%Rpush imiss_file\n",
    "%R imiss <- read.table(imiss_file, header=T)\n",
    "%R str(imiss)\n",
    "%R write.table(imiss[which(imiss$F_MISS >= 0.01),c(1,2)], file='fail-imiss-qc.txt', quote=F, row.names=F, sep=\"\\t\")\n",
    "imiss_fail_count = !cat fail-imiss-qc.txt | wc -l\n",
    "print \"Number of samples that failed callrate QC: \" + str(imiss_fail_count)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "PLINK v1.90b3.45 64-bit (13 Jan 2017)      https://www.cog-genomics.org/plink2\n",
      "(C) 2005-2017 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to clean-merged_neighbor.log.\n",
      "Options in effect:\n",
      "  --bfile clean-merged_neighbor\n",
      "  --make-bed\n",
      "  --out clean-merged_neighbor\n",
      "  --remove fail-imiss-qc.txt\n",
      "\n",
      "995 MB RAM detected; reserving 497 MB for main workspace.\n",
      "Note: --make-bed input and output filenames match.  Appending '~' to input\n",
      "filenames.\n",
      "547812 variants loaded from .bim file.\n",
      "5183 people (2299 males, 2752 females, 132 ambiguous) loaded from .fam.\n",
      "Ambiguous sex IDs written to clean-merged_neighbor.nosex .\n",
      "5051 phenotype values loaded from .fam.\n",
      "--remove: 5164 people remaining.\n",
      "Using 1 thread.\n",
      "Before main variant filters, 5164 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 0%\b\b1%\b\b2%\b\b3%\b\b4%\b\b5%\b\b6%\b\b7%\b\b8%\b\b9%\b\b10%\b\b\b11%\b\b\b12%\b\b\b13%\b\b\b14%\b\b\b15%\b\b\b16%\b\b\b17%\b\b\b18%\b\b\b19%\b\b\b20%\b\b\b21%\b\b\b22%\b\b\b23%\b\b\b24%\b\b\b25%\b\b\b26%\b\b\b27%\b\b\b28%\b\b\b29%\b\b\b30%\b\b\b31%\b\b\b32%\b\b\b33%\b\b\b34%\b\b\b35%\b\b\b36%\b\b\b37%\b\b\b38%\b\b\b39%\b\b\b40%\b\b\b41%\b\b\b42%\b\b\b43%\b\b\b44%\b\b\b45%\b\b\b46%\b\b\b47%\b\b\b48%\b\b\b49%\b\b\b50%\b\b\b51%\b\b\b52%\b\b\b53%\b\b\b54%\b\b\b55%\b\b\b56%\b\b\b57%\b\b\b58%\b\b\b59%\b\b\b60%\b\b\b61%\b\b\b62%\b\b\b63%\b\b\b64%\b\b\b65%\b\b\b66%\b\b\b67%\b\b\b68%\b\b\b69%\b\b\b70%\b\b\b71%\b\b\b72%\b\b\b73%\b\b\b74%\b\b\b75%\b\b\b76%\b\b\b77%\b\b\b78%\b\b\b79%\b\b\b80%\b\b\b81%\b\b\b82%\b\b\b83%\b\b\b84%\b\b\b85%\b\b\b86%\b\b\b87%\b\b\b88%\b\b\b89%\b\b\b90%\b\b\b91%\b\b\b92%\b\b\b93%\b\b\b94%\b\b\b95%\b\b\b96%\b\b\b97%\b\b\b98%\b\b\b99%\b\b\b\b done.\n",
      "Warning: 11469 het. haploid genotypes present (see clean-merged_neighbor.hh );\n",
      "many commands treat these as missing.\n",
      "Warning: Nonmissing nonmale Y chromosome genotype(s) present; many commands\n",
      "treat these as missing.\n",
      "Total genotyping rate in remaining samples is 0.999535.\n",
      "547812 variants and 5164 people pass filters and QC.\n",
      "Phenotype data is quantitative.\n",
      "--make-bed to clean-merged_neighbor.bed + clean-merged_neighbor.bim +\n",
      "clean-merged_neighbor.fam ... 0%\b\b1%\b\b2%\b\b3%\b\b4%\b\b5%\b\b6%\b\b7%\b\b8%\b\b9%\b\b10%\b\b\b11%\b\b\b12%\b\b\b13%\b\b\b14%\b\b\b15%\b\b\b16%\b\b\b17%\b\b\b18%\b\b\b19%\b\b\b20%\b\b\b21%\b\b\b22%\b\b\b23%\b\b\b24%\b\b\b25%\b\b\b26%\b\b\b27%\b\b\b28%\b\b\b29%\b\b\b30%\b\b\b31%\b\b\b32%\b\b\b33%\b\b\b34%\b\b\b35%\b\b\b36%\b\b\b37%\b\b\b38%\b\b\b39%\b\b\b40%\b\b\b41%\b\b\b42%\b\b\b43%\b\b\b44%\b\b\b45%\b\b\b46%\b\b\b47%\b\b\b48%\b\b\b49%\b\b\b50%\b\b\b51%\b\b\b52%\b\b\b53%\b\b\b54%\b\b\b55%\b\b\b56%\b\b\b57%\b\b\b58%\b\b\b59%\b\b\b60%\b\b\b61%\b\b\b62%\b\b\b63%\b\b\b64%\b\b\b65%\b\b\b66%\b\b\b67%\b\b\b68%\b\b\b69%\b\b\b70%\b\b\b71%\b\b\b72%\b\b\b73%\b\b\b74%\b\b\b75%\b\b\b76%\b\b\b77%\b\b\b78%\b\b\b79%\b\b\b80%\b\b\b81%\b\b\b82%\b\b\b83%\b\b\b84%\b\b\b85%\b\b\b86%\b\b\b87%\b\b\b88%\b\b\b89%\b\b\b90%\b\b\b91%\b\b\b92%\b\b\b93%\b\b\b94%\b\b\b95%\b\b\b96%\b\b\b97%\b\b\b98%\b\b\b99%\b\b\bdone.\n"
     ]
    }
   ],
   "source": [
    "'''Use plink to exclude those samples that failed callrate QC.'''\n",
    "!plink --bfile $output_filestem --remove fail-imiss-qc.txt --make-bed --out $output_filestem"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "PLINK v1.90b3.45 64-bit (13 Jan 2017)      https://www.cog-genomics.org/plink2\n",
      "(C) 2005-2017 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to clean-merged_neighbor.log.\n",
      "Options in effect:\n",
      "  --bfile clean-merged_neighbor\n",
      "  --maf 0.01\n",
      "  --make-bed\n",
      "  --out clean-merged_neighbor\n",
      "\n",
      "995 MB RAM detected; reserving 497 MB for main workspace.\n",
      "Note: --make-bed input and output filenames match.  Appending '~' to input\n",
      "filenames.\n",
      "547812 variants loaded from .bim file.\n",
      "5164 people (2291 males, 2742 females, 131 ambiguous) loaded from .fam.\n",
      "Ambiguous sex IDs written to clean-merged_neighbor.nosex .\n",
      "5033 phenotype values loaded from .fam.\n",
      "Using 1 thread.\n",
      "Before main variant filters, 5164 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 0%\b\b1%\b\b2%\b\b3%\b\b4%\b\b5%\b\b6%\b\b7%\b\b8%\b\b9%\b\b10%\b\b\b11%\b\b\b12%\b\b\b13%\b\b\b14%\b\b\b15%\b\b\b16%\b\b\b17%\b\b\b18%\b\b\b19%\b\b\b20%\b\b\b21%\b\b\b22%\b\b\b23%\b\b\b24%\b\b\b25%\b\b\b26%\b\b\b27%\b\b\b28%\b\b\b29%\b\b\b30%\b\b\b31%\b\b\b32%\b\b\b33%\b\b\b34%\b\b\b35%\b\b\b36%\b\b\b37%\b\b\b38%\b\b\b39%\b\b\b40%\b\b\b41%\b\b\b42%\b\b\b43%\b\b\b44%\b\b\b45%\b\b\b46%\b\b\b47%\b\b\b48%\b\b\b49%\b\b\b50%\b\b\b51%\b\b\b52%\b\b\b53%\b\b\b54%\b\b\b55%\b\b\b56%\b\b\b57%\b\b\b58%\b\b\b59%\b\b\b60%\b\b\b61%\b\b\b62%\b\b\b63%\b\b\b64%\b\b\b65%\b\b\b66%\b\b\b67%\b\b\b68%\b\b\b69%\b\b\b70%\b\b\b71%\b\b\b72%\b\b\b73%\b\b\b74%\b\b\b75%\b\b\b76%\b\b\b77%\b\b\b78%\b\b\b79%\b\b\b80%\b\b\b81%\b\b\b82%\b\b\b83%\b\b\b84%\b\b\b85%\b\b\b86%\b\b\b87%\b\b\b88%\b\b\b89%\b\b\b90%\b\b\b91%\b\b\b92%\b\b\b93%\b\b\b94%\b\b\b95%\b\b\b96%\b\b\b97%\b\b\b98%\b\b\b99%\b\b\b\b done.\n",
      "Warning: 11469 het. haploid genotypes present (see clean-merged_neighbor.hh );\n",
      "many commands treat these as missing.\n",
      "Warning: Nonmissing nonmale Y chromosome genotype(s) present; many commands\n",
      "treat these as missing.\n",
      "Total genotyping rate is 0.999535.\n",
      "19345 variants removed due to minor allele threshold(s)\n",
      "(--maf/--max-maf/--mac/--max-mac).\n",
      "528467 variants and 5164 people pass filters and QC.\n",
      "Phenotype data is quantitative.\n",
      "--make-bed to clean-merged_neighbor.bed + clean-merged_neighbor.bim +\n",
      "clean-merged_neighbor.fam ... 0%\b\b1%\b\b2%\b\b3%\b\b4%\b\b5%\b\b6%\b\b7%\b\b8%\b\b9%\b\b10%\b\b\b11%\b\b\b12%\b\b\b13%\b\b\b14%\b\b\b15%\b\b\b16%\b\b\b17%\b\b\b18%\b\b\b19%\b\b\b20%\b\b\b21%\b\b\b22%\b\b\b23%\b\b\b24%\b\b\b25%\b\b\b26%\b\b\b27%\b\b\b28%\b\b\b29%\b\b\b30%\b\b\b31%\b\b\b32%\b\b\b33%\b\b\b34%\b\b\b35%\b\b\b36%\b\b\b37%\b\b\b38%\b\b\b39%\b\b\b40%\b\b\b41%\b\b\b42%\b\b\b43%\b\b\b44%\b\b\b45%\b\b\b46%\b\b\b47%\b\b\b48%\b\b\b49%\b\b\b50%\b\b\b51%\b\b\b52%\b\b\b53%\b\b\b54%\b\b\b55%\b\b\b56%\b\b\b57%\b\b\b58%\b\b\b59%\b\b\b60%\b\b\b61%\b\b\b62%\b\b\b63%\b\b\b64%\b\b\b65%\b\b\b66%\b\b\b67%\b\b\b68%\b\b\b69%\b\b\b70%\b\b\b71%\b\b\b72%\b\b\b73%\b\b\b74%\b\b\b75%\b\b\b76%\b\b\b77%\b\b\b78%\b\b\b79%\b\b\b80%\b\b\b81%\b\b\b82%\b\b\b83%\b\b\b84%\b\b\b85%\b\b\b86%\b\b\b87%\b\b\b88%\b\b\b89%\b\b\b90%\b\b\b91%\b\b\b92%\b\b\b93%\b\b\b94%\b\b\b95%\b\b\b96%\b\b\b97%\b\b\b98%\b\b\b99%\b\b\bdone.\n"
     ]
    }
   ],
   "source": [
    "'''Exclude rare variants: genetic markers with less than 1% minor allele frequency.'''\n",
    "!plink --bfile $output_filestem --maf 0.01 --make-bed --out $output_filestem"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "PLINK v1.90b3.45 64-bit (13 Jan 2017)      https://www.cog-genomics.org/plink2\n",
      "(C) 2005-2017 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to clean-merged_neighbor.log.\n",
      "Options in effect:\n",
      "  --bfile clean-merged_neighbor\n",
      "  --indep-pairwise 50 5 0.2\n",
      "  --out clean-merged_neighbor\n",
      "\n",
      "995 MB RAM detected; reserving 497 MB for main workspace.\n",
      "528467 variants loaded from .bim file.\n",
      "5164 people (2291 males, 2742 females, 131 ambiguous) loaded from .fam.\n",
      "Ambiguous sex IDs written to clean-merged_neighbor.nosex .\n",
      "5033 phenotype values loaded from .fam.\n",
      "Using 1 thread.\n",
      "Before main variant filters, 5164 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 0%\b\b1%\b\b2%\b\b3%\b\b4%\b\b5%\b\b6%\b\b7%\b\b8%\b\b9%\b\b10%\b\b\b11%\b\b\b12%\b\b\b13%\b\b\b14%\b\b\b15%\b\b\b16%\b\b\b17%\b\b\b18%\b\b\b19%\b\b\b20%\b\b\b21%\b\b\b22%\b\b\b23%\b\b\b24%\b\b\b25%\b\b\b26%\b\b\b27%\b\b\b28%\b\b\b29%\b\b\b30%\b\b\b31%\b\b\b32%\b\b\b33%\b\b\b34%\b\b\b35%\b\b\b36%\b\b\b37%\b\b\b38%\b\b\b39%\b\b\b40%\b\b\b41%\b\b\b42%\b\b\b43%\b\b\b44%\b\b\b45%\b\b\b46%\b\b\b47%\b\b\b48%\b\b\b49%\b\b\b50%\b\b\b51%\b\b\b52%\b\b\b53%\b\b\b54%\b\b\b55%\b\b\b56%\b\b\b57%\b\b\b58%\b\b\b59%\b\b\b60%\b\b\b61%\b\b\b62%\b\b\b63%\b\b\b64%\b\b\b65%\b\b\b66%\b\b\b67%\b\b\b68%\b\b\b69%\b\b\b70%\b\b\b71%\b\b\b72%\b\b\b73%\b\b\b74%\b\b\b75%\b\b\b76%\b\b\b77%\b\b\b78%\b\b\b79%\b\b\b80%\b\b\b81%\b\b\b82%\b\b\b83%\b\b\b84%\b\b\b85%\b\b\b86%\b\b\b87%\b\b\b88%\b\b\b89%\b\b\b90%\b\b\b91%\b\b\b92%\b\b\b93%\b\b\b94%\b\b\b95%\b\b\b96%\b\b\b97%\b\b\b98%\b\b\b99%\b\b\b\b done.\n",
      "Warning: 11452 het. haploid genotypes present (see clean-merged_neighbor.hh );\n",
      "many commands treat these as missing.\n",
      "Warning: Nonmissing nonmale Y chromosome genotype(s) present; many commands\n",
      "treat these as missing.\n",
      "Total genotyping rate is 0.999529.\n",
      "528467 variants and 5164 people pass filters and QC.\n",
      "Phenotype data is quantitative.\n",
      "Pruned 31484 variants from chromosome 1, leaving 8830.\n",
      "Pruned 34454 variants from chromosome 2, leaving 8428.\n",
      "Pruned 28405 variants from chromosome 3, leaving 7255.\n",
      "Pruned 24781 variants from chromosome 4, leaving 6599.\n",
      "Pruned 25657 variants from chromosome 5, leaving 6630.\n",
      "Pruned 27776 variants from chromosome 6, leaving 6779.\n",
      "Pruned 22693 variants from chromosome 7, leaving 5918.\n",
      "Pruned 23939 variants from chromosome 8, leaving 5526.\n",
      "Pruned 20166 variants from chromosome 9, leaving 5139.\n",
      "Pruned 21924 variants from chromosome 10, leaving 5749.\n",
      "Pruned 20411 variants from chromosome 11, leaving 5244.\n",
      "Pruned 20101 variants from chromosome 12, leaving 5569.\n",
      "Pruned 15707 variants from chromosome 13, leaving 4149.\n",
      "Pruned 13646 variants from chromosome 14, leaving 3758.\n",
      "Pruned 12159 variants from chromosome 15, leaving 3527.\n",
      "Pruned 12140 variants from chromosome 16, leaving 3822.\n",
      "Pruned 10174 variants from chromosome 17, leaving 3598.\n",
      "Pruned 12144 variants from chromosome 18, leaving 3605.\n",
      "Pruned 6319 variants from chromosome 19, leaving 2740.\n",
      "Pruned 10164 variants from chromosome 20, leaving 3237.\n",
      "Pruned 6006 variants from chromosome 21, leaving 1808.\n",
      "Pruned 5868 variants from chromosome 22, leaving 1977.\n",
      "Pruned 9560 variants from chromosome 23, leaving 2819.\n",
      "Pruned 2 variants from chromosome 24, leaving 2.\n",
      "Pruned 0 variants from chromosome 25, leaving 3.\n",
      "Pruned 43 variants from chromosome 26, leaving 33.\n",
      "Pruning complete.  415723 of 528467 variants removed.\n",
      "Marker lists written to clean-merged_neighbor.prune.in and\n",
      "clean-merged_neighbor.prune.out .\n"
     ]
    }
   ],
   "source": [
    "'''Exclude related samples.'''\n",
    "'''Generate LD-pruned variants first.'''\n",
    "!plink --bfile $output_filestem --indep-pairwise 50 5 0.2 --out $output_filestem"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Unlinked (LD-pruned) genetic variant count: ['112744']\n",
      "PLINK v1.90b3.45 64-bit (13 Jan 2017)      https://www.cog-genomics.org/plink2\n",
      "(C) 2005-2017 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to clean-merged_neighbor.log.\n",
      "Options in effect:\n",
      "  --bfile clean-merged_neighbor\n",
      "  --extract clean-merged_neighbor.prune.in\n",
      "  --genome\n",
      "  --out clean-merged_neighbor\n",
      "\n",
      "995 MB RAM detected; reserving 497 MB for main workspace.\n",
      "528467 variants loaded from .bim file.\n",
      "5164 people (2291 males, 2742 females, 131 ambiguous) loaded from .fam.\n",
      "Ambiguous sex IDs written to clean-merged_neighbor.nosex .\n",
      "5033 phenotype values loaded from .fam.\n",
      "--extract: 112744 variants remaining.\n",
      "Using 1 thread.\n",
      "Before main variant filters, 5164 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 0%\b\b1%\b\b2%\b\b3%\b\b4%\b\b5%\b\b6%\b\b7%\b\b8%\b\b9%\b\b10%\b\b\b11%\b\b\b12%\b\b\b13%\b\b\b14%\b\b\b15%\b\b\b16%\b\b\b17%\b\b\b18%\b\b\b19%\b\b\b20%\b\b\b21%\b\b\b22%\b\b\b23%\b\b\b24%\b\b\b25%\b\b\b26%\b\b\b27%\b\b\b28%\b\b\b29%\b\b\b30%\b\b\b31%\b\b\b32%\b\b\b33%\b\b\b34%\b\b\b35%\b\b\b36%\b\b\b37%\b\b\b38%\b\b\b39%\b\b\b40%\b\b\b41%\b\b\b42%\b\b\b43%\b\b\b44%\b\b\b45%\b\b\b46%\b\b\b47%\b\b\b48%\b\b\b49%\b\b\b50%\b\b\b51%\b\b\b52%\b\b\b53%\b\b\b54%\b\b\b55%\b\b\b56%\b\b\b57%\b\b\b58%\b\b\b59%\b\b\b60%\b\b\b61%\b\b\b62%\b\b\b63%\b\b\b64%\b\b\b65%\b\b\b66%\b\b\b67%\b\b\b68%\b\b\b69%\b\b\b70%\b\b\b71%\b\b\b72%\b\b\b73%\b\b\b74%\b\b\b75%\b\b\b76%\b\b\b77%\b\b\b78%\b\b\b79%\b\b\b80%\b\b\b81%\b\b\b82%\b\b\b83%\b\b\b84%\b\b\b85%\b\b\b86%\b\b\b87%\b\b\b88%\b\b\b89%\b\b\b90%\b\b\b91%\b\b\b92%\b\b\b93%\b\b\b94%\b\b\b95%\b\b\b96%\b\b\b97%\b\b\b98%\b\b\b99%\b\b\b\b done.\n",
      "Warning: 2513 het. haploid genotypes present (see clean-merged_neighbor.hh );\n",
      "many commands treat these as missing.\n",
      "Warning: Nonmissing nonmale Y chromosome genotype(s) present; many commands\n",
      "treat these as missing.\n",
      "Total genotyping rate is 0.999529.\n",
      "112744 variants and 5164 people pass filters and QC.\n",
      "Phenotype data is quantitative.\n",
      "Excluding 2854 variants on non-autosomes from IBD calculation.\n",
      "IBD calculations complete.  \n",
      "Finished writing clean-merged_neighbor.genome .\n"
     ]
    }
   ],
   "source": [
    "'''If the above output pruned too few or too many SNPs, adjust the linkage R^2 value \n",
    "Increase R^2 to retain more SNPs, decrease to prune more.\n",
    "The goal is to end up with about 100k SNPs for a GWAS study (>= 500k SNPs genome-wide).\n",
    "Then proceed with identification of related samples.'''\n",
    "prune_in_file = output_filestem + '.prune.in'\n",
    "unlinked_variant_count = !cat $prune_in_file | wc -l\n",
    "print \"Unlinked (LD-pruned) genetic variant count: \" + str(unlinked_variant_count)\n",
    "!plink --bfile $output_filestem --extract $prune_in_file --genome --out $output_filestem\n",
    "#That command should generate a *.genome file."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Line count in genome file (used for relationship estimation): ['13330867 clean-merged_neighbor.genome']\n",
      "FID1\tIID1\tFID2\tIID2\tRT\tEZ\tZ0\tZ1\tZ2\tPI_HAT\tPHE\tDST\tPPC\tRATIO\r\n",
      "ME8001\tME8001\tME8002\tME8002\tUN\tNA\t1.0000\t0.0000\t0.0000\t0.0000\tNA\t0.745242\t0.0535\t1.8941\r\n",
      "ME8001\tME8001\tME8003\tME8003\tUN\tNA\t1.0000\t0.0000\t0.0000\t0.0000\tNA\t0.744817\t0.0055\t1.8363\r\n",
      "ME8001\tME8001\tME8004\tME8004\tUN\tNA\t1.0000\t0.0000\t0.0000\t0.0000\tNA\t0.746071\t0.5046\t2.0008\r\n",
      "ME8001\tME8001\tME8005\tME8005\tUN\tNA\t1.0000\t0.0000\t0.0000\t0.0000\tNA\t0.747932\t0.0500\t1.8920\r\n",
      "ME8001\tME8001\tME8006\tME8006\tUN\tNA\t1.0000\t0.0000\t0.0000\t0.0000\tNA\t0.746616\t0.2204\t1.9484\r\n",
      "ME8001\tME8001\tME8007\tME8007\tUN\tNA\t1.0000\t0.0000\t0.0000\t0.0000\tNA\t0.747042\t0.2199\t1.9482\r\n",
      "ME8001\tME8001\tME8008\tME8008\tUN\tNA\t1.0000\t0.0000\t0.0000\t0.0000\tNA\t0.745406\t0.0870\t1.9104\r\n",
      "ME8001\tME8001\tME8009\tME8009\tUN\tNA\t1.0000\t0.0000\t0.0000\t0.0000\tNA\t0.744062\t0.0177\t1.8630\r\n",
      "ME8001\tME8001\tME8010\tME8010\tUN\tNA\t1.0000\t0.0000\t0.0000\t0.0000\tNA\t0.745033\t0.0328\t1.8795\r\n"
     ]
    }
   ],
   "source": [
    "genome_file = output_filestem + '.genome'\n",
    "#Clean up the whitespace:\n",
    "!perl -i.bak -lpe 's/^\\s+//; s/\\s+/\\t/g;' $genome_file\n",
    "genome_file_line_count = !wc -l $genome_file\n",
    "print \"Line count in genome file (used for relationship estimation): \" + str(genome_file_line_count)\n",
    "!head $genome_file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of samples that failed IBD QC: ['193']\n"
     ]
    }
   ],
   "source": [
    "'''Generate a file containing FID:IID pairs of individuals to be excluded on the basis of relatedness.\n",
    "Exclude the individual with greater missingness from any pair of individuals that are related more than first cousins.'''\n",
    "imiss, removed = dict(), set()\n",
    "\n",
    "with open(imiss_file) as imiss_in, open(genome_file) as genome_in, open('fail-IBD-QC.txt', 'w') as fail_IBD_out:\n",
    "    #Read the missingness into a dictionary.\n",
    "    header = imiss_in.readline()\n",
    "    for line in imiss_in:\n",
    "        fid, iid, miss_pheno, n_miss, n_geno, f_miss = line.rstrip().split(\"\\t\") #File was already whitespace-cleaned.\n",
    "        imiss[(fid, iid)] = float(f_miss) #Key by FID:IID tuple, value by F_MISS (missing frequency).\n",
    "        \n",
    "    #Now exclude related samples. Drop whichever one has higher missingness.\n",
    "    header = genome_in.readline()\n",
    "    for line in genome_in:\n",
    "        fid1, iid1, fid2, iid2, rt, ez, z0, z1, z2, pi_hat, phe, dst, ppc, ratio = line.rstrip().split(\"\\t\")\n",
    "        if float(pi_hat) > 0.185:\n",
    "            if imiss[(fid1, iid1)] >= imiss[(fid2, iid2)]: #The first sample has higher missingness.\n",
    "                if (fid1, iid1) not in removed:\n",
    "                    fail_IBD_out.write(\"\\t\".join([fid1, iid1]) + \"\\n\")\n",
    "                    removed.add((fid1, iid1)) #Remember that we already excluded this FID:IID pair.\n",
    "            else: #The second sample has higher missingness.\n",
    "                if (fid2, iid2) not in removed:\n",
    "                    fail_IBD_out.write(\"\\t\".join([fid2, iid2]) + \"\\n\")\n",
    "                    removed.add((fid2, iid2))\n",
    "\n",
    "fail_IBD_count = !cat fail-IBD-QC.txt | wc -l\n",
    "print \"Number of samples that failed IBD QC: \" + str(fail_IBD_count)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "PLINK v1.90b3.45 64-bit (13 Jan 2017)      https://www.cog-genomics.org/plink2\n",
      "(C) 2005-2017 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to clean-merged_neighbor.log.\n",
      "Options in effect:\n",
      "  --bfile clean-merged_neighbor\n",
      "  --make-bed\n",
      "  --out clean-merged_neighbor\n",
      "  --remove fail-IBD-QC.txt\n",
      "\n",
      "995 MB RAM detected; reserving 497 MB for main workspace.\n",
      "Note: --make-bed input and output filenames match.  Appending '~' to input\n",
      "filenames.\n",
      "528467 variants loaded from .bim file.\n",
      "5164 people (2291 males, 2742 females, 131 ambiguous) loaded from .fam.\n",
      "Ambiguous sex IDs written to clean-merged_neighbor.nosex .\n",
      "5033 phenotype values loaded from .fam.\n",
      "--remove: 4971 people remaining.\n",
      "Using 1 thread.\n",
      "Before main variant filters, 4971 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 0%\b\b1%\b\b2%\b\b3%\b\b4%\b\b5%\b\b6%\b\b7%\b\b8%\b\b9%\b\b10%\b\b\b11%\b\b\b12%\b\b\b13%\b\b\b14%\b\b\b15%\b\b\b16%\b\b\b17%\b\b\b18%\b\b\b19%\b\b\b20%\b\b\b21%\b\b\b22%\b\b\b23%\b\b\b24%\b\b\b25%\b\b\b26%\b\b\b27%\b\b\b28%\b\b\b29%\b\b\b30%\b\b\b31%\b\b\b32%\b\b\b33%\b\b\b34%\b\b\b35%\b\b\b36%\b\b\b37%\b\b\b38%\b\b\b39%\b\b\b40%\b\b\b41%\b\b\b42%\b\b\b43%\b\b\b44%\b\b\b45%\b\b\b46%\b\b\b47%\b\b\b48%\b\b\b49%\b\b\b50%\b\b\b51%\b\b\b52%\b\b\b53%\b\b\b54%\b\b\b55%\b\b\b56%\b\b\b57%\b\b\b58%\b\b\b59%\b\b\b60%\b\b\b61%\b\b\b62%\b\b\b63%\b\b\b64%\b\b\b65%\b\b\b66%\b\b\b67%\b\b\b68%\b\b\b69%\b\b\b70%\b\b\b71%\b\b\b72%\b\b\b73%\b\b\b74%\b\b\b75%\b\b\b76%\b\b\b77%\b\b\b78%\b\b\b79%\b\b\b80%\b\b\b81%\b\b\b82%\b\b\b83%\b\b\b84%\b\b\b85%\b\b\b86%\b\b\b87%\b\b\b88%\b\b\b89%\b\b\b90%\b\b\b91%\b\b\b92%\b\b\b93%\b\b\b94%\b\b\b95%\b\b\b96%\b\b\b97%\b\b\b98%\b\b\b99%\b\b\b\b done.\n",
      "Warning: 11441 het. haploid genotypes present (see clean-merged_neighbor.hh );\n",
      "many commands treat these as missing.\n",
      "Warning: Nonmissing nonmale Y chromosome genotype(s) present; many commands\n",
      "treat these as missing.\n",
      "Total genotyping rate in remaining samples is 0.99954.\n",
      "528467 variants and 4971 people pass filters and QC.\n",
      "Phenotype data is quantitative.\n",
      "--make-bed to clean-merged_neighbor.bed + clean-merged_neighbor.bim +\n",
      "clean-merged_neighbor.fam ... 0%\b\b1%\b\b2%\b\b3%\b\b4%\b\b5%\b\b6%\b\b7%\b\b8%\b\b9%\b\b10%\b\b\b11%\b\b\b12%\b\b\b13%\b\b\b14%\b\b\b15%\b\b\b16%\b\b\b17%\b\b\b18%\b\b\b19%\b\b\b20%\b\b\b21%\b\b\b22%\b\b\b23%\b\b\b24%\b\b\b25%\b\b\b26%\b\b\b27%\b\b\b28%\b\b\b29%\b\b\b30%\b\b\b31%\b\b\b32%\b\b\b33%\b\b\b34%\b\b\b35%\b\b\b36%\b\b\b37%\b\b\b38%\b\b\b39%\b\b\b40%\b\b\b41%\b\b\b42%\b\b\b43%\b\b\b44%\b\b\b45%\b\b\b46%\b\b\b47%\b\b\b48%\b\b\b49%\b\b\b50%\b\b\b51%\b\b\b52%\b\b\b53%\b\b\b54%\b\b\b55%\b\b\b56%\b\b\b57%\b\b\b58%\b\b\b59%\b\b\b60%\b\b\b61%\b\b\b62%\b\b\b63%\b\b\b64%\b\b\b65%\b\b\b66%\b\b\b67%\b\b\b68%\b\b\b69%\b\b\b70%\b\b\b71%\b\b\b72%\b\b\b73%\b\b\b74%\b\b\b75%\b\b\b76%\b\b\b77%\b\b\b78%\b\b\b79%\b\b\b80%\b\b\b81%\b\b\b82%\b\b\b83%\b\b\b84%\b\b\b85%\b\b\b86%\b\b\b87%\b\b\b88%\b\b\b89%\b\b\b90%\b\b\b91%\b\b\b92%\b\b\b93%\b\b\b94%\b\b\b95%\b\b\b96%\b\b\b97%\b\b\b98%\b\b\b99%\b\b\bdone.\n"
     ]
    }
   ],
   "source": [
    "'''Now exclude the related samples from the study.'''\n",
    "!plink --bfile $output_filestem --remove fail-IBD-QC.txt --make-bed --out $output_filestem"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "PLINK v1.90b3.45 64-bit (13 Jan 2017)      https://www.cog-genomics.org/plink2\n",
      "(C) 2005-2017 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to clean-merged_neighbor.log.\n",
      "Options in effect:\n",
      "  --bfile clean-merged_neighbor\n",
      "  --geno 0.01\n",
      "  --make-bed\n",
      "  --out clean-merged_neighbor\n",
      "\n",
      "995 MB RAM detected; reserving 497 MB for main workspace.\n",
      "Note: --make-bed input and output filenames match.  Appending '~' to input\n",
      "filenames.\n",
      "528467 variants loaded from .bim file.\n",
      "4971 people (2245 males, 2655 females, 71 ambiguous) loaded from .fam.\n",
      "Ambiguous sex IDs written to clean-merged_neighbor.nosex .\n",
      "4900 phenotype values loaded from .fam.\n",
      "Using 1 thread.\n",
      "Before main variant filters, 4971 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 0%\b\b1%\b\b2%\b\b3%\b\b4%\b\b5%\b\b6%\b\b7%\b\b8%\b\b9%\b\b10%\b\b\b11%\b\b\b12%\b\b\b13%\b\b\b14%\b\b\b15%\b\b\b16%\b\b\b17%\b\b\b18%\b\b\b19%\b\b\b20%\b\b\b21%\b\b\b22%\b\b\b23%\b\b\b24%\b\b\b25%\b\b\b26%\b\b\b27%\b\b\b28%\b\b\b29%\b\b\b30%\b\b\b31%\b\b\b32%\b\b\b33%\b\b\b34%\b\b\b35%\b\b\b36%\b\b\b37%\b\b\b38%\b\b\b39%\b\b\b40%\b\b\b41%\b\b\b42%\b\b\b43%\b\b\b44%\b\b\b45%\b\b\b46%\b\b\b47%\b\b\b48%\b\b\b49%\b\b\b50%\b\b\b51%\b\b\b52%\b\b\b53%\b\b\b54%\b\b\b55%\b\b\b56%\b\b\b57%\b\b\b58%\b\b\b59%\b\b\b60%\b\b\b61%\b\b\b62%\b\b\b63%\b\b\b64%\b\b\b65%\b\b\b66%\b\b\b67%\b\b\b68%\b\b\b69%\b\b\b70%\b\b\b71%\b\b\b72%\b\b\b73%\b\b\b74%\b\b\b75%\b\b\b76%\b\b\b77%\b\b\b78%\b\b\b79%\b\b\b80%\b\b\b81%\b\b\b82%\b\b\b83%\b\b\b84%\b\b\b85%\b\b\b86%\b\b\b87%\b\b\b88%\b\b\b89%\b\b\b90%\b\b\b91%\b\b\b92%\b\b\b93%\b\b\b94%\b\b\b95%\b\b\b96%\b\b\b97%\b\b\b98%\b\b\b99%\b\b\b\b done.\n",
      "Warning: 11441 het. haploid genotypes present (see clean-merged_neighbor.hh );\n",
      "many commands treat these as missing.\n",
      "Warning: Nonmissing nonmale Y chromosome genotype(s) present; many commands\n",
      "treat these as missing.\n",
      "Total genotyping rate is 0.99954.\n",
      "25 variants removed due to missing genotype data (--geno).\n",
      "528442 variants and 4971 people pass filters and QC.\n",
      "Phenotype data is quantitative.\n",
      "--make-bed to clean-merged_neighbor.bed + clean-merged_neighbor.bim +\n",
      "clean-merged_neighbor.fam ... 0%\b\b1%\b\b2%\b\b3%\b\b4%\b\b5%\b\b6%\b\b7%\b\b8%\b\b9%\b\b10%\b\b\b11%\b\b\b12%\b\b\b13%\b\b\b14%\b\b\b15%\b\b\b16%\b\b\b17%\b\b\b18%\b\b\b19%\b\b\b20%\b\b\b21%\b\b\b22%\b\b\b23%\b\b\b24%\b\b\b25%\b\b\b26%\b\b\b27%\b\b\b28%\b\b\b29%\b\b\b30%\b\b\b31%\b\b\b32%\b\b\b33%\b\b\b34%\b\b\b35%\b\b\b36%\b\b\b37%\b\b\b38%\b\b\b39%\b\b\b40%\b\b\b41%\b\b\b42%\b\b\b43%\b\b\b44%\b\b\b45%\b\b\b46%\b\b\b47%\b\b\b48%\b\b\b49%\b\b\b50%\b\b\b51%\b\b\b52%\b\b\b53%\b\b\b54%\b\b\b55%\b\b\b56%\b\b\b57%\b\b\b58%\b\b\b59%\b\b\b60%\b\b\b61%\b\b\b62%\b\b\b63%\b\b\b64%\b\b\b65%\b\b\b66%\b\b\b67%\b\b\b68%\b\b\b69%\b\b\b70%\b\b\b71%\b\b\b72%\b\b\b73%\b\b\b74%\b\b\b75%\b\b\b76%\b\b\b77%\b\b\b78%\b\b\b79%\b\b\b80%\b\b\b81%\b\b\b82%\b\b\b83%\b\b\b84%\b\b\b85%\b\b\b86%\b\b\b87%\b\b\b88%\b\b\b89%\b\b\b90%\b\b\b91%\b\b\b92%\b\b\b93%\b\b\b94%\b\b\b95%\b\b\b96%\b\b\b97%\b\b\b98%\b\b\b99%\b\b\bdone.\n"
     ]
    }
   ],
   "source": [
    "'''Now that related samples have been excluded, repeat the marker callrate, sample callrate, and MAF filtering.'''\n",
    "\n",
    "'''Exclude genetic variants that are not present in at least 99% of samples.'''\n",
    "!plink --bfile $output_filestem --geno 0.01 --make-bed --out $output_filestem"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "PLINK v1.90b3.45 64-bit (13 Jan 2017)      https://www.cog-genomics.org/plink2\n",
      "(C) 2005-2017 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to clean-merged_neighbor.log.\n",
      "Options in effect:\n",
      "  --bfile clean-merged_neighbor\n",
      "  --missing\n",
      "  --out clean-merged_neighbor\n",
      "\n",
      "995 MB RAM detected; reserving 497 MB for main workspace.\n",
      "528442 variants loaded from .bim file.\n",
      "4971 people (2245 males, 2655 females, 71 ambiguous) loaded from .fam.\n",
      "Ambiguous sex IDs written to clean-merged_neighbor.nosex .\n",
      "4900 phenotype values loaded from .fam.\n",
      "Using 1 thread.\n",
      "Before main variant filters, 4971 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 0%\b\b1%\b\b2%\b\b3%\b\b4%\b\b5%\b\b6%\b\b7%\b\b8%\b\b9%\b\b10%\b\b\b11%\b\b\b12%\b\b\b13%\b\b\b14%\b\b\b15%\b\b\b16%\b\b\b17%\b\b\b18%\b\b\b19%\b\b\b20%\b\b\b21%\b\b\b22%\b\b\b23%\b\b\b24%\b\b\b25%\b\b\b26%\b\b\b27%\b\b\b28%\b\b\b29%\b\b\b30%\b\b\b31%\b\b\b32%\b\b\b33%\b\b\b34%\b\b\b35%\b\b\b36%\b\b\b37%\b\b\b38%\b\b\b39%\b\b\b40%\b\b\b41%\b\b\b42%\b\b\b43%\b\b\b44%\b\b\b45%\b\b\b46%\b\b\b47%\b\b\b48%\b\b\b49%\b\b\b50%\b\b\b51%\b\b\b52%\b\b\b53%\b\b\b54%\b\b\b55%\b\b\b56%\b\b\b57%\b\b\b58%\b\b\b59%\b\b\b60%\b\b\b61%\b\b\b62%\b\b\b63%\b\b\b64%\b\b\b65%\b\b\b66%\b\b\b67%\b\b\b68%\b\b\b69%\b\b\b70%\b\b\b71%\b\b\b72%\b\b\b73%\b\b\b74%\b\b\b75%\b\b\b76%\b\b\b77%\b\b\b78%\b\b\b79%\b\b\b80%\b\b\b81%\b\b\b82%\b\b\b83%\b\b\b84%\b\b\b85%\b\b\b86%\b\b\b87%\b\b\b88%\b\b\b89%\b\b\b90%\b\b\b91%\b\b\b92%\b\b\b93%\b\b\b94%\b\b\b95%\b\b\b96%\b\b\b97%\b\b\b98%\b\b\b99%\b\b\b\b done.\n",
      "Warning: 11436 het. haploid genotypes present (see clean-merged_neighbor.hh );\n",
      "many commands treat these as missing.\n",
      "Warning: Nonmissing nonmale Y chromosome genotype(s) present; many commands\n",
      "treat these as missing.\n",
      "Total genotyping rate is 0.999541.\n",
      "--missing: Sample missing data report written to clean-merged_neighbor.imiss,\n",
      "and variant-based missing data report written to clean-merged_neighbor.lmiss.\n"
     ]
    }
   ],
   "source": [
    "'''Exclude individuals with less than 99% callrate (1% or more of genetic variants are missing in that sample).'''\n",
    "#First, generate missingness reports:\n",
    "!plink --bfile $output_filestem --missing --out $output_filestem\n",
    "imiss_file = output_filestem + '.imiss'\n",
    "#Clean up the whitespace:\n",
    "!perl -i.bak -lpe 's/^\\s+//; s/\\s+/\\t/g;' $imiss_file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
